// @doist/todoist-sdk/dist/cjs/transport/http-dispatcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultDispatcher = getDefaultDispatcher;
exports.resetDefaultDispatcherForTests = resetDefaultDispatcherForTests;
exports.suppressExperimentalWarningsSync = suppressExperimentalWarningsSync;
// Use effectively-disabled keep-alive so short-lived CLI processes do not stay
// open waiting on idle sockets. Undici requires positive values, so we use 1ms.
const KEEP_ALIVE_OPTIONS = {
    keepAliveTimeout: 1,
    keepAliveMaxTimeout: 1,
};
let defaultDispatcherPromise;
function getDefaultDispatcher() {
    if (!isNodeEnvironment()) {
        return Promise.resolve(undefined);
    }
    if (!defaultDispatcherPromise) {
        defaultDispatcherPromise = createDefaultDispatcher().catch((error) => {
            defaultDispatcherPromise = undefined;
            throw error;
        });
    }
    return defaultDispatcherPromise;
}
async function resetDefaultDispatcherForTests() {
    if (!defaultDispatcherPromise) {
        return;
    }
    const dispatcherPromise = defaultDispatcherPromise;
    defaultDispatcherPromise = undefined;
    await dispatcherPromise.then((dispatcher) => dispatcher.close(), () => undefined);
}
async function createDefaultDispatcher() {
    // Dynamic import so non-Node consumers (browser, edge runtimes) don't pull
    // undici into their bundle. `getDefaultDispatcher` already short-circuits
    // outside Node, so this branch only runs when undici is safe to load.
    const { EnvHttpProxyAgent, interceptors } = await Promise.resolve().then(() => __importStar(require('undici')));
    const agent = new EnvHttpProxyAgent(KEEP_ALIVE_OPTIONS);
    // Some runtimes report `process.versions.node` (so `isNodeEnvironment()`
    // passes) but ship only a partial undici: `interceptors.decompress` is
    // absent and dispatchers have no `.compose`. Bun is the common case. There
    // the proxy agent alone is enough — Bun's `fetch` decompresses
    // gzip/deflate/br/zstd natively — so skip the interceptor instead of
    // crashing on the missing API.
    if (typeof interceptors.decompress !== 'function') {
        return agent;
    }
    // Compose the response-decompression interceptor so gzip/deflate/br/zstd
    // bodies are decoded before consumers parse them. Required on Node 24+:
    // attaching any custom dispatcher to the global `fetch` strips the
    // `content-encoding` header but does not actually decompress the body,
    // so callers receive raw gzipped bytes and `JSON.parse` fails.
    // See https://github.com/Doist/todoist-cli/issues/318.
    const decompress = suppressExperimentalWarningsSync(() => interceptors.decompress());
    return agent.compose(decompress);
}
// undici emits an `ExperimentalWarning` the first time `interceptors.decompress()`
// runs. The interceptor is stable for our gzipped-JSON-over-HTTPS use case;
// silence the warning during dispatcher init only so it does not leak to every
// consumer's stderr on the first request.
//
// `fn` must be synchronous so the override covers a single critical section
// (microseconds) — no unrelated `ExperimentalWarning` from elsewhere can
// interleave and be lost. We suppress every `ExperimentalWarning` rather than
// pattern-matching the message text: the message wording is an undici
// implementation detail (not a stable API), and the suppression window is
// narrow enough that a coarse type filter is safe.
//
// Exported for direct unit testing — the integration path through
// `getDefaultDispatcher()` cannot reliably exercise the helper because both
// the dispatcher singleton and undici's internal `warningEmitted` flag are
// once-per-process.
function suppressExperimentalWarningsSync(fn) {
    const originalEmit = process.emitWarning;
    process.emitWarning = ((warning, typeOrOptions, ...rest) => {
        const type = typeof typeOrOptions === 'string'
            ? typeOrOptions
            : typeof typeOrOptions === 'object' && typeOrOptions !== null
                ? typeOrOptions.type
                : undefined;
        if (type === 'ExperimentalWarning')
            return;
        originalEmit.call(process, warning, typeOrOptions, ...rest);
    });
    try {
        return fn();
    }
    finally {
        process.emitWarning = originalEmit;
    }
}
function isNodeEnvironment() {
    var _a;
    return typeof process !== 'undefined' && Boolean((_a = process.versions) === null || _a === void 0 ? void 0 : _a.node);
}
