// @doist/todoist-sdk/dist/cjs/transport/sync-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.performSyncRequest = performSyncRequest;
const endpoints_1 = require("../consts/endpoints");
const types_1 = require("../types");
const sync_1 = require("../types/sync");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const http_client_1 = require("./http-client");
function serializeUserUpdateArgs(args) {
    return Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, args), (0, request_helpers_1.spreadIfDefined)(args.dateFormat, (v) => ({
        dateFormat: sync_1.DATE_FORMAT_TO_API[v],
    }))), (0, request_helpers_1.spreadIfDefined)(args.timeFormat, (v) => ({
        timeFormat: sync_1.TIME_FORMAT_TO_API[v],
    }))), (0, request_helpers_1.spreadIfDefined)(args.startDay, (v) => ({
        startDay: sync_1.DAY_OF_WEEK_TO_API[v],
    }))), (0, request_helpers_1.spreadIfDefined)(args.nextWeek, (v) => ({
        nextWeek: sync_1.DAY_OF_WEEK_TO_API[v],
    })));
}
function serializeTaskUpdateDateCompleteArgs(args) {
    return Object.assign(Object.assign(Object.assign({}, args), { isForward: args.isForward ? 1 : 0 }), (0, request_helpers_1.spreadIfDefined)(args.resetSubtasks, (v) => ({
        resetSubtasks: v ? 1 : 0,
    })));
}
function serializeUpdateGoalsArgs(args) {
    return Object.assign(Object.assign(Object.assign(Object.assign({}, args), (0, request_helpers_1.spreadIfDefined)(args.vacationMode, (v) => ({ vacationMode: v ? 1 : 0 }))), (0, request_helpers_1.spreadIfDefined)(args.karmaDisabled, (v) => ({
        karmaDisabled: v ? 1 : 0,
    }))), (0, request_helpers_1.spreadIfDefined)(args.ignoreDays, (v) => ({
        ignoreDays: v.map((day) => {
            const apiValue = sync_1.DAY_OF_WEEK_TO_API[day];
            if (apiValue === undefined) {
                throw new types_1.TodoistArgumentError(`Invalid ignoreDays entry "${String(day)}". Expected a DayOfWeek (e.g. "Monday").`);
            }
            return apiValue;
        }),
    })));
}
function preprocessSyncCommands(commands) {
    return commands.map((cmd) => {
        if (cmd.type === 'user_update')
            return Object.assign(Object.assign({}, cmd), { args: serializeUserUpdateArgs(cmd.args) });
        if (cmd.type === 'item_update_date_complete')
            return Object.assign(Object.assign({}, cmd), { args: serializeTaskUpdateDateCompleteArgs(cmd.args) });
        if (cmd.type === 'update_goals')
            return Object.assign(Object.assign({}, cmd), { args: serializeUpdateGoalsArgs(cmd.args) });
        return cmd;
    });
}
/**
 * Makes a request to the Sync API and handles error checking.
 *
 * @param ctx - HTTP identity (auth token, sync API base URI, optional custom fetch).
 * @param syncRequest - The sync request payload.
 * @param options - Optional request identifier and commands flag.
 * @returns The sync response data.
 * @throws TodoistRequestError if sync status contains errors.
 */
async function performSyncRequest(ctx, syncRequest, options = {}) {
    var _a;
    const { requestId, hasSyncCommands = false } = options;
    const processedRequest = ((_a = syncRequest.commands) === null || _a === void 0 ? void 0 : _a.length)
        ? Object.assign(Object.assign({}, syncRequest), { commands: preprocessSyncCommands(syncRequest.commands) }) : syncRequest;
    const response = await (0, http_client_1.request)({
        httpMethod: 'POST',
        baseUri: ctx.syncApiBase,
        relativePath: endpoints_1.ENDPOINT_SYNC,
        apiToken: ctx.authToken,
        customFetch: ctx.customFetch,
        payload: processedRequest,
        requestId: requestId,
        hasSyncCommands: hasSyncCommands,
    });
    // Check for sync errors and throw if any are found
    if (response.data.syncStatus) {
        Object.entries(response.data.syncStatus).forEach(([_, value]) => {
            if (value === 'ok')
                return;
            throw new types_1.TodoistRequestError(value.error, value.httpCode, value.errorExtra);
        });
    }
    return (0, validators_1.parseSyncResponse)(response.data);
}
