// @doist/todoist-sdk/dist/esm/types/projects/types.js
import { z } from 'zod';
import { getProjectUrl } from '../../utils/url-helpers.js';
/**
 * Base schema for all project types in Todoist.
 * Contains common fields shared between personal and workspace projects.
 */
export const BaseProjectSchema = z.object({
    id: z.string(),
    canAssignTasks: z.boolean(),
    childOrder: z.number().int(),
    color: z.string(),
    createdAt: z.coerce.date().nullable(),
    isArchived: z.boolean(),
    isDeleted: z.boolean(),
    isFavorite: z.boolean(),
    isFrozen: z.boolean(),
    name: z.string(),
    updatedAt: z.coerce.date().nullable(),
    viewStyle: z.string(),
    defaultOrder: z.number().int(),
    description: z.string(),
    isCollapsed: z.boolean(),
    isShared: z.boolean(),
});
/**
 * Schema for personal projects in Todoist.
 */
export const PersonalProjectSchema = BaseProjectSchema.extend({
    parentId: z.string().nullable(),
    inboxProject: z.boolean().optional().default(false),
}).transform((data) => {
    return Object.assign(Object.assign({}, data), { url: getProjectUrl(data.id, data.name) });
});
/** Available project visibility levels. */
export const PROJECT_VISIBILITIES = ['restricted', 'team', 'public'];
export const ProjectVisibilitySchema = z.enum(PROJECT_VISIBILITIES);
/**
 * Schema for workspace projects in Todoist.
 */
export const WorkspaceProjectSchema = BaseProjectSchema.extend({
    access: z.object({ visibility: ProjectVisibilitySchema }).optional(),
    collaboratorRoleDefault: z.string(),
    folderId: z.string().nullable(),
    isInviteOnly: z.boolean().nullable(),
    isLinkSharingEnabled: z.boolean(),
    role: z.string().nullable(),
    status: z.string(),
    workspaceId: z.string(),
}).transform((data) => {
    return Object.assign(Object.assign({}, data), { url: getProjectUrl(data.id, data.name) });
});
/**
 * Schema for any project type in Todoist (personal or workspace).
 * Dispatches to WorkspaceProjectSchema when workspaceId is present, PersonalProjectSchema otherwise.
 */
export const ProjectSchema = z.unknown().transform((val) => {
    if (typeof val === 'object' && val !== null && 'workspaceId' in val) {
        return WorkspaceProjectSchema.parse(val);
    }
    return PersonalProjectSchema.parse(val);
});
/** Available project view styles. */
export const PROJECT_VIEW_STYLES = ['list', 'board', 'calendar'];
