// @doist/todoist-sdk/dist/cjs/types/comments/types.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommentSchema = exports.RawCommentSchema = exports.AttachmentSchema = exports.UPLOAD_STATES = void 0;
const zod_1 = require("zod");
/** Available file attachment upload states. */
exports.UPLOAD_STATES = ['pending', 'completed'];
exports.AttachmentSchema = zod_1.z
    .object({
    resourceType: zod_1.z.string(),
})
    .extend({
    fileName: zod_1.z.string().nullable().optional(),
    fileSize: zod_1.z.number().int().nullable().optional(),
    fileType: zod_1.z.string().nullable().optional(),
    fileUrl: zod_1.z.string().nullable().optional(),
    fileDuration: zod_1.z.number().int().nullable().optional(),
    uploadState: zod_1.z.enum(exports.UPLOAD_STATES).nullable().optional(),
    image: zod_1.z.string().nullable().optional(),
    imageWidth: zod_1.z.number().int().nullable().optional(),
    imageHeight: zod_1.z.number().int().nullable().optional(),
    url: zod_1.z.string().nullable().optional(),
    title: zod_1.z.string().nullable().optional(),
});
exports.RawCommentSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    itemId: zod_1.z.string().optional(),
    projectId: zod_1.z.string().optional(),
    content: zod_1.z.string(),
    postedAt: zod_1.z.coerce.date(),
    fileAttachment: exports.AttachmentSchema.nullable(),
    postedUid: zod_1.z.string(),
    uidsToNotify: zod_1.z.array(zod_1.z.string()).nullable(),
    reactions: zod_1.z.record(zod_1.z.string(), zod_1.z.array(zod_1.z.string())).nullable(),
    isDeleted: zod_1.z.boolean(),
})
    .refine((data) => {
    // At least one of itemId or projectId must be present
    return ((data.itemId !== undefined && data.projectId === undefined) ||
        (data.itemId === undefined && data.projectId !== undefined));
}, {
    message: 'Exactly one of itemId or projectId must be provided',
});
exports.CommentSchema = exports.RawCommentSchema.transform((data) => {
    const { itemId } = data, rest = __rest(data, ["itemId"]);
    return Object.assign(Object.assign({}, rest), { 
        // Map itemId to taskId for backwards compatibility
        taskId: itemId });
});
