// @doist/todoist-sdk/dist/esm/types/insights/types.js
import { z } from 'zod';
export const DayActivitySchema = z.object({
    // Kept as string: this is a calendar-day bucket (e.g. '2025-01-13'), not a timestamp.
    // Coercing to Date would create UTC midnight, causing day-shift for non-UTC consumers.
    date: z.string(),
    totalCount: z.number().int(),
});
export const WeekRollupSchema = z.object({
    fromDate: z.string(),
    toDate: z.string(),
    totalCount: z.number().int(),
});
export const ProjectActivityStatsSchema = z.object({
    dayItems: z.array(DayActivitySchema),
    weekItems: z.array(WeekRollupSchema).nullable(),
});
/** Available project health statuses. */
export const HEALTH_STATUSES = [
    'UNKNOWN',
    'ON_TRACK',
    'AT_RISK',
    'CRITICAL',
    'EXCELLENT',
    'ERROR',
];
export const TaskRecommendationSchema = z.object({
    taskId: z.string(),
    recommendation: z.string(),
});
export const ProjectHealthSchema = z.object({
    status: z.enum(HEALTH_STATUSES),
    description: z.string().nullable().optional(),
    descriptionSummary: z.string().nullable().optional(),
    taskRecommendations: z.array(TaskRecommendationSchema).nullable().optional(),
    projectId: z.string().nullable().optional(),
    updatedAt: z.coerce.date().nullable().optional(),
    isStale: z.boolean().default(false),
    updateInProgress: z.boolean().default(false),
});
export const ProjectMetricsSchema = z.object({
    totalTasks: z.number().int(),
    completedTasks: z.number().int(),
    overdueTasks: z.number().int(),
    tasksCreatedThisWeek: z.number().int(),
    tasksCompletedThisWeek: z.number().int(),
    averageCompletionTime: z.number().nullable(),
});
export const TaskContextSchema = z.object({
    id: z.string(),
    content: z.string(),
    due: z.string().nullable().optional(),
    deadline: z.string().nullable().optional(),
    priority: z.string(),
    isCompleted: z.boolean(),
    createdAt: z.coerce.date(),
    updatedAt: z.coerce.date(),
    completedAt: z.coerce.date().nullable(),
    completedByUid: z.string().nullable(),
    labels: z.array(z.string()),
});
export const ProjectHealthContextSchema = z.object({
    projectId: z.string(),
    projectName: z.string(),
    projectDescription: z.string().nullable(),
    projectMetrics: ProjectMetricsSchema,
    tasks: z.array(TaskContextSchema),
    language: z.string().nullable().optional(),
});
export const ProjectProgressSchema = z.object({
    projectId: z.string(),
    completedCount: z.number().int(),
    activeCount: z.number().int(),
    progressPercent: z.number().int(),
});
export const ProjectInsightSchema = z.object({
    projectId: z.string(),
    health: ProjectHealthSchema.nullable(),
    progress: ProjectProgressSchema.nullable(),
});
export const WorkspaceInsightsSchema = z.object({
    folderId: z.string().nullable(),
    projectInsights: z.array(ProjectInsightSchema),
});
