// @doist/todoist-sdk/dist/esm/transport/http-dispatcher.js
// Use effectively-disabled keep-alive so short-lived CLI processes do not stay
// open waiting on idle sockets. Undici requires positive values, so we use 1ms.
const KEEP_ALIVE_OPTIONS = {
    keepAliveTimeout: 1,
    keepAliveMaxTimeout: 1,
};
let defaultDispatcherPromise;
export function getDefaultDispatcher() {
    if (!isNodeEnvironment()) {
        return Promise.resolve(undefined);
    }
    if (!defaultDispatcherPromise) {
        defaultDispatcherPromise = createDefaultDispatcher().catch((error) => {
            defaultDispatcherPromise = undefined;
            throw error;
        });
    }
    return defaultDispatcherPromise;
}
export async function resetDefaultDispatcherForTests() {
    if (!defaultDispatcherPromise) {
        return;
    }
    const dispatcherPromise = defaultDispatcherPromise;
    defaultDispatcherPromise = undefined;
    await dispatcherPromise.then((dispatcher) => dispatcher.close(), () => undefined);
}
async function createDefaultDispatcher() {
    // Dynamic import so non-Node consumers (browser, edge runtimes) don't pull
    // undici into their bundle. `getDefaultDispatcher` already short-circuits
    // outside Node, so this branch only runs when undici is safe to load.
    const { EnvHttpProxyAgent, interceptors } = await import('undici');
    const agent = new EnvHttpProxyAgent(KEEP_ALIVE_OPTIONS);
    // Some runtimes report `process.versions.node` (so `isNodeEnvironment()`
    // passes) but ship only a partial undici: `interceptors.decompress` is
    // absent and dispatchers have no `.compose`. Bun is the common case. There
    // the proxy agent alone is enough — Bun's `fetch` decompresses
    // gzip/deflate/br/zstd natively — so skip the interceptor instead of
    // crashing on the missing API.
    if (typeof interceptors.decompress !== 'function') {
        return agent;
    }
    // Compose the response-decompression interceptor so gzip/deflate/br/zstd
    // bodies are decoded before consumers parse them. Required on Node 24+:
    // attaching any custom dispatcher to the global `fetch` strips the
    // `content-encoding` header but does not actually decompress the body,
    // so callers receive raw gzipped bytes and `JSON.parse` fails.
    // See https://github.com/Doist/todoist-cli/issues/318.
    const decompress = suppressExperimentalWarningsSync(() => interceptors.decompress());
    return agent.compose(decompress);
}
// undici emits an `ExperimentalWarning` the first time `interceptors.decompress()`
// runs. The interceptor is stable for our gzipped-JSON-over-HTTPS use case;
// silence the warning during dispatcher init only so it does not leak to every
// consumer's stderr on the first request.
//
// `fn` must be synchronous so the override covers a single critical section
// (microseconds) — no unrelated `ExperimentalWarning` from elsewhere can
// interleave and be lost. We suppress every `ExperimentalWarning` rather than
// pattern-matching the message text: the message wording is an undici
// implementation detail (not a stable API), and the suppression window is
// narrow enough that a coarse type filter is safe.
//
// Exported for direct unit testing — the integration path through
// `getDefaultDispatcher()` cannot reliably exercise the helper because both
// the dispatcher singleton and undici's internal `warningEmitted` flag are
// once-per-process.
export function suppressExperimentalWarningsSync(fn) {
    const originalEmit = process.emitWarning;
    process.emitWarning = ((warning, typeOrOptions, ...rest) => {
        const type = typeof typeOrOptions === 'string'
            ? typeOrOptions
            : typeof typeOrOptions === 'object' && typeOrOptions !== null
                ? typeOrOptions.type
                : undefined;
        if (type === 'ExperimentalWarning')
            return;
        originalEmit.call(process, warning, typeOrOptions, ...rest);
    });
    try {
        return fn();
    }
    finally {
        process.emitWarning = originalEmit;
    }
}
function isNodeEnvironment() {
    var _a;
    return typeof process !== 'undefined' && Boolean((_a = process.versions) === null || _a === void 0 ? void 0 : _a.node);
}
