// @doist/todoist-sdk/dist/esm/transport/sync-request.js
import { ENDPOINT_SYNC } from '../consts/endpoints.js';
import { TodoistArgumentError, TodoistRequestError } from '../types/index.js';
import { DATE_FORMAT_TO_API, TIME_FORMAT_TO_API, DAY_OF_WEEK_TO_API, } from '../types/sync/index.js';
import { spreadIfDefined } from '../utils/request-helpers.js';
import { parseSyncResponse } from '../utils/validators.js';
import { request } from './http-client.js';
function serializeUserUpdateArgs(args) {
    return Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, args), spreadIfDefined(args.dateFormat, (v) => ({
        dateFormat: DATE_FORMAT_TO_API[v],
    }))), spreadIfDefined(args.timeFormat, (v) => ({
        timeFormat: TIME_FORMAT_TO_API[v],
    }))), spreadIfDefined(args.startDay, (v) => ({
        startDay: DAY_OF_WEEK_TO_API[v],
    }))), spreadIfDefined(args.nextWeek, (v) => ({
        nextWeek: DAY_OF_WEEK_TO_API[v],
    })));
}
function serializeTaskUpdateDateCompleteArgs(args) {
    return Object.assign(Object.assign(Object.assign({}, args), { isForward: args.isForward ? 1 : 0 }), spreadIfDefined(args.resetSubtasks, (v) => ({
        resetSubtasks: v ? 1 : 0,
    })));
}
function serializeUpdateGoalsArgs(args) {
    return Object.assign(Object.assign(Object.assign(Object.assign({}, args), spreadIfDefined(args.vacationMode, (v) => ({ vacationMode: v ? 1 : 0 }))), spreadIfDefined(args.karmaDisabled, (v) => ({
        karmaDisabled: v ? 1 : 0,
    }))), spreadIfDefined(args.ignoreDays, (v) => ({
        ignoreDays: v.map((day) => {
            const apiValue = DAY_OF_WEEK_TO_API[day];
            if (apiValue === undefined) {
                throw new TodoistArgumentError(`Invalid ignoreDays entry "${String(day)}". Expected a DayOfWeek (e.g. "Monday").`);
            }
            return apiValue;
        }),
    })));
}
function preprocessSyncCommands(commands) {
    return commands.map((cmd) => {
        if (cmd.type === 'user_update')
            return Object.assign(Object.assign({}, cmd), { args: serializeUserUpdateArgs(cmd.args) });
        if (cmd.type === 'item_update_date_complete')
            return Object.assign(Object.assign({}, cmd), { args: serializeTaskUpdateDateCompleteArgs(cmd.args) });
        if (cmd.type === 'update_goals')
            return Object.assign(Object.assign({}, cmd), { args: serializeUpdateGoalsArgs(cmd.args) });
        return cmd;
    });
}
/**
 * Makes a request to the Sync API and handles error checking.
 *
 * @param ctx - HTTP identity (auth token, sync API base URI, optional custom fetch).
 * @param syncRequest - The sync request payload.
 * @param options - Optional request identifier and commands flag.
 * @returns The sync response data.
 * @throws TodoistRequestError if sync status contains errors.
 */
export async function performSyncRequest(ctx, syncRequest, options = {}) {
    var _a;
    const { requestId, hasSyncCommands = false } = options;
    const processedRequest = ((_a = syncRequest.commands) === null || _a === void 0 ? void 0 : _a.length)
        ? Object.assign(Object.assign({}, syncRequest), { commands: preprocessSyncCommands(syncRequest.commands) }) : syncRequest;
    const response = await request({
        httpMethod: 'POST',
        baseUri: ctx.syncApiBase,
        relativePath: ENDPOINT_SYNC,
        apiToken: ctx.authToken,
        customFetch: ctx.customFetch,
        payload: processedRequest,
        requestId: requestId,
        hasSyncCommands: hasSyncCommands,
    });
    // Check for sync errors and throw if any are found
    if (response.data.syncStatus) {
        Object.entries(response.data.syncStatus).forEach(([_, value]) => {
            if (value === 'ok')
                return;
            throw new TodoistRequestError(value.error, value.httpCode, value.errorExtra);
        });
    }
    return parseSyncResponse(response.data);
}
