// @doist/todoist-sdk/dist/esm/types/workspaces/types.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
import { StringOrNumberSchema } from '../common.js';
import { DueDateSchema, DeadlineSchema } from '../tasks/types.js';
/** Available project collaborator roles. */
export const COLLABORATOR_ROLES = [
    'CREATOR',
    'ADMIN',
    'READ_WRITE',
    'EDIT_ONLY',
    'COMPLETE_ONLY',
];
/**
 * Available workspace roles.
 */
export const WORKSPACE_ROLES = ['ADMIN', 'MEMBER', 'GUEST'];
export const WorkspaceRoleSchema = z.enum(WORKSPACE_ROLES);
export const WorkspaceUserSchema = z.object({
    userId: z.string(),
    workspaceId: z.string(),
    userEmail: z.string(),
    fullName: z.string(),
    timezone: z.string(),
    role: WorkspaceRoleSchema,
    imageId: z.string().nullable(),
    isDeleted: z.boolean().default(false),
});
export const WorkspaceInvitationSchema = z.object({
    id: z.string().default('0'),
    inviterId: z.string(),
    userEmail: z.string(),
    workspaceId: z.string(),
    role: WorkspaceRoleSchema,
    isExistingUser: z.boolean(),
});
export const PlanPriceSchema = z.object({
    currency: z.string(),
    amount: z.union([z.number(), z.string()]),
    interval: z.string().optional(),
});
export const FormattedPriceListingSchema = z.object({
    currency: z.string().optional(),
    amount: z.number().optional(),
    interval: z.string().optional(),
    formatted: z.string().optional(),
});
/** Available workspace plan names. */
export const WORKSPACE_CURRENT_PLANS = ['Business', 'Starter'];
/** Available workspace plan statuses. */
export const WORKSPACE_PLAN_STATUSES = [
    'Active',
    'Downgraded',
    'Cancelled',
    'NeverSubscribed',
];
export const WorkspacePlanDetailsSchema = z.object({
    currentMemberCount: z.number(),
    currentPlan: z.enum(WORKSPACE_CURRENT_PLANS),
    currentPlanStatus: z.enum(WORKSPACE_PLAN_STATUSES),
    downgradeAt: z.string().nullable(),
    currentActiveProjects: z.number(),
    maximumActiveProjects: z.number(),
    priceList: z.array(FormattedPriceListingSchema),
    workspaceId: StringOrNumberSchema,
    isTrialing: z.boolean(),
    trialEndsAt: z.string().nullable(),
    cancelAtPeriodEnd: z.boolean(),
    hasTrialed: z.boolean(),
    planPrice: PlanPriceSchema.nullable(),
    hasBillingPortal: z.boolean(),
    hasBillingPortalSwitchToAnnual: z.boolean(),
});
export const JoinWorkspaceResultSchema = z.object({
    customSortingApplied: z.boolean(),
    projectSortPreference: z.string(),
    role: WorkspaceRoleSchema,
    userId: z.string(),
    workspaceId: z.string(),
});
/**
 * Available workspace plans.
 */
export const WORKSPACE_PLANS = ['STARTER', 'BUSINESS'];
export const WorkspacePlanSchema = z.enum(WORKSPACE_PLANS);
/**
 * Workspace resource limits.
 */
export const WorkspaceLimitsSchema = z
    .object({
    current: z.record(z.string(), z.any()).nullable(),
    next: z.record(z.string(), z.any()).nullable(),
})
    .catchall(z.any());
/**
 * Workspace properties (flexible object for unknown fields).
 */
export const WorkspacePropertiesSchema = z.record(z.string(), z.unknown());
/**
 * Represents a workspace in Todoist.
 */
export const WorkspaceSchema = z
    .object({
    id: StringOrNumberSchema,
    name: z.string(),
    plan: WorkspacePlanSchema,
    role: WorkspaceRoleSchema.optional(),
    inviteCode: z.string(),
    isLinkSharingEnabled: z.boolean(),
    isGuestAllowed: z.boolean(),
    limits: WorkspaceLimitsSchema.optional(),
    logoBig: z.string().nullish(),
    logoMedium: z.string().nullish(),
    logoSmall: z.string().nullish(),
    logoS640: z.string().nullish(),
    createdAt: z.coerce.date().optional(),
    dateCreated: z.coerce.date().optional(),
    creatorId: StringOrNumberSchema,
    properties: WorkspacePropertiesSchema,
})
    .transform((_a) => {
    var { dateCreated, createdAt } = _a, rest = __rest(_a, ["dateCreated", "createdAt"]);
    return (Object.assign(Object.assign({}, rest), { createdAt: createdAt !== null && createdAt !== void 0 ? createdAt : dateCreated }));
});
export const MemberActivityInfoSchema = z.object({
    userId: z.string(),
    tasksAssigned: z.number().int(),
    tasksOverdue: z.number().int(),
});
export const WorkspaceUserTaskSchema = z.object({
    id: z.string(),
    content: z.string(),
    responsibleUid: z.string().nullable(),
    due: DueDateSchema.nullable(),
    deadline: DeadlineSchema.nullable(),
    labels: z.array(z.string()),
    notesCount: z.number().int(),
    projectId: z.string(),
    projectName: z.string(),
    priority: z.number().int(),
    description: z.string(),
    isOverdue: z.boolean(),
});
