// @doist/todoist-sdk/dist/cjs/types/sync/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncResponseSchema = exports.SyncErrorSchema = void 0;
const zod_1 = require("zod");
const types_1 = require("../goals/types");
const types_2 = require("../labels/types");
const types_3 = require("../projects/types");
const types_4 = require("../sections/types");
const types_5 = require("../tasks/types");
const types_6 = require("../workspaces/types");
const resources_1 = require("./resources");
exports.SyncErrorSchema = zod_1.z.object({
    error: zod_1.z.string(),
    errorCode: zod_1.z.number(),
    errorExtra: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()),
    errorTag: zod_1.z.string(),
    httpCode: zod_1.z.number(),
});
exports.SyncResponseSchema = zod_1.z.looseObject({
    syncToken: zod_1.z.string().optional(),
    fullSync: zod_1.z.boolean().optional(),
    syncStatus: zod_1.z.record(zod_1.z.string(), zod_1.z.union([zod_1.z.literal('ok'), exports.SyncErrorSchema])).optional(),
    tempIdMapping: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
    items: zod_1.z.array(types_5.TaskSchema).optional(),
    projects: zod_1.z.array(types_3.ProjectSchema).optional(),
    sections: zod_1.z.array(types_4.SectionSchema).optional(),
    labels: zod_1.z.array(types_2.LabelSchema).optional(),
    notes: zod_1.z.array(resources_1.NoteSchema).optional(),
    projectNotes: zod_1.z.array(resources_1.NoteSchema).optional(),
    filters: zod_1.z.array(resources_1.FilterSchema).optional(),
    reminders: zod_1.z.array(resources_1.ReminderSchema).optional(),
    remindersLocation: zod_1.z.array(resources_1.LocationReminderSchema).optional(),
    locations: zod_1.z.array(zod_1.z.record(zod_1.z.string(), zod_1.z.unknown())).optional(),
    user: resources_1.SyncUserSchema.optional(),
    liveNotifications: zod_1.z.array(resources_1.LiveNotificationSchema).optional(),
    collaborators: zod_1.z.array(resources_1.CollaboratorSchema).optional(),
    collaboratorStates: zod_1.z.array(resources_1.CollaboratorStateSchema).optional(),
    userSettings: resources_1.UserSettingsSchema.optional(),
    notificationSettings: zod_1.z.record(zod_1.z.string(), zod_1.z.boolean()).optional(),
    userPlanLimits: resources_1.UserPlanLimitsSchema.optional(),
    completedInfo: zod_1.z.array(resources_1.CompletedInfoSchema).optional(),
    stats: zod_1.z.record(zod_1.z.string(), zod_1.z.unknown()).optional(),
    workspaces: zod_1.z.array(resources_1.SyncWorkspaceSchema).optional(),
    workspaceUsers: zod_1.z.array(types_6.WorkspaceUserSchema).optional(),
    workspaceFilters: zod_1.z.array(resources_1.WorkspaceFilterSchema).optional(),
    viewOptions: zod_1.z.array(resources_1.ViewOptionsSchema).optional(),
    projectViewOptionsDefaults: zod_1.z.array(resources_1.ProjectViewOptionsDefaultsSchema).optional(),
    roleActions: zod_1.z.array(zod_1.z.record(zod_1.z.string(), zod_1.z.unknown())).optional(),
    folders: zod_1.z.array(resources_1.FolderSchema).optional(),
    goals: zod_1.z.array(types_1.GoalSchema).optional(),
    workspaceGoals: zod_1.z.array(resources_1.WorkspaceGoalSchema).optional(),
    dayOrders: zod_1.z.record(zod_1.z.string(), zod_1.z.number()).optional(),
    calendars: zod_1.z.array(resources_1.CalendarSchema).optional(),
    calendarAccounts: zod_1.z.array(resources_1.CalendarAccountSchema).optional(),
    suggestions: zod_1.z.array(resources_1.SuggestionSchema).optional(),
    tooltips: resources_1.TooltipsSchema.optional(),
});
