// @doist/todoist-sdk/dist/cjs/consts/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENDPOINT_REST_ID_MAPPINGS = exports.ENDPOINT_REST_EMAILS = exports.ENDPOINT_REST_BACKUPS_DOWNLOAD = exports.ENDPOINT_REST_BACKUPS = exports.ENDPOINT_REST_ACCESS_TOKENS_MIGRATE = exports.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_ID = exports.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_FILE = exports.ENDPOINT_REST_TEMPLATES_CREATE_FROM_FILE = exports.ENDPOINT_REST_TEMPLATES_URL = exports.ENDPOINT_REST_TEMPLATES_FILE = exports.ENDPOINT_REST_TASKS_COMPLETED = exports.ENDPOINT_REST_PROJECTS_MOVE_TO_PERSONAL = exports.ENDPOINT_REST_PROJECTS_MOVE_TO_WORKSPACE = exports.SECTION_UNARCHIVE = exports.SECTION_ARCHIVE = exports.ENDPOINT_REST_PROJECT_JOIN = exports.ENDPOINT_REST_PROJECT_FULL = exports.ENDPOINT_REST_PROJECTS_PERMISSIONS = exports.ENDPOINT_REST_PROJECTS_ARCHIVED_COUNT = exports.PROJECT_UNARCHIVE = exports.PROJECT_ARCHIVE = exports.ENDPOINT_REST_UPLOADS = exports.ENDPOINT_REST_ACTIVITIES = exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_USER = exports.ENDPOINT_REST_PROJECT_COLLABORATORS = exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS_SEARCH = exports.ENDPOINT_REST_PROJECTS = exports.ENDPOINT_REST_TASK_MOVE = exports.ENDPOINT_REST_TASK_REOPEN = exports.ENDPOINT_REST_TASK_CLOSE = exports.ENDPOINT_REST_LOCATION_REMINDERS = exports.ENDPOINT_REST_REMINDERS = exports.ENDPOINT_REST_COMMENTS = exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS_SEARCH = exports.ENDPOINT_REST_LABELS = exports.ENDPOINT_REST_SECTIONS_SEARCH = exports.ENDPOINT_REST_SECTIONS = exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS = exports.API_BASE_URI = exports.API_VERSION = exports.TODOIST_WEB_URI = void 0;
exports.APP_ICON_SIZES = exports.ENDPOINT_REST_USER_AUTHORIZATIONS_DELETE = exports.ENDPOINT_REST_USER_AUTHORIZATIONS = exports.ENDPOINT_REST_APPS_UI_EXTENSIONS_INSTALLED = exports.ENDPOINT_REST_APPS_UI_EXTENSIONS = exports.ENDPOINT_REST_APPS_INSTALLATIONS = exports.ENDPOINT_REST_APPS = exports.ENDPOINT_WORKSPACE_MEMBERS = exports.ENDPOINT_PRICING = exports.ENDPOINT_PRICES = exports.ENDPOINT_WORKSPACE_BILLING_PORTAL = exports.ENDPOINT_WORKSPACE_TRIAL = exports.ENDPOINT_WORKSPACE_UPGRADE = exports.ENDPOINT_PRO_PLAN_DETAILS = exports.ENDPOINT_PRO_BILLING_PORTAL = exports.ENDPOINT_PRO_TRIAL = exports.ENDPOINT_PRO_UPGRADE = exports.ENDPOINT_PAYMENTS_REACTIVATE_PLAN = exports.ENDPOINT_PAYMENTS_CANCEL_PLAN = exports.ENDPOINT_PAYMENTS_SUBSCRIPTION_INFO = exports.ENDPOINT_WORKSPACE_USERS = exports.ENDPOINT_WORKSPACE_PLAN_DETAILS = exports.ENDPOINT_WORKSPACE_LOGO = exports.ENDPOINT_WORKSPACE_JOIN = exports.ENDPOINT_WORKSPACE_INVITATIONS_DELETE = exports.ENDPOINT_WORKSPACE_INVITATIONS_ALL = exports.ENDPOINT_WORKSPACE_INVITATIONS = exports.ENDPOINT_REST_WORKSPACES = exports.GOAL_TASKS = exports.GOAL_UNCOMPLETE = exports.GOAL_COMPLETE = exports.ENDPOINT_REST_GOALS_SEARCH = exports.ENDPOINT_REST_GOALS = exports.ENDPOINT_REST_FOLDERS = exports.ENDPOINT_REGISTER = exports.ENDPOINT_REVOKE = exports.ENDPOINT_GET_TOKEN = exports.ENDPOINT_AUTHORIZATION = exports.ENDPOINT_SYNC = exports.ENDPOINT_SYNC_QUICK_ADD = exports.ENDPOINT_REST_MOVED_IDS = void 0;
exports.getSyncBaseUri = getSyncBaseUri;
exports.getApiRootBaseUri = getApiRootBaseUri;
exports.getAuthBaseUri = getAuthBaseUri;
exports.getWorkspaceInvitationAcceptEndpoint = getWorkspaceInvitationAcceptEndpoint;
exports.getWorkspaceInvitationRejectEndpoint = getWorkspaceInvitationRejectEndpoint;
exports.getProjectInsightsActivityStatsEndpoint = getProjectInsightsActivityStatsEndpoint;
exports.getProjectInsightsHealthEndpoint = getProjectInsightsHealthEndpoint;
exports.getProjectInsightsHealthContextEndpoint = getProjectInsightsHealthContextEndpoint;
exports.getProjectInsightsProgressEndpoint = getProjectInsightsProgressEndpoint;
exports.getProjectInsightsHealthAnalyzeEndpoint = getProjectInsightsHealthAnalyzeEndpoint;
exports.getWorkspaceInsightsEndpoint = getWorkspaceInsightsEndpoint;
exports.getWorkspaceUserTasksEndpoint = getWorkspaceUserTasksEndpoint;
exports.getWorkspaceInviteUsersEndpoint = getWorkspaceInviteUsersEndpoint;
exports.getWorkspaceUserEndpoint = getWorkspaceUserEndpoint;
exports.getWorkspaceActiveProjectsEndpoint = getWorkspaceActiveProjectsEndpoint;
exports.getWorkspaceArchivedProjectsEndpoint = getWorkspaceArchivedProjectsEndpoint;
exports.getAppEndpoint = getAppEndpoint;
exports.getAppSecretsEndpoint = getAppSecretsEndpoint;
exports.getAppClientSecretEndpoint = getAppClientSecretEndpoint;
exports.getAppTokensEndpoint = getAppTokensEndpoint;
exports.getAppTestTokenEndpoint = getAppTestTokenEndpoint;
exports.getAppDistributionTokenEndpoint = getAppDistributionTokenEndpoint;
exports.getAppVerificationTokenEndpoint = getAppVerificationTokenEndpoint;
exports.getAppWebhookEndpoint = getAppWebhookEndpoint;
exports.getAppIconEndpoint = getAppIconEndpoint;
exports.getAppInstallationEndpoint = getAppInstallationEndpoint;
exports.getAppByDistributionTokenEndpoint = getAppByDistributionTokenEndpoint;
exports.getUiExtensionEndpoint = getUiExtensionEndpoint;
exports.getUiExtensionIconEndpoint = getUiExtensionIconEndpoint;
exports.getUiExtensionsByIntegrationEndpoint = getUiExtensionsByIntegrationEndpoint;
const BASE_URI = 'https://api.todoist.com';
const TODOIST_URI = 'https://todoist.com';
exports.TODOIST_WEB_URI = 'https://app.todoist.com/app';
// The API version is not configurable, to ensure
// compatibility between the API and the client.
exports.API_VERSION = 'v1';
exports.API_BASE_URI = `/api/${exports.API_VERSION}/`;
const API_AUTHORIZATION_BASE_URI = '/oauth/';
function getSyncBaseUri(domainBase = BASE_URI) {
    return new URL(exports.API_BASE_URI, domainBase).toString();
}
function getApiRootBaseUri(domainBase = BASE_URI) {
    return new URL('/', domainBase).toString();
}
function getAuthBaseUri(domainBase = TODOIST_URI) {
    return new URL(API_AUTHORIZATION_BASE_URI, domainBase).toString();
}
exports.ENDPOINT_REST_TASKS = 'tasks';
exports.ENDPOINT_REST_TASKS_FILTER = exports.ENDPOINT_REST_TASKS + '/filter';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_completion_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE = exports.ENDPOINT_REST_TASKS + '/completed/by_due_date';
exports.ENDPOINT_REST_TASKS_COMPLETED_SEARCH = 'completed/search';
exports.ENDPOINT_REST_SECTIONS = 'sections';
exports.ENDPOINT_REST_SECTIONS_SEARCH = exports.ENDPOINT_REST_SECTIONS + '/search';
exports.ENDPOINT_REST_LABELS = 'labels';
exports.ENDPOINT_REST_LABELS_SEARCH = exports.ENDPOINT_REST_LABELS + '/search';
exports.ENDPOINT_REST_LABELS_SHARED = exports.ENDPOINT_REST_LABELS + '/shared';
exports.ENDPOINT_REST_LABELS_SHARED_RENAME = exports.ENDPOINT_REST_LABELS_SHARED + '/rename';
exports.ENDPOINT_REST_LABELS_SHARED_REMOVE = exports.ENDPOINT_REST_LABELS_SHARED + '/remove';
exports.ENDPOINT_REST_COMMENTS = 'comments';
exports.ENDPOINT_REST_REMINDERS = 'reminders';
exports.ENDPOINT_REST_LOCATION_REMINDERS = 'location_reminders';
exports.ENDPOINT_REST_TASK_CLOSE = 'close';
exports.ENDPOINT_REST_TASK_REOPEN = 'reopen';
exports.ENDPOINT_REST_TASK_MOVE = 'move';
exports.ENDPOINT_REST_PROJECTS = 'projects';
exports.ENDPOINT_REST_PROJECTS_SEARCH = exports.ENDPOINT_REST_PROJECTS + '/search';
exports.ENDPOINT_REST_PROJECTS_ARCHIVED = exports.ENDPOINT_REST_PROJECTS + '/archived';
exports.ENDPOINT_REST_PROJECT_COLLABORATORS = 'collaborators';
exports.ENDPOINT_REST_USER = 'user';
exports.ENDPOINT_REST_PRODUCTIVITY = exports.ENDPOINT_REST_TASKS + '/completed/stats';
exports.ENDPOINT_REST_ACTIVITIES = 'activities';
exports.ENDPOINT_REST_UPLOADS = 'uploads';
exports.PROJECT_ARCHIVE = 'archive';
exports.PROJECT_UNARCHIVE = 'unarchive';
exports.ENDPOINT_REST_PROJECTS_ARCHIVED_COUNT = exports.ENDPOINT_REST_PROJECTS + '/archived/count';
exports.ENDPOINT_REST_PROJECTS_PERMISSIONS = exports.ENDPOINT_REST_PROJECTS + '/permissions';
exports.ENDPOINT_REST_PROJECT_FULL = 'full';
exports.ENDPOINT_REST_PROJECT_JOIN = 'join';
exports.SECTION_ARCHIVE = 'archive';
exports.SECTION_UNARCHIVE = 'unarchive';
exports.ENDPOINT_REST_PROJECTS_MOVE_TO_WORKSPACE = exports.ENDPOINT_REST_PROJECTS + '/move_to_workspace';
exports.ENDPOINT_REST_PROJECTS_MOVE_TO_PERSONAL = exports.ENDPOINT_REST_PROJECTS + '/move_to_personal';
exports.ENDPOINT_REST_TASKS_COMPLETED = exports.ENDPOINT_REST_TASKS + '/completed';
exports.ENDPOINT_REST_TEMPLATES_FILE = 'templates/file';
exports.ENDPOINT_REST_TEMPLATES_URL = 'templates/url';
exports.ENDPOINT_REST_TEMPLATES_CREATE_FROM_FILE = 'templates/create_project_from_file';
exports.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_FILE = 'templates/import_into_project_from_file';
exports.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_ID = 'templates/import_into_project_from_template_id';
exports.ENDPOINT_REST_ACCESS_TOKENS_MIGRATE = 'access_tokens/migrate_personal_token';
exports.ENDPOINT_REST_BACKUPS = 'backups';
exports.ENDPOINT_REST_BACKUPS_DOWNLOAD = 'backups/download';
exports.ENDPOINT_REST_EMAILS = 'emails';
exports.ENDPOINT_REST_ID_MAPPINGS = 'id_mappings';
exports.ENDPOINT_REST_MOVED_IDS = 'moved_ids';
exports.ENDPOINT_SYNC_QUICK_ADD = exports.ENDPOINT_REST_TASKS + '/quick';
exports.ENDPOINT_SYNC = 'sync';
exports.ENDPOINT_AUTHORIZATION = 'authorize';
exports.ENDPOINT_GET_TOKEN = 'access_token';
exports.ENDPOINT_REVOKE = 'revoke';
exports.ENDPOINT_REGISTER = 'register';
exports.ENDPOINT_REST_FOLDERS = 'folders';
exports.ENDPOINT_REST_GOALS = 'goals';
exports.ENDPOINT_REST_GOALS_SEARCH = exports.ENDPOINT_REST_GOALS + '/search';
exports.GOAL_COMPLETE = 'complete';
exports.GOAL_UNCOMPLETE = 'uncomplete';
exports.GOAL_TASKS = 'items';
// Workspace endpoints
exports.ENDPOINT_REST_WORKSPACES = 'workspaces';
exports.ENDPOINT_WORKSPACE_INVITATIONS = 'workspaces/invitations';
exports.ENDPOINT_WORKSPACE_INVITATIONS_ALL = 'workspaces/invitations/all';
exports.ENDPOINT_WORKSPACE_INVITATIONS_DELETE = 'workspaces/invitations/delete';
exports.ENDPOINT_WORKSPACE_JOIN = 'workspaces/join';
exports.ENDPOINT_WORKSPACE_LOGO = 'workspaces/logo';
exports.ENDPOINT_WORKSPACE_PLAN_DETAILS = 'workspaces/plan_details';
exports.ENDPOINT_WORKSPACE_USERS = 'workspaces/users';
// Billing endpoints
exports.ENDPOINT_PAYMENTS_SUBSCRIPTION_INFO = 'payments/get_subscription_info';
exports.ENDPOINT_PAYMENTS_CANCEL_PLAN = 'payments/cancel_plan_with_redirect_to_stripe';
exports.ENDPOINT_PAYMENTS_REACTIVATE_PLAN = 'payments/reactivate_plan';
exports.ENDPOINT_PRO_UPGRADE = 'pro/upgrade';
exports.ENDPOINT_PRO_TRIAL = 'pro/trial';
exports.ENDPOINT_PRO_BILLING_PORTAL = 'pro/billing_portal';
exports.ENDPOINT_PRO_PLAN_DETAILS = 'pro/plan_details';
exports.ENDPOINT_WORKSPACE_UPGRADE = 'workspaces/upgrade';
exports.ENDPOINT_WORKSPACE_TRIAL = 'workspaces/trial';
exports.ENDPOINT_WORKSPACE_BILLING_PORTAL = 'workspaces/billing_portal';
exports.ENDPOINT_PRICES = 'prices';
exports.ENDPOINT_PRICING = 'pricing';
// Workspace invitation actions (require invite_code parameter)
function getWorkspaceInvitationAcceptEndpoint(inviteCode) {
    return `workspaces/invitations/${inviteCode}/accept`;
}
function getWorkspaceInvitationRejectEndpoint(inviteCode) {
    return `workspaces/invitations/${inviteCode}/reject`;
}
// Insights endpoints
function getProjectInsightsActivityStatsEndpoint(projectId) {
    return `projects/${projectId}/insights/activity_stats`;
}
function getProjectInsightsHealthEndpoint(projectId) {
    return `projects/${projectId}/insights/health`;
}
function getProjectInsightsHealthContextEndpoint(projectId) {
    return `projects/${projectId}/insights/health/context`;
}
function getProjectInsightsProgressEndpoint(projectId) {
    return `projects/${projectId}/insights/progress`;
}
function getProjectInsightsHealthAnalyzeEndpoint(projectId) {
    return `projects/${projectId}/insights/health/analyze`;
}
function getWorkspaceInsightsEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/insights`;
}
// Workspace members
exports.ENDPOINT_WORKSPACE_MEMBERS = 'workspaces/members';
// Workspace user management (require workspace_id and/or user_id parameters)
function getWorkspaceUserTasksEndpoint(workspaceId, userId) {
    return `workspaces/${workspaceId}/users/${userId}/tasks`;
}
function getWorkspaceInviteUsersEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/users/invite`;
}
function getWorkspaceUserEndpoint(workspaceId, userId) {
    return `workspaces/${workspaceId}/users/${userId}`;
}
// Workspace projects (require workspace_id parameter)
function getWorkspaceActiveProjectsEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/projects/active`;
}
function getWorkspaceArchivedProjectsEndpoint(workspaceId) {
    return `workspaces/${workspaceId}/projects/archived`;
}
// App management endpoints (dev:app_console scope)
exports.ENDPOINT_REST_APPS = 'apps';
exports.ENDPOINT_REST_APPS_INSTALLATIONS = 'apps/installations';
exports.ENDPOINT_REST_APPS_UI_EXTENSIONS = 'apps/ui_extensions';
exports.ENDPOINT_REST_APPS_UI_EXTENSIONS_INSTALLED = 'apps/ui_extensions/installed';
exports.ENDPOINT_REST_USER_AUTHORIZATIONS = 'user/authorizations';
exports.ENDPOINT_REST_USER_AUTHORIZATIONS_DELETE = 'user/authorizations/delete';
/** Available app icon sizes (small 60x60, medium 300x300, large 600x600). */
exports.APP_ICON_SIZES = ['small', 'medium', 'large'];
function getAppEndpoint(appId) {
    return `apps/${appId}`;
}
function getAppSecretsEndpoint(appId) {
    return `apps/${appId}/secrets`;
}
function getAppClientSecretEndpoint(appId) {
    return `apps/${appId}/client_secret`;
}
function getAppTokensEndpoint(appId) {
    return `apps/${appId}/tokens`;
}
function getAppTestTokenEndpoint(appId) {
    return `apps/${appId}/test_token`;
}
function getAppDistributionTokenEndpoint(appId) {
    return `apps/${appId}/distribution_token`;
}
function getAppVerificationTokenEndpoint(appId) {
    return `apps/${appId}/verification_token`;
}
function getAppWebhookEndpoint(appId) {
    return `apps/${appId}/webhook`;
}
function getAppIconEndpoint(appId, size) {
    return size === 'medium' ? `apps/${appId}/icon` : `apps/${appId}/icon/${size}`;
}
function getAppInstallationEndpoint(installationId) {
    return `apps/installations/${installationId}`;
}
function getAppByDistributionTokenEndpoint(distributionToken) {
    return `apps/by_distribution_token/${distributionToken}`;
}
function getUiExtensionEndpoint(uiExtensionId) {
    return `apps/ui_extensions/${uiExtensionId}`;
}
function getUiExtensionIconEndpoint(uiExtensionId) {
    return `apps/ui_extensions/${uiExtensionId}/icon`;
}
function getUiExtensionsByIntegrationEndpoint(integrationId) {
    return `apps/ui_extensions/integration/${integrationId}`;
}
