// @doist/todoist-sdk/dist/cjs/authentication.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_ENDPOINT_AUTH_METHODS = exports.PERMISSIONS = void 0;
exports.getAuthStateParameter = getAuthStateParameter;
exports.getAuthorizationUrl = getAuthorizationUrl;
exports.getAuthToken = getAuthToken;
exports.refreshAuthToken = refreshAuthToken;
exports.revokeToken = revokeToken;
exports.migratePersonalToken = migratePersonalToken;
exports.registerClient = registerClient;
const uuid_1 = require("uuid");
const endpoints_1 = require("./consts/endpoints");
const http_client_1 = require("./transport/http-client");
const types_1 = require("./types");
/** Available OAuth2 permission scopes. */
exports.PERMISSIONS = [
    'task:add',
    'data:read',
    'data:read_write',
    'data:delete',
    'project:delete',
    'backups:read',
    'billing:read',
    'billing:read_write',
    'dev:app_console',
];
/** Supported token endpoint authentication methods for dynamic client registration. */
exports.TOKEN_ENDPOINT_AUTH_METHODS = [
    'client_secret_post',
    'client_secret_basic',
    'none',
];
/**
 * Creates a Basic Authentication header value from client credentials.
 * @param clientId - The OAuth client ID
 * @param clientSecret - The OAuth client secret
 * @returns The Basic Auth header value (without the 'Basic ' prefix)
 */
function createBasicAuthHeader(clientId, clientSecret) {
    const credentials = `${clientId}:${clientSecret}`;
    return Buffer.from(credentials).toString('base64');
}
/**
 * Generates a random state parameter for OAuth2 authorization.
 * The state parameter helps prevent CSRF attacks.
 *
 * @example
 * ```typescript
 * const state = getAuthStateParameter()
 * // Store state in session
 * const authUrl = getAuthorizationUrl(clientId, ['data:read'], state)
 * ```
 *
 * @returns A random UUID v4 string
 */
function getAuthStateParameter() {
    return (0, uuid_1.v4)();
}
/**
 * Generates the authorization URL for the OAuth2 flow.
 *
 * The `clientId` can be either a traditional client ID string (e.g. from
 * {@link registerClient}) or an HTTPS URL pointing to a client metadata document,
 * as defined in {@link https://drafts.ietf.org/doc/draft-ietf-oauth-client-id-metadata-document/ RFC draft-ietf-oauth-client-id-metadata-document}.
 *
 * @example
 * ```typescript
 * const url = getAuthorizationUrl({
 *   clientId: 'your-client-id',
 *   permissions: ['data:read', 'task:add'],
 *   state,
 * })
 * // Redirect user to url
 * ```
 *
 * @returns The full authorization URL to redirect users to
 * @see https://developer.todoist.com/api/v1/#tag/Authorization/OAuth
 */
function getAuthorizationUrl({ clientId, permissions, state, baseUrl, }) {
    if (!(permissions === null || permissions === void 0 ? void 0 : permissions.length)) {
        throw new Error('At least one scope value should be passed for permissions.');
    }
    const scope = permissions.join(',');
    return `${(0, endpoints_1.getAuthBaseUri)(baseUrl)}${endpoints_1.ENDPOINT_AUTHORIZATION}?client_id=${encodeURIComponent(clientId)}&scope=${scope}&state=${state}`;
}
/**
 * Exchanges an authorization code for an access token.
 *
 * @example
 * ```typescript
 * const { accessToken } = await getAuthToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   code: authCode
 * })
 * ```
 *
 * @returns The access token response
 * @throws {@link TodoistRequestError} If the token exchange fails
 */
/**
 * Posts a payload to the OAuth token endpoint and returns the token response,
 * mapping any failure to a {@link TodoistRequestError} with `errorMessage`.
 * Shared by {@link getAuthToken} and {@link refreshAuthToken}.
 */
async function requestAuthToken(payload, errorMessage, options) {
    var _a;
    try {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: (0, endpoints_1.getAuthBaseUri)(options === null || options === void 0 ? void 0 : options.baseUrl),
            relativePath: endpoints_1.ENDPOINT_GET_TOKEN,
            apiToken: undefined,
            payload,
            customFetch: options === null || options === void 0 ? void 0 : options.customFetch,
        });
        if (response.status !== 200 || !((_a = response.data) === null || _a === void 0 ? void 0 : _a.accessToken)) {
            throw new types_1.TodoistRequestError(errorMessage, response.status, response.data);
        }
        return response.data;
    }
    catch (error) {
        // Re-throw with custom message for authentication failures
        const err = error;
        throw new types_1.TodoistRequestError(errorMessage, err.httpStatusCode, err.responseData);
    }
}
async function getAuthToken(args, options) {
    if (typeof options === 'string') {
        throw new TypeError('Passing baseUrl as a string is no longer supported. Use an options object instead: getAuthToken(args, { baseUrl })');
    }
    return requestAuthToken(Object.assign({}, args), 'Authentication token exchange failed.', options);
}
/**
 * Exchanges a refresh token for a new access token using the OAuth2
 * `refresh_token` grant.
 *
 * Only relevant for OAuth apps that have refresh tokens enabled (the default for
 * new apps). Apps without them never receive a `refreshToken` from
 * {@link getAuthToken}, so they have nothing to pass here — their access tokens
 * are long-lived instead.
 *
 * Omit `clientSecret` for public clients registered with
 * `tokenEndpointAuthMethod: 'none'`.
 *
 * @example
 * ```typescript
 * const { accessToken, refreshToken } = await refreshAuthToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   refreshToken: storedRefreshToken,
 * })
 * ```
 *
 * @returns The refreshed token. The server rotates the refresh token, so the
 * response carries a new `refreshToken` — persist it in place of the old one.
 * @throws {@link TodoistRequestError} If the refresh fails
 */
async function refreshAuthToken(args, options) {
    const payload = {
        clientId: args.clientId,
        refreshToken: args.refreshToken,
        grantType: 'refresh_token',
    };
    if (args.clientSecret !== undefined) {
        payload.clientSecret = args.clientSecret;
    }
    return requestAuthToken(payload, 'Authentication token refresh failed.', options);
}
/**
 * Revokes a token using the RFC 7009 OAuth 2.0 Token Revocation standard.
 *
 * This function uses HTTP Basic Authentication with client credentials and follows
 * the RFC 7009 specification for token revocation.
 *
 * @example
 * ```typescript
 * await revokeToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   token: 'access-token-to-revoke'
 * })
 * ```
 *
 * @returns True if revocation was successful
 * @see https://datatracker.ietf.org/doc/html/rfc7009
 * @see https://developer.todoist.com/api/v1/#tag/Authorization
 */
async function revokeToken(args, options) {
    if (typeof options === 'string') {
        throw new TypeError('Passing baseUrl as a string is no longer supported. Use an options object instead: revokeToken(args, { baseUrl })');
    }
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    const { clientId, clientSecret, token } = args;
    // Create Basic Auth header as per RFC 7009
    const basicAuth = createBasicAuthHeader(clientId, clientSecret);
    const customHeaders = {
        Authorization: `Basic ${basicAuth}`,
    };
    // Request body only contains the token and optional token_type_hint
    const requestBody = {
        token,
        token_type_hint: 'access_token',
    };
    const response = await (0, http_client_1.request)({
        httpMethod: 'POST',
        baseUri: (0, endpoints_1.getSyncBaseUri)(baseUrl),
        relativePath: endpoints_1.ENDPOINT_REVOKE,
        apiToken: undefined,
        payload: requestBody,
        requestId: undefined,
        hasSyncCommands: false,
        customHeaders: customHeaders,
        customFetch,
    });
    return (0, http_client_1.isSuccess)(response);
}
/**
 * Migrates a personal API token to an OAuth access token.
 *
 * This allows applications to transition users from personal API tokens
 * to proper OAuth tokens without requiring the user to go through the
 * full OAuth authorization flow.
 *
 * @example
 * ```typescript
 * const { accessToken } = await migratePersonalToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   personalToken: 'user-personal-token',
 *   scope: 'data:read_write,data:delete'
 * })
 * ```
 *
 * @returns The new OAuth token response
 * @throws {@link TodoistRequestError} If the migration fails
 */
async function migratePersonalToken(args, options) {
    var _a;
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    try {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: (0, endpoints_1.getSyncBaseUri)(baseUrl),
            relativePath: endpoints_1.ENDPOINT_REST_ACCESS_TOKENS_MIGRATE,
            apiToken: undefined,
            payload: Object.assign(Object.assign({}, args), { scope: args.scope.join(',') }),
            customFetch,
        });
        if (response.status !== 200 || !((_a = response.data) === null || _a === void 0 ? void 0 : _a.accessToken)) {
            throw new types_1.TodoistRequestError('Personal token migration failed.', response.status, response.data);
        }
        return response.data;
    }
    catch (error) {
        const err = error;
        throw new types_1.TodoistRequestError('Personal token migration failed.', err.httpStatusCode, err.responseData);
    }
}
/**
 * Registers a new OAuth client via Dynamic Client Registration (RFC 7591).
 *
 * @example
 * ```typescript
 * const client = await registerClient({
 *   redirectUris: ['https://example.com/callback'],
 *   clientName: 'My App',
 *   scope: ['data:read_write', 'task:add'],
 * })
 * // Use client.clientId and client.clientSecret for OAuth flows
 * ```
 *
 * @returns The registered client details
 * @throws {@link TodoistRequestError} If the registration fails
 * @see {@link https://datatracker.ietf.org/doc/html/rfc7591 RFC 7591}
 */
async function registerClient(args, options) {
    var _a, _b;
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    try {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: (0, endpoints_1.getAuthBaseUri)(baseUrl),
            relativePath: endpoints_1.ENDPOINT_REGISTER,
            apiToken: undefined,
            payload: Object.assign(Object.assign({}, args), { scope: (_a = args.scope) === null || _a === void 0 ? void 0 : _a.join(' ') }),
            customFetch,
        });
        if (!(0, http_client_1.isSuccess)(response) || !((_b = response.data) === null || _b === void 0 ? void 0 : _b.clientId)) {
            throw new types_1.TodoistRequestError('Dynamic client registration failed.', response.status, response.data);
        }
        const _c = response.data, { clientIdIssuedAt, clientSecretExpiresAt, scope } = _c, rest = __rest(_c, ["clientIdIssuedAt", "clientSecretExpiresAt", "scope"]);
        return Object.assign(Object.assign({}, rest), { scope: scope ? scope.split(' ') : undefined, clientIdIssuedAt: clientIdIssuedAt !== undefined ? new Date(clientIdIssuedAt * 1000) : undefined, clientSecretExpiresAt: clientSecretExpiresAt === undefined
                ? undefined
                : clientSecretExpiresAt === 0
                    ? null
                    : new Date(clientSecretExpiresAt * 1000) });
    }
    catch (error) {
        const err = error;
        throw new types_1.TodoistRequestError('Dynamic client registration failed.', err.httpStatusCode, err.responseData);
    }
}
