// @doist/todoist-sdk/dist/cjs/types/activity/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityEventSchema = exports.ActivityEventExtraDataSchema = void 0;
const zod_1 = require("zod");
/**
 * Flexible object containing event-specific data.
 * Uses z.record to accept any properties for forward compatibility.
 */
exports.ActivityEventExtraDataSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.any()).nullable();
/**
 * Activity log event schema. Accepts unknown fields for forward compatibility.
 */
exports.ActivityEventSchema = zod_1.z
    .object({
    objectType: zod_1.z.string(),
    objectId: zod_1.z.string(),
    eventType: zod_1.z.string(),
    eventDate: zod_1.z.coerce.date(),
    id: zod_1.z
        .union([zod_1.z.string(), zod_1.z.number()])
        .transform((val) => { var _a; return (_a = val === null || val === void 0 ? void 0 : val.toString()) !== null && _a !== void 0 ? _a : null; })
        .nullable(),
    parentProjectId: zod_1.z.string().nullable(),
    parentItemId: zod_1.z.string().nullable(),
    initiatorId: zod_1.z.string().nullable(),
    extraData: exports.ActivityEventExtraDataSchema,
})
    .catchall(zod_1.z.any());
