// @doist/todoist-sdk/dist/cjs/types/insights/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceInsightsSchema = exports.ProjectInsightSchema = exports.ProjectProgressSchema = exports.ProjectHealthContextSchema = exports.TaskContextSchema = exports.ProjectMetricsSchema = exports.ProjectHealthSchema = exports.TaskRecommendationSchema = exports.HEALTH_STATUSES = exports.ProjectActivityStatsSchema = exports.WeekRollupSchema = exports.DayActivitySchema = void 0;
const zod_1 = require("zod");
exports.DayActivitySchema = zod_1.z.object({
    // Kept as string: this is a calendar-day bucket (e.g. '2025-01-13'), not a timestamp.
    // Coercing to Date would create UTC midnight, causing day-shift for non-UTC consumers.
    date: zod_1.z.string(),
    totalCount: zod_1.z.number().int(),
});
exports.WeekRollupSchema = zod_1.z.object({
    fromDate: zod_1.z.string(),
    toDate: zod_1.z.string(),
    totalCount: zod_1.z.number().int(),
});
exports.ProjectActivityStatsSchema = zod_1.z.object({
    dayItems: zod_1.z.array(exports.DayActivitySchema),
    weekItems: zod_1.z.array(exports.WeekRollupSchema).nullable(),
});
/** Available project health statuses. */
exports.HEALTH_STATUSES = [
    'UNKNOWN',
    'ON_TRACK',
    'AT_RISK',
    'CRITICAL',
    'EXCELLENT',
    'ERROR',
];
exports.TaskRecommendationSchema = zod_1.z.object({
    taskId: zod_1.z.string(),
    recommendation: zod_1.z.string(),
});
exports.ProjectHealthSchema = zod_1.z.object({
    status: zod_1.z.enum(exports.HEALTH_STATUSES),
    description: zod_1.z.string().nullable().optional(),
    descriptionSummary: zod_1.z.string().nullable().optional(),
    taskRecommendations: zod_1.z.array(exports.TaskRecommendationSchema).nullable().optional(),
    projectId: zod_1.z.string().nullable().optional(),
    updatedAt: zod_1.z.coerce.date().nullable().optional(),
    isStale: zod_1.z.boolean().default(false),
    updateInProgress: zod_1.z.boolean().default(false),
});
exports.ProjectMetricsSchema = zod_1.z.object({
    totalTasks: zod_1.z.number().int(),
    completedTasks: zod_1.z.number().int(),
    overdueTasks: zod_1.z.number().int(),
    tasksCreatedThisWeek: zod_1.z.number().int(),
    tasksCompletedThisWeek: zod_1.z.number().int(),
    averageCompletionTime: zod_1.z.number().nullable(),
});
exports.TaskContextSchema = zod_1.z.object({
    id: zod_1.z.string(),
    content: zod_1.z.string(),
    due: zod_1.z.string().nullable().optional(),
    deadline: zod_1.z.string().nullable().optional(),
    priority: zod_1.z.string(),
    isCompleted: zod_1.z.boolean(),
    createdAt: zod_1.z.coerce.date(),
    updatedAt: zod_1.z.coerce.date(),
    completedAt: zod_1.z.coerce.date().nullable(),
    completedByUid: zod_1.z.string().nullable(),
    labels: zod_1.z.array(zod_1.z.string()),
});
exports.ProjectHealthContextSchema = zod_1.z.object({
    projectId: zod_1.z.string(),
    projectName: zod_1.z.string(),
    projectDescription: zod_1.z.string().nullable(),
    projectMetrics: exports.ProjectMetricsSchema,
    tasks: zod_1.z.array(exports.TaskContextSchema),
    language: zod_1.z.string().nullable().optional(),
});
exports.ProjectProgressSchema = zod_1.z.object({
    projectId: zod_1.z.string(),
    completedCount: zod_1.z.number().int(),
    activeCount: zod_1.z.number().int(),
    progressPercent: zod_1.z.number().int(),
});
exports.ProjectInsightSchema = zod_1.z.object({
    projectId: zod_1.z.string(),
    health: exports.ProjectHealthSchema.nullable(),
    progress: exports.ProjectProgressSchema.nullable(),
});
exports.WorkspaceInsightsSchema = zod_1.z.object({
    folderId: zod_1.z.string().nullable(),
    projectInsights: zod_1.z.array(exports.ProjectInsightSchema),
});
