// @doist/todoist-sdk/dist/cjs/types/projects/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROJECT_VIEW_STYLES = exports.ProjectSchema = exports.WorkspaceProjectSchema = exports.ProjectVisibilitySchema = exports.PROJECT_VISIBILITIES = exports.PersonalProjectSchema = exports.BaseProjectSchema = void 0;
const zod_1 = require("zod");
const url_helpers_1 = require("../../utils/url-helpers");
/**
 * Base schema for all project types in Todoist.
 * Contains common fields shared between personal and workspace projects.
 */
exports.BaseProjectSchema = zod_1.z.object({
    id: zod_1.z.string(),
    canAssignTasks: zod_1.z.boolean(),
    childOrder: zod_1.z.number().int(),
    color: zod_1.z.string(),
    createdAt: zod_1.z.coerce.date().nullable(),
    isArchived: zod_1.z.boolean(),
    isDeleted: zod_1.z.boolean(),
    isFavorite: zod_1.z.boolean(),
    isFrozen: zod_1.z.boolean(),
    name: zod_1.z.string(),
    updatedAt: zod_1.z.coerce.date().nullable(),
    viewStyle: zod_1.z.string(),
    defaultOrder: zod_1.z.number().int(),
    description: zod_1.z.string(),
    isCollapsed: zod_1.z.boolean(),
    isShared: zod_1.z.boolean(),
});
/**
 * Schema for personal projects in Todoist.
 */
exports.PersonalProjectSchema = exports.BaseProjectSchema.extend({
    parentId: zod_1.z.string().nullable(),
    inboxProject: zod_1.z.boolean().optional().default(false),
}).transform((data) => {
    return Object.assign(Object.assign({}, data), { url: (0, url_helpers_1.getProjectUrl)(data.id, data.name) });
});
/** Available project visibility levels. */
exports.PROJECT_VISIBILITIES = ['restricted', 'team', 'public'];
exports.ProjectVisibilitySchema = zod_1.z.enum(exports.PROJECT_VISIBILITIES);
/**
 * Schema for workspace projects in Todoist.
 */
exports.WorkspaceProjectSchema = exports.BaseProjectSchema.extend({
    access: zod_1.z.object({ visibility: exports.ProjectVisibilitySchema }).optional(),
    collaboratorRoleDefault: zod_1.z.string(),
    folderId: zod_1.z.string().nullable(),
    isInviteOnly: zod_1.z.boolean().nullable(),
    isLinkSharingEnabled: zod_1.z.boolean(),
    role: zod_1.z.string().nullable(),
    status: zod_1.z.string(),
    workspaceId: zod_1.z.string(),
}).transform((data) => {
    return Object.assign(Object.assign({}, data), { url: (0, url_helpers_1.getProjectUrl)(data.id, data.name) });
});
/**
 * Schema for any project type in Todoist (personal or workspace).
 * Dispatches to WorkspaceProjectSchema when workspaceId is present, PersonalProjectSchema otherwise.
 */
exports.ProjectSchema = zod_1.z.unknown().transform((val) => {
    if (typeof val === 'object' && val !== null && 'workspaceId' in val) {
        return exports.WorkspaceProjectSchema.parse(val);
    }
    return exports.PersonalProjectSchema.parse(val);
});
/** Available project view styles. */
exports.PROJECT_VIEW_STYLES = ['list', 'board', 'calendar'];
