// @doist/todoist-sdk/dist/esm/types/tasks/types.js
import { z } from 'zod';
import { hasUncompletablePrefix } from '../../utils/uncompletable-helpers.js';
import { getTaskUrl } from '../../utils/url-helpers.js';
export const DueDateSchema = z
    .object({
    isRecurring: z.boolean(),
    string: z.string(),
    date: z.string(),
})
    .extend({
    datetime: z.string().nullable().optional(),
    timezone: z.string().nullable().optional(),
    lang: z.string().nullable().optional(),
});
/** Available duration units for task deadlines. */
export const DURATION_UNITS = ['minute', 'day'];
export const DurationSchema = z.object({
    amount: z.number().positive('Value should be greater than zero'),
    unit: z.enum(DURATION_UNITS),
});
export const DeadlineSchema = z.object({
    date: z.string(),
    lang: z.string(),
});
export const TaskSchema = z
    .object({
    id: z.string(),
    userId: z.string(),
    projectId: z.string(),
    sectionId: z.string().nullable(),
    parentId: z.string().nullable(),
    addedByUid: z.string().nullable(),
    assignedByUid: z.string().nullable(),
    responsibleUid: z.string().nullable(),
    labels: z.array(z.string()),
    deadline: DeadlineSchema.nullable(),
    duration: DurationSchema.nullable(),
    checked: z.boolean(),
    isDeleted: z.boolean(),
    addedAt: z.coerce.date().nullable(),
    completedAt: z.coerce.date().nullable(),
    updatedAt: z.coerce.date().nullable(),
    due: DueDateSchema.nullable(),
    priority: z.number().int(),
    childOrder: z.number().int(),
    content: z.string(),
    description: z.string(),
    dayOrder: z.number().int(),
    isCollapsed: z.boolean(),
    isUncompletable: z.boolean().default(false),
})
    .transform((data) => {
    // Auto-detect uncompletable status from content prefix
    const isUncompletable = hasUncompletablePrefix(data.content);
    return Object.assign(Object.assign({}, data), { isUncompletable, url: getTaskUrl(data.id, data.content) });
});
