// @doist/todoist-sdk/dist/cjs/types/webhooks/comments.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommentDeletedPayloadSchema = exports.CommentUpdatedPayloadSchema = exports.CommentAddedPayloadSchema = exports.WebhookCommentSchema = exports.WebhookProjectCommentSchema = exports.WebhookItemCommentSchema = exports.COMMENT_WEBHOOK_EVENTS = exports.CommentWebhookEventEnum = void 0;
const zod_1 = require("zod");
const types_1 = require("../comments/types");
const types_2 = require("../tasks/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of comment (`note:*`) webhook event names. Single source
 * of truth for the discriminator values — `COMMENT_WEBHOOK_EVENTS` and every
 * `z.literal()` below derive from it.
 *
 * The on-the-wire event names remain `note:*` for historical reasons; the
 * SDK types use `Comment*` to match the rest of the public API where the
 * resource is named `Comment`.
 */
exports.CommentWebhookEventEnum = {
    Added: 'note:added',
    Updated: 'note:updated',
    Deleted: 'note:deleted',
};
/**
 * Webhook events that carry a comment as their `eventData`.
 */
exports.COMMENT_WEBHOOK_EVENTS = [
    exports.CommentWebhookEventEnum.Added,
    exports.CommentWebhookEventEnum.Updated,
    exports.CommentWebhookEventEnum.Deleted,
];
/**
 * Fields common to both item and project comment payloads.
 *
 * `postedAt` and `postedUid` are nullable here (unlike
 * {@link RawCommentSchema}) because webhook payloads may omit them — for
 * example when the poster could not be resolved, or on a deleted-note
 * tombstone.
 */
const WebhookCommentBaseSchema = zod_1.z.object({
    id: zod_1.z.string(),
    content: zod_1.z.string(),
    postedAt: zod_1.z.coerce.date().nullable(),
    postedUid: zod_1.z.string().nullable(),
    fileAttachment: types_1.AttachmentSchema.nullable(),
    uidsToNotify: zod_1.z.array(zod_1.z.string()).nullable(),
    reactions: zod_1.z.record(zod_1.z.string(), zod_1.z.array(zod_1.z.string())).nullable(),
    isDeleted: zod_1.z.boolean(),
});
/**
 * A comment attached to a task, delivered in a webhook payload.
 *
 * Webhook item-comment payloads carry two fields that don't appear on the
 * REST comment resource: `url` (the direct task URL) and `item` (the full
 * embedded {@link Task}). Both are parsed here so consumers handling a
 * `note:added` event also get the task it belongs to.
 *
 * `itemId` is exposed as `taskId` to match the rest of the SDK surface.
 */
exports.WebhookItemCommentSchema = WebhookCommentBaseSchema.extend({
    itemId: zod_1.z.string(),
    url: zod_1.z.string(),
    item: types_2.TaskSchema,
})
    .strict()
    .transform((_a) => {
    var { itemId } = _a, rest = __rest(_a, ["itemId"]);
    return (Object.assign(Object.assign({}, rest), { taskId: itemId }));
});
/**
 * A comment attached to a project, delivered in a webhook payload.
 *
 * Project comments do not carry the extra `url` or embedded project fields
 * that item comments carry for the parent task.
 */
exports.WebhookProjectCommentSchema = WebhookCommentBaseSchema.extend({
    projectId: zod_1.z.string(),
}).strict();
/**
 * The `eventData` of a `note:*` webhook payload — either an item comment
 * (includes `taskId` and an embedded {@link Task}) or a project comment
 * (includes `projectId`). Narrow with `'taskId' in eventData`.
 */
exports.WebhookCommentSchema = zod_1.z.union([exports.WebhookItemCommentSchema, exports.WebhookProjectCommentSchema]);
exports.CommentAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.CommentWebhookEventEnum.Added), eventData: exports.WebhookCommentSchema }));
exports.CommentUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.CommentWebhookEventEnum.Updated), eventData: exports.WebhookCommentSchema }));
exports.CommentDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.CommentWebhookEventEnum.Deleted), eventData: exports.WebhookCommentSchema }));
