// @doist/todoist-sdk/dist/cjs/types/webhooks/filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterDeletedPayloadSchema = exports.FilterUpdatedPayloadSchema = exports.FilterAddedPayloadSchema = exports.FILTER_WEBHOOK_EVENTS = exports.FilterWebhookEventEnum = void 0;
const zod_1 = require("zod");
const filters_1 = require("../sync/resources/filters");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of filter (`filter:*`) webhook event names. Single
 * source of truth for the discriminator values — `FILTER_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
exports.FilterWebhookEventEnum = {
    Added: 'filter:added',
    Updated: 'filter:updated',
    Deleted: 'filter:deleted',
};
/** Webhook events that carry a filter as their `eventData`. */
exports.FILTER_WEBHOOK_EVENTS = [
    exports.FilterWebhookEventEnum.Added,
    exports.FilterWebhookEventEnum.Updated,
    exports.FilterWebhookEventEnum.Deleted,
];
/**
 * Webhook filter payloads match the shared {@link Filter} resource shape,
 * so `eventData` reuses `FilterSchema` directly — no webhook-specific
 * schema is needed.
 */
exports.FilterAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.FilterWebhookEventEnum.Added), eventData: filters_1.FilterSchema }));
exports.FilterUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.FilterWebhookEventEnum.Updated), eventData: filters_1.FilterSchema }));
exports.FilterDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.FilterWebhookEventEnum.Deleted), eventData: filters_1.FilterSchema }));
