// @doist/todoist-sdk/dist/cjs/types/webhooks/labels.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabelDeletedPayloadSchema = exports.LabelUpdatedPayloadSchema = exports.LabelAddedPayloadSchema = exports.WebhookLabelSchema = exports.LABEL_WEBHOOK_EVENTS = exports.LabelWebhookEventEnum = void 0;
const zod_1 = require("zod");
const types_1 = require("../labels/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of label (`label:*`) webhook event names. Single source
 * of truth for the discriminator values — `LABEL_WEBHOOK_EVENTS` and every
 * `z.literal()` below derive from it.
 */
exports.LabelWebhookEventEnum = {
    Added: 'label:added',
    Updated: 'label:updated',
    Deleted: 'label:deleted',
};
/** Webhook events that carry a label as their `eventData`. */
exports.LABEL_WEBHOOK_EVENTS = [
    exports.LabelWebhookEventEnum.Added,
    exports.LabelWebhookEventEnum.Updated,
    exports.LabelWebhookEventEnum.Deleted,
];
/**
 * A label as delivered in a webhook payload.
 *
 * Webhook label payloads differ from the REST label shape in two ways:
 * they use `item_order` (not `order`) and always include `is_deleted`. This
 * schema reuses {@link LabelSchema}, swapping `order` for `itemOrder` on
 * the input side and renaming it back after validation, and surfaces
 * `isDeleted` for downstream filtering.
 */
exports.WebhookLabelSchema = types_1.LabelSchema.omit({ order: true })
    .extend({
    itemOrder: zod_1.z.number().int().nullable(),
    isDeleted: zod_1.z.boolean(),
})
    .transform((_a) => {
    var { itemOrder } = _a, rest = __rest(_a, ["itemOrder"]);
    return (Object.assign(Object.assign({}, rest), { order: itemOrder }));
});
exports.LabelAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.LabelWebhookEventEnum.Added), eventData: exports.WebhookLabelSchema }));
exports.LabelUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.LabelWebhookEventEnum.Updated), eventData: exports.WebhookLabelSchema }));
exports.LabelDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.LabelWebhookEventEnum.Deleted), eventData: exports.WebhookLabelSchema }));
