// @doist/todoist-sdk/dist/cjs/types/webhooks/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectUnarchivedPayloadSchema = exports.ProjectArchivedPayloadSchema = exports.ProjectDeletedPayloadSchema = exports.ProjectUpdatedPayloadSchema = exports.ProjectAddedPayloadSchema = exports.PROJECT_WEBHOOK_EVENTS = exports.ProjectWebhookEventEnum = void 0;
const zod_1 = require("zod");
const types_1 = require("../projects/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of project (`project:*`) webhook event names. Single
 * source of truth for the discriminator values — `PROJECT_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
exports.ProjectWebhookEventEnum = {
    Added: 'project:added',
    Updated: 'project:updated',
    Deleted: 'project:deleted',
    Archived: 'project:archived',
    Unarchived: 'project:unarchived',
};
/** Webhook events that carry a project as their `eventData`. */
exports.PROJECT_WEBHOOK_EVENTS = [
    exports.ProjectWebhookEventEnum.Added,
    exports.ProjectWebhookEventEnum.Updated,
    exports.ProjectWebhookEventEnum.Deleted,
    exports.ProjectWebhookEventEnum.Archived,
    exports.ProjectWebhookEventEnum.Unarchived,
];
/**
 * Webhook project payloads match the shape of the shared {@link Project}
 * resource — either a personal or a workspace project. No webhook-specific
 * schema is needed; `ProjectSchema` already dispatches on `workspaceId`.
 */
exports.ProjectAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ProjectWebhookEventEnum.Added), eventData: types_1.ProjectSchema }));
exports.ProjectUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ProjectWebhookEventEnum.Updated), eventData: types_1.ProjectSchema }));
exports.ProjectDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ProjectWebhookEventEnum.Deleted), eventData: types_1.ProjectSchema }));
exports.ProjectArchivedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ProjectWebhookEventEnum.Archived), eventData: types_1.ProjectSchema }));
exports.ProjectUnarchivedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ProjectWebhookEventEnum.Unarchived), eventData: types_1.ProjectSchema }));
