// @doist/todoist-sdk/dist/cjs/types/webhooks/reminders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReminderFiredPayloadSchema = exports.WebhookFiredReminderSchema = exports.REMINDER_WEBHOOK_EVENTS = exports.ReminderWebhookEventEnum = void 0;
const zod_1 = require("zod");
const reminders_1 = require("../sync/resources/reminders");
const types_1 = require("../tasks/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of reminder (`reminder:*`) webhook event names. Single
 * source of truth for the discriminator values — `REMINDER_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
exports.ReminderWebhookEventEnum = {
    Fired: 'reminder:fired',
};
/** Webhook events that carry a fired-reminder as their `eventData`. */
exports.REMINDER_WEBHOOK_EVENTS = [
    exports.ReminderWebhookEventEnum.Fired,
];
/**
 * A fired-reminder notification as delivered in a webhook payload.
 *
 * `reminder:fired` carries a subset of the full Reminder record (closer
 * to a fired-notification projection), so it doesn't line up with any of
 * the {@link Reminder} discriminated variants directly. This schema
 * reuses {@link ReminderBaseSchema} for the common identity fields and
 * adds the fields always present on a fired notification.
 */
exports.WebhookFiredReminderSchema = reminders_1.ReminderBaseSchema.extend({
    type: zod_1.z.enum(reminders_1.REMINDER_TYPES),
    isUrgent: zod_1.z.boolean(),
    minuteOffset: zod_1.z.number().int().nullable(),
    due: types_1.DueDateSchema.nullable(),
});
exports.ReminderFiredPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.ReminderWebhookEventEnum.Fired), eventData: exports.WebhookFiredReminderSchema }));
