// @doist/todoist-sdk/dist/cjs/types/webhooks/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskDeletedPayloadSchema = exports.TaskUncompletedPayloadSchema = exports.TaskCompletedPayloadSchema = exports.TaskUpdatedPayloadSchema = exports.TaskAddedPayloadSchema = exports.TaskUpdatedExtraSchema = exports.UpdateIntentSchema = exports.UPDATE_INTENTS = exports.TASK_WEBHOOK_EVENTS = exports.TaskWebhookEventEnum = void 0;
const zod_1 = require("zod");
const types_1 = require("../tasks/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of task (`item:*`) webhook event names. Single source of
 * truth for the discriminator values — `TASK_WEBHOOK_EVENTS` and every
 * `z.literal()` below derive from it.
 */
exports.TaskWebhookEventEnum = {
    Added: 'item:added',
    Updated: 'item:updated',
    Completed: 'item:completed',
    Uncompleted: 'item:uncompleted',
    Deleted: 'item:deleted',
};
/**
 * Webhook events that carry a {@link Task} as their `eventData`.
 *
 * The on-the-wire event names remain `item:*` for historical reasons; the
 * SDK types call them `Task*` to match the rest of the public API where the
 * resource is named `Task`.
 */
exports.TASK_WEBHOOK_EVENTS = [
    exports.TaskWebhookEventEnum.Added,
    exports.TaskWebhookEventEnum.Updated,
    exports.TaskWebhookEventEnum.Completed,
    exports.TaskWebhookEventEnum.Uncompleted,
    exports.TaskWebhookEventEnum.Deleted,
];
/**
 * Possible values for the `updateIntent` field on `item:updated`'s
 * `eventDataExtra`. `item_updated` is the default; the others narrow the
 * reason for the update (for example, a recurring completion or a change of
 * assignee).
 *
 * @see https://developer.todoist.com/sync/v1/#webhooks (Events Extra)
 */
exports.UPDATE_INTENTS = [
    'item_updated',
    'item_completed',
    'item_uncompleted',
    'responsible_uid_changed',
];
/**
 * An unrecognised value from a future backend release degrades gracefully to
 * `'item_updated'` rather than crashing the whole payload parse.
 */
exports.UpdateIntentSchema = zod_1.z.enum(exports.UPDATE_INTENTS).catch('item_updated');
/**
 * Extra info delivered alongside an `item:updated` event when the initiator
 * is the acting user. Contains the pre-update task state plus the reason for
 * the update — useful for diffing or for distinguishing manual edits from
 * recurring-task completions.
 *
 * @see https://developer.todoist.com/sync/v1/#webhooks (Events Extra)
 */
exports.TaskUpdatedExtraSchema = zod_1.z.object({
    oldItem: types_1.TaskSchema,
    updateIntent: exports.UpdateIntentSchema,
});
exports.TaskAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.TaskWebhookEventEnum.Added), eventData: types_1.TaskSchema }));
exports.TaskUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.TaskWebhookEventEnum.Updated), eventData: types_1.TaskSchema, eventDataExtra: exports.TaskUpdatedExtraSchema.optional() }));
exports.TaskCompletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.TaskWebhookEventEnum.Completed), eventData: types_1.TaskSchema }));
exports.TaskUncompletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.TaskWebhookEventEnum.Uncompleted), eventData: types_1.TaskSchema }));
exports.TaskDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.TaskWebhookEventEnum.Deleted), eventData: types_1.TaskSchema }));
