// @doist/todoist-sdk/dist/esm/types/webhooks/filters.js
import { z } from 'zod';
import { FilterSchema } from '../sync/resources/filters.js';
import { BaseWebhookEnvelopeShape } from './envelope.js';
/**
 * Enum-like object of filter (`filter:*`) webhook event names. Single
 * source of truth for the discriminator values — `FILTER_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
export const FilterWebhookEventEnum = {
    Added: 'filter:added',
    Updated: 'filter:updated',
    Deleted: 'filter:deleted',
};
/** Webhook events that carry a filter as their `eventData`. */
export const FILTER_WEBHOOK_EVENTS = [
    FilterWebhookEventEnum.Added,
    FilterWebhookEventEnum.Updated,
    FilterWebhookEventEnum.Deleted,
];
/**
 * Webhook filter payloads match the shared {@link Filter} resource shape,
 * so `eventData` reuses `FilterSchema` directly — no webhook-specific
 * schema is needed.
 */
export const FilterAddedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(FilterWebhookEventEnum.Added), eventData: FilterSchema }));
export const FilterUpdatedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(FilterWebhookEventEnum.Updated), eventData: FilterSchema }));
export const FilterDeletedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(FilterWebhookEventEnum.Deleted), eventData: FilterSchema }));
