// @doist/todoist-sdk/dist/esm/types/webhooks/labels.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
import { LabelSchema } from '../labels/types.js';
import { BaseWebhookEnvelopeShape } from './envelope.js';
/**
 * Enum-like object of label (`label:*`) webhook event names. Single source
 * of truth for the discriminator values — `LABEL_WEBHOOK_EVENTS` and every
 * `z.literal()` below derive from it.
 */
export const LabelWebhookEventEnum = {
    Added: 'label:added',
    Updated: 'label:updated',
    Deleted: 'label:deleted',
};
/** Webhook events that carry a label as their `eventData`. */
export const LABEL_WEBHOOK_EVENTS = [
    LabelWebhookEventEnum.Added,
    LabelWebhookEventEnum.Updated,
    LabelWebhookEventEnum.Deleted,
];
/**
 * A label as delivered in a webhook payload.
 *
 * Webhook label payloads differ from the REST label shape in two ways:
 * they use `item_order` (not `order`) and always include `is_deleted`. This
 * schema reuses {@link LabelSchema}, swapping `order` for `itemOrder` on
 * the input side and renaming it back after validation, and surfaces
 * `isDeleted` for downstream filtering.
 */
export const WebhookLabelSchema = LabelSchema.omit({ order: true })
    .extend({
    itemOrder: z.number().int().nullable(),
    isDeleted: z.boolean(),
})
    .transform((_a) => {
    var { itemOrder } = _a, rest = __rest(_a, ["itemOrder"]);
    return (Object.assign(Object.assign({}, rest), { order: itemOrder }));
});
export const LabelAddedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(LabelWebhookEventEnum.Added), eventData: WebhookLabelSchema }));
export const LabelUpdatedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(LabelWebhookEventEnum.Updated), eventData: WebhookLabelSchema }));
export const LabelDeletedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(LabelWebhookEventEnum.Deleted), eventData: WebhookLabelSchema }));
