// @doist/todoist-sdk/dist/esm/types/webhooks/projects.js
import { z } from 'zod';
import { ProjectSchema } from '../projects/types.js';
import { BaseWebhookEnvelopeShape } from './envelope.js';
/**
 * Enum-like object of project (`project:*`) webhook event names. Single
 * source of truth for the discriminator values — `PROJECT_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
export const ProjectWebhookEventEnum = {
    Added: 'project:added',
    Updated: 'project:updated',
    Deleted: 'project:deleted',
    Archived: 'project:archived',
    Unarchived: 'project:unarchived',
};
/** Webhook events that carry a project as their `eventData`. */
export const PROJECT_WEBHOOK_EVENTS = [
    ProjectWebhookEventEnum.Added,
    ProjectWebhookEventEnum.Updated,
    ProjectWebhookEventEnum.Deleted,
    ProjectWebhookEventEnum.Archived,
    ProjectWebhookEventEnum.Unarchived,
];
/**
 * Webhook project payloads match the shape of the shared {@link Project}
 * resource — either a personal or a workspace project. No webhook-specific
 * schema is needed; `ProjectSchema` already dispatches on `workspaceId`.
 */
export const ProjectAddedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ProjectWebhookEventEnum.Added), eventData: ProjectSchema }));
export const ProjectUpdatedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ProjectWebhookEventEnum.Updated), eventData: ProjectSchema }));
export const ProjectDeletedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ProjectWebhookEventEnum.Deleted), eventData: ProjectSchema }));
export const ProjectArchivedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ProjectWebhookEventEnum.Archived), eventData: ProjectSchema }));
export const ProjectUnarchivedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ProjectWebhookEventEnum.Unarchived), eventData: ProjectSchema }));
