// @doist/todoist-sdk/dist/esm/types/webhooks/reminders.js
import { z } from 'zod';
import { REMINDER_TYPES, ReminderBaseSchema } from '../sync/resources/reminders.js';
import { DueDateSchema } from '../tasks/types.js';
import { BaseWebhookEnvelopeShape } from './envelope.js';
/**
 * Enum-like object of reminder (`reminder:*`) webhook event names. Single
 * source of truth for the discriminator values — `REMINDER_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
export const ReminderWebhookEventEnum = {
    Fired: 'reminder:fired',
};
/** Webhook events that carry a fired-reminder as their `eventData`. */
export const REMINDER_WEBHOOK_EVENTS = [
    ReminderWebhookEventEnum.Fired,
];
/**
 * A fired-reminder notification as delivered in a webhook payload.
 *
 * `reminder:fired` carries a subset of the full Reminder record (closer
 * to a fired-notification projection), so it doesn't line up with any of
 * the {@link Reminder} discriminated variants directly. This schema
 * reuses {@link ReminderBaseSchema} for the common identity fields and
 * adds the fields always present on a fired notification.
 */
export const WebhookFiredReminderSchema = ReminderBaseSchema.extend({
    type: z.enum(REMINDER_TYPES),
    isUrgent: z.boolean(),
    minuteOffset: z.number().int().nullable(),
    due: DueDateSchema.nullable(),
});
export const ReminderFiredPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(ReminderWebhookEventEnum.Fired), eventData: WebhookFiredReminderSchema }));
