// @doist/todoist-sdk/dist/cjs/types/ui-extensions/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UiExtensionSchema = exports.SettingsExtensionSchema = exports.ComposerExtensionSchema = exports.ContextMenuExtensionSchema = exports.UiExtensionBaseSchema = exports.ComposerExtensionContextSchema = exports.COMPOSER_EXTENSION_CONTEXTS = exports.ContextMenuExtensionContextSchema = exports.CONTEXT_MENU_EXTENSION_CONTEXTS = exports.UI_EXTENSION_TYPES = exports.UiExtensionTypeEnum = void 0;
const zod_1 = require("zod");
const common_1 = require("../common");
/**
 * Discriminator values for the UI extension variants.
 * Exposed as an enum-like object because the values are used in `z.literal()`
 * across multiple variant schemas (see AGENTS.md "Discriminated unions" exception).
 */
exports.UiExtensionTypeEnum = {
    ContextMenu: 'context-menu',
    Composer: 'composer',
    Settings: 'settings',
};
/**
 * All UI extension type values.
 */
exports.UI_EXTENSION_TYPES = [
    exports.UiExtensionTypeEnum.ContextMenu,
    exports.UiExtensionTypeEnum.Composer,
    exports.UiExtensionTypeEnum.Settings,
];
/**
 * Context locations where a context-menu extension can appear.
 */
exports.CONTEXT_MENU_EXTENSION_CONTEXTS = ['project', 'task'];
exports.ContextMenuExtensionContextSchema = zod_1.z.enum(exports.CONTEXT_MENU_EXTENSION_CONTEXTS);
/**
 * Composer locations where a composer extension can appear.
 */
exports.COMPOSER_EXTENSION_CONTEXTS = ['task', 'comment'];
exports.ComposerExtensionContextSchema = zod_1.z.enum(exports.COMPOSER_EXTENSION_CONTEXTS);
/**
 * Shared fields across all UI extension variants.
 */
exports.UiExtensionBaseSchema = zod_1.z.object({
    id: common_1.StringOrNumberSchema,
    integrationId: common_1.StringOrNumberSchema,
    extensionId: zod_1.z.string(),
    url: zod_1.z.string(),
    icon: zod_1.z.string().nullable(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    width: zod_1.z.number().int().nullable(),
    height: zod_1.z.number().int().nullable(),
    defVersion: zod_1.z.number().int(),
    minimumCardistVersion: zod_1.z.string(),
});
/**
 * Schema for a context-menu UI extension.
 */
exports.ContextMenuExtensionSchema = exports.UiExtensionBaseSchema.extend({
    extensionType: zod_1.z.literal(exports.UiExtensionTypeEnum.ContextMenu),
    contextType: exports.ContextMenuExtensionContextSchema,
});
/**
 * Schema for a composer UI extension.
 */
exports.ComposerExtensionSchema = exports.UiExtensionBaseSchema.extend({
    extensionType: zod_1.z.literal(exports.UiExtensionTypeEnum.Composer),
    composerType: exports.ComposerExtensionContextSchema,
});
/**
 * Schema for a settings UI extension.
 */
exports.SettingsExtensionSchema = exports.UiExtensionBaseSchema.extend({
    extensionType: zod_1.z.literal(exports.UiExtensionTypeEnum.Settings),
});
/**
 * Discriminated union of all UI extension variants.
 */
exports.UiExtensionSchema = zod_1.z.discriminatedUnion('extensionType', [
    exports.ContextMenuExtensionSchema,
    exports.ComposerExtensionSchema,
    exports.SettingsExtensionSchema,
]);
