// @doist/todoist-sdk/dist/esm/types/comments/types.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
/** Available file attachment upload states. */
export const UPLOAD_STATES = ['pending', 'completed'];
export const AttachmentSchema = z
    .object({
    resourceType: z.string(),
})
    .extend({
    fileName: z.string().nullable().optional(),
    fileSize: z.number().int().nullable().optional(),
    fileType: z.string().nullable().optional(),
    fileUrl: z.string().nullable().optional(),
    fileDuration: z.number().int().nullable().optional(),
    uploadState: z.enum(UPLOAD_STATES).nullable().optional(),
    image: z.string().nullable().optional(),
    imageWidth: z.number().int().nullable().optional(),
    imageHeight: z.number().int().nullable().optional(),
    url: z.string().nullable().optional(),
    title: z.string().nullable().optional(),
});
export const RawCommentSchema = z
    .object({
    id: z.string(),
    itemId: z.string().optional(),
    projectId: z.string().optional(),
    content: z.string(),
    postedAt: z.coerce.date(),
    fileAttachment: AttachmentSchema.nullable(),
    postedUid: z.string(),
    uidsToNotify: z.array(z.string()).nullable(),
    reactions: z.record(z.string(), z.array(z.string())).nullable(),
    isDeleted: z.boolean(),
})
    .refine((data) => {
    // At least one of itemId or projectId must be present
    return ((data.itemId !== undefined && data.projectId === undefined) ||
        (data.itemId === undefined && data.projectId !== undefined));
}, {
    message: 'Exactly one of itemId or projectId must be provided',
});
export const CommentSchema = RawCommentSchema.transform((data) => {
    const { itemId } = data, rest = __rest(data, ["itemId"]);
    return Object.assign(Object.assign({}, rest), { 
        // Map itemId to taskId for backwards compatibility
        taskId: itemId });
});
