// @doist/todoist-sdk/dist/esm/types/sync/resources/suggestions.js
import { z } from 'zod';
/** Available template types. */
export const TEMPLATE_TYPES = ['project', 'setup'];
export const TemplateSuggestionSchema = z.looseObject({
    id: z.string(),
    name: z.string(),
    templateType: z.enum(TEMPLATE_TYPES),
});
export const WorkspaceTemplateSuggestionSchema = TemplateSuggestionSchema.extend({
    workspaceId: z.string().nullable(),
});
/** Available suggestion section types. */
export const SUGGESTION_SECTION_TYPES = ['templates', 'most_used_user_templates'];
export const SyncTemplateSuggestionsSchema = z.looseObject({
    type: z.enum(SUGGESTION_SECTION_TYPES),
    content: z.object({
        templates: z.array(TemplateSuggestionSchema),
        locale: z.string(),
    }),
    isDeleted: z.boolean(),
});
export const SyncWorkspaceTemplateSuggestionsSchema = z.looseObject({
    type: z.literal('most_used_workspace_templates'),
    content: z.object({
        templates: z.array(WorkspaceTemplateSuggestionSchema),
        locale: z.string(),
    }),
    isDeleted: z.boolean(),
});
export const SuggestionSchema = z.union([
    SyncWorkspaceTemplateSuggestionsSchema,
    SyncTemplateSuggestionsSchema,
]);
