// @doist/todoist-sdk/dist/esm/types/sync/resources/view-options.js
import { z } from 'zod';
/** Available view types. */
export const VIEW_TYPES = [
    'TODAY',
    'UPCOMING',
    'PROJECT',
    'LABEL',
    'FILTER',
    'WORKSPACE_FILTER',
    'SEARCH',
    'TEMPLATE_PREVIEW',
    'TASK_DETAIL',
    'AUTOMATION',
    'ASSIGNED',
    'OVERDUE',
    'WORKSPACE_OVERVIEW',
];
export const ViewTypeSchema = z.enum(VIEW_TYPES);
/** Available view modes. */
export const VIEW_MODES = ['LIST', 'BOARD', 'CALENDAR'];
export const ViewModeSchema = z.enum(VIEW_MODES);
/** Available grouping options. */
export const GROUPED_BY_OPTIONS = [
    'ASSIGNEE',
    'ADDED_DATE',
    'DUE_DATE',
    'DEADLINE',
    'LABEL',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
];
export const GroupedBySchema = z.enum(GROUPED_BY_OPTIONS).nullable();
/** Available sorting options. */
export const SORTED_BY_OPTIONS = [
    'MANUAL',
    'ALPHABETICALLY',
    'ASSIGNEE',
    'DUE_DATE',
    'DEADLINE',
    'ADDED_DATE',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
];
export const SortedBySchema = z.enum(SORTED_BY_OPTIONS).nullable();
/** Available sort directions. */
export const SORT_ORDERS = ['ASC', 'DESC'];
export const SortOrderSchema = z.enum(SORT_ORDERS).nullable();
/** Available calendar layout modes. */
export const CALENDAR_LAYOUTS = ['WEEK', 'MONTH'];
export const CalendarSettingsSchema = z.looseObject({
    layout: z.enum(CALENDAR_LAYOUTS).optional(),
});
export const ViewOptionsSchema = z.looseObject({
    viewType: ViewTypeSchema,
    objectId: z.string().optional(),
    groupedBy: GroupedBySchema.optional(),
    filteredBy: z.string().nullable().optional(),
    viewMode: ViewModeSchema.optional(),
    showCompletedTasks: z.boolean().optional(),
    sortedBy: SortedBySchema.optional(),
    sortOrder: SortOrderSchema.optional(),
});
export const ProjectViewOptionsDefaultsSchema = z.looseObject({
    projectId: z.string(),
    viewMode: ViewModeSchema.nullable().optional(),
    groupedBy: GroupedBySchema.optional(),
    sortedBy: SortedBySchema.optional(),
    sortOrder: SortOrderSchema.optional(),
    showCompletedTasks: z.boolean().optional(),
    filteredBy: z.string().nullable().optional(),
    calendarSettings: CalendarSettingsSchema.nullable().optional(),
});
