// @doist/todoist-sdk/dist/cjs/types/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TodoistArgumentError = exports.TodoistRequestError = void 0;
const ts_custom_error_1 = require("ts-custom-error");
const authenticationErrorCodes = [401, 403];
class TodoistRequestError extends ts_custom_error_1.CustomError {
    constructor(message, httpStatusCode, responseData) {
        super(message);
        this.message = message;
        this.httpStatusCode = httpStatusCode;
        this.responseData = responseData;
        this.isAuthenticationError = () => {
            if (!this.httpStatusCode) {
                return false;
            }
            return authenticationErrorCodes.includes(this.httpStatusCode);
        };
        Object.defineProperty(this, 'name', { value: 'TodoistRequestError' });
    }
}
exports.TodoistRequestError = TodoistRequestError;
class TodoistArgumentError extends ts_custom_error_1.CustomError {
    constructor(message) {
        super(message);
        this.message = message;
        Object.defineProperty(this, 'name', { value: 'TodoistArgumentError' });
    }
}
exports.TodoistArgumentError = TodoistArgumentError;
