// @doist/todoist-sdk/dist/cjs/test-utils/test-defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_INSTALLATION_ID = exports.DEFAULT_APP_ID = exports.DEFAULT_GOAL_NO_PROGRESS = exports.DEFAULT_GOAL = exports.DEFAULT_GOAL_ID = exports.DEFAULT_COLLABORATOR_STATE = exports.DEFAULT_COLLABORATOR = exports.DEFAULT_FOLDER = exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK = exports.INVALID_COMMENT = exports.DEFAULT_LOCATION_REMINDER = exports.DEFAULT_ABSOLUTE_REMINDER = exports.DEFAULT_RELATIVE_REMINDER = exports.DEFAULT_REMINDER_ID = exports.DEFAULT_COMMENT = exports.DEFAULT_RAW_COMMENT = exports.INVALID_ATTACHMENT = exports.DEFAULT_ATTACHMENT = exports.INVALID_USER = exports.DEFAULT_USER = exports.INVALID_LABEL = exports.DEFAULT_LABEL = exports.INVALID_SECTION = exports.DEFAULT_SECTION = exports.DEFAULT_WORKSPACE_PROJECT = exports.PROJECT_WITH_OPTIONALS_AS_NULL = exports.INVALID_PROJECT = exports.DEFAULT_PROJECT = exports.TASK_WITH_OPTIONALS_AS_NULL = exports.INVALID_TASK = exports.DEFAULT_TASK = exports.DEFAULT_DEADLINE = exports.DEFAULT_DURATION = exports.DEFAULT_DUE_DATE = exports.INVALID_ENTITY_ID = exports.DEFAULT_REQUEST_ID = exports.DEFAULT_AUTH_TOKEN = exports.DEFAULT_PROJECT_VIEW_STYLE = exports.DEFAULT_PROJECT_NAME = exports.DEFAULT_PROJECT_ID = exports.DEFAULT_ORDER = exports.DEFAULT_TASK_PRIORITY = exports.DEFAULT_TASK_DESCRIPTION = exports.DEFAULT_TASK_CONTENT = exports.DEFAULT_TASK_ID = void 0;
exports.DEFAULT_NOTE = exports.DEFAULT_UI_EXTENSION = exports.DEFAULT_USER_AUTHORIZATION = exports.DEFAULT_APP_INSTALLATION = exports.DEFAULT_APP_WEBHOOK = exports.DEFAULT_APP_DISTRIBUTION_TOKEN = exports.DEFAULT_APP_TEST_TOKEN = exports.DEFAULT_APP_VERIFICATION_TOKEN = exports.DEFAULT_APP_SECRETS = exports.DEFAULT_APP_WITH_USER_COUNT = exports.DEFAULT_APP = exports.DEFAULT_ACCESS_TOKEN_ID = exports.DEFAULT_DISTRIBUTION_TOKEN = exports.DEFAULT_UI_EXTENSION_ID = void 0;
const url_helpers_1 = require("../utils/url-helpers");
exports.DEFAULT_TASK_ID = '1234';
exports.DEFAULT_TASK_CONTENT = 'This is a task';
exports.DEFAULT_TASK_DESCRIPTION = 'A description';
exports.DEFAULT_TASK_PRIORITY = 1;
exports.DEFAULT_ORDER = 3;
exports.DEFAULT_PROJECT_ID = '123';
exports.DEFAULT_PROJECT_NAME = 'This is a project';
exports.DEFAULT_PROJECT_VIEW_STYLE = 'list';
const DEFAULT_LABEL_ID = '456';
const DEFAULT_LABEL_NAME = 'This is a label';
const DEFAULT_SECTION_ID = '456';
const DEFAULT_SECTION_NAME = 'This is a section';
const DEFAULT_PARENT_ID = '5678';
const DEFAULT_ASSIGNEE = '1234';
const DEFAULT_CREATOR = '1234';
const DEFAULT_DATE = '2020-09-08T12:00:00Z';
function defaultDate() {
    return new Date(DEFAULT_DATE);
}
const DEFAULT_ENTITY_COLOR = 'berry_red';
const DEFAULT_LABELS = ['personal', 'work', 'hobby'];
const DEFAULT_USER_ID = '5';
const DEFAULT_USER_NAME = 'A User';
const DEFAULT_USER_EMAIL = 'atestuser@doist.com';
const DEFAULT_COMMENT_ID = '4';
const DEFAULT_COMMENT_CONTENT = 'A comment';
const DEFAULT_COMMENT_REACTIONS = { '👍': ['1234', '5678'] };
const DEFAULT_CAN_ASSIGN_TASKS = true;
const DEFAULT_IS_ARCHIVED = false;
const DEFAULT_IS_DELETED = false;
const DEFAULT_IS_FROZEN = false;
const DEFAULT_IS_COLLAPSED = false;
// URL constants using the helper functions
const DEFAULT_TASK_URL = (0, url_helpers_1.getTaskUrl)(exports.DEFAULT_TASK_ID, exports.DEFAULT_TASK_CONTENT);
const DEFAULT_PROJECT_URL = (0, url_helpers_1.getProjectUrl)(exports.DEFAULT_PROJECT_ID, exports.DEFAULT_PROJECT_NAME);
const DEFAULT_SECTION_URL = (0, url_helpers_1.getSectionUrl)(DEFAULT_SECTION_ID, DEFAULT_SECTION_NAME);
exports.DEFAULT_AUTH_TOKEN = 'AToken';
exports.DEFAULT_REQUEST_ID = 'ARequestID';
exports.INVALID_ENTITY_ID = 1234;
exports.DEFAULT_DUE_DATE = {
    isRecurring: false,
    string: 'a date string',
    date: DEFAULT_DATE,
    lang: 'en',
    timezone: null,
};
exports.DEFAULT_DURATION = {
    amount: 10,
    unit: 'minute',
};
exports.DEFAULT_DEADLINE = {
    date: '2020-09-08',
    lang: 'en',
};
exports.DEFAULT_TASK = {
    id: exports.DEFAULT_TASK_ID,
    userId: DEFAULT_CREATOR,
    projectId: exports.DEFAULT_PROJECT_ID,
    sectionId: DEFAULT_SECTION_ID,
    parentId: DEFAULT_PARENT_ID,
    addedByUid: DEFAULT_CREATOR,
    assignedByUid: DEFAULT_CREATOR,
    responsibleUid: DEFAULT_ASSIGNEE,
    labels: DEFAULT_LABELS,
    deadline: exports.DEFAULT_DEADLINE,
    duration: exports.DEFAULT_DURATION,
    checked: false,
    isDeleted: DEFAULT_IS_DELETED,
    addedAt: defaultDate(),
    completedAt: null,
    updatedAt: defaultDate(),
    due: exports.DEFAULT_DUE_DATE,
    priority: exports.DEFAULT_TASK_PRIORITY,
    childOrder: exports.DEFAULT_ORDER,
    content: exports.DEFAULT_TASK_CONTENT,
    description: exports.DEFAULT_TASK_DESCRIPTION,
    dayOrder: exports.DEFAULT_ORDER,
    isCollapsed: DEFAULT_IS_COLLAPSED,
    isUncompletable: false,
    url: DEFAULT_TASK_URL,
};
exports.INVALID_TASK = Object.assign(Object.assign({}, exports.DEFAULT_TASK), { due: '2020-01-31' });
exports.TASK_WITH_OPTIONALS_AS_NULL = {
    userId: DEFAULT_CREATOR,
    id: exports.DEFAULT_TASK_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    sectionId: null,
    parentId: null,
    addedByUid: DEFAULT_CREATOR,
    assignedByUid: null,
    responsibleUid: null,
    labels: [],
    deadline: null,
    duration: null,
    checked: false,
    isDeleted: DEFAULT_IS_DELETED,
    addedAt: defaultDate(),
    completedAt: null,
    updatedAt: defaultDate(),
    due: null,
    priority: exports.DEFAULT_TASK_PRIORITY,
    childOrder: exports.DEFAULT_ORDER,
    content: exports.DEFAULT_TASK_CONTENT,
    description: exports.DEFAULT_TASK_DESCRIPTION,
    dayOrder: exports.DEFAULT_ORDER,
    isCollapsed: DEFAULT_IS_COLLAPSED,
    isUncompletable: false,
    url: DEFAULT_TASK_URL,
};
exports.DEFAULT_PROJECT = {
    id: exports.DEFAULT_PROJECT_ID,
    name: exports.DEFAULT_PROJECT_NAME,
    color: DEFAULT_ENTITY_COLOR,
    childOrder: exports.DEFAULT_ORDER,
    parentId: exports.DEFAULT_PROJECT_ID,
    isFavorite: false,
    isShared: false,
    inboxProject: false,
    viewStyle: exports.DEFAULT_PROJECT_VIEW_STYLE,
    canAssignTasks: DEFAULT_CAN_ASSIGN_TASKS,
    isArchived: DEFAULT_IS_ARCHIVED,
    isDeleted: DEFAULT_IS_DELETED,
    isFrozen: DEFAULT_IS_FROZEN,
    createdAt: defaultDate(),
    updatedAt: defaultDate(),
    defaultOrder: exports.DEFAULT_ORDER,
    description: '',
    isCollapsed: DEFAULT_IS_COLLAPSED,
    url: DEFAULT_PROJECT_URL,
};
exports.INVALID_PROJECT = Object.assign(Object.assign({}, exports.DEFAULT_PROJECT), { name: 123 });
exports.PROJECT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.DEFAULT_PROJECT), { parentId: null });
exports.DEFAULT_WORKSPACE_PROJECT = {
    id: exports.DEFAULT_PROJECT_ID,
    name: exports.DEFAULT_PROJECT_NAME,
    color: DEFAULT_ENTITY_COLOR,
    childOrder: exports.DEFAULT_ORDER,
    isFavorite: false,
    isShared: true,
    viewStyle: exports.DEFAULT_PROJECT_VIEW_STYLE,
    canAssignTasks: DEFAULT_CAN_ASSIGN_TASKS,
    isArchived: DEFAULT_IS_ARCHIVED,
    isDeleted: DEFAULT_IS_DELETED,
    isFrozen: DEFAULT_IS_FROZEN,
    createdAt: defaultDate(),
    updatedAt: defaultDate(),
    defaultOrder: exports.DEFAULT_ORDER,
    description: '',
    isCollapsed: DEFAULT_IS_COLLAPSED,
    url: DEFAULT_PROJECT_URL,
    workspaceId: '100',
    folderId: null,
    collaboratorRoleDefault: 'READ_WRITE',
    isInviteOnly: false,
    isLinkSharingEnabled: true,
    role: 'ADMIN',
    status: 'ACTIVE',
};
exports.DEFAULT_SECTION = {
    id: DEFAULT_SECTION_ID,
    userId: DEFAULT_USER_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    addedAt: new Date('2025-03-28T14:01:23.334881Z'),
    updatedAt: new Date('2025-03-28T14:01:23.334885Z'),
    archivedAt: null,
    name: DEFAULT_SECTION_NAME,
    description: null,
    sectionOrder: exports.DEFAULT_ORDER,
    isArchived: false,
    isDeleted: false,
    isCollapsed: false,
    url: DEFAULT_SECTION_URL,
};
exports.INVALID_SECTION = Object.assign(Object.assign({}, exports.DEFAULT_SECTION), { projectId: undefined });
exports.DEFAULT_LABEL = {
    id: DEFAULT_LABEL_ID,
    name: DEFAULT_LABEL_NAME,
    color: DEFAULT_ENTITY_COLOR,
    order: exports.DEFAULT_ORDER,
    isFavorite: false,
};
exports.INVALID_LABEL = Object.assign(Object.assign({}, exports.DEFAULT_LABEL), { isFavorite: 'true' });
exports.DEFAULT_USER = {
    id: DEFAULT_USER_ID,
    name: DEFAULT_USER_NAME,
    email: DEFAULT_USER_EMAIL,
};
exports.INVALID_USER = Object.assign(Object.assign({}, exports.DEFAULT_USER), { email: undefined });
exports.DEFAULT_ATTACHMENT = {
    resourceType: 'file',
    fileType: 'image/png',
    fileUrl: 'https://someurl.com/image.jpg',
    uploadState: 'completed',
};
exports.INVALID_ATTACHMENT = Object.assign(Object.assign({}, exports.DEFAULT_ATTACHMENT), { uploadState: 'something random' });
exports.DEFAULT_RAW_COMMENT = {
    id: DEFAULT_COMMENT_ID,
    postedUid: DEFAULT_USER_ID,
    content: DEFAULT_COMMENT_CONTENT,
    fileAttachment: exports.DEFAULT_ATTACHMENT,
    uidsToNotify: null,
    isDeleted: false,
    postedAt: defaultDate(),
    reactions: DEFAULT_COMMENT_REACTIONS,
    itemId: exports.DEFAULT_TASK_ID,
};
exports.DEFAULT_COMMENT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { taskId: exports.DEFAULT_RAW_COMMENT.itemId, itemId: undefined });
exports.DEFAULT_REMINDER_ID = '6XGgmFQrx44wfGHr';
exports.DEFAULT_RELATIVE_REMINDER = {
    id: exports.DEFAULT_REMINDER_ID,
    itemId: exports.DEFAULT_TASK_ID,
    notifyUid: DEFAULT_USER_ID,
    isDeleted: false,
    type: 'relative',
    minuteOffset: 30,
    due: exports.DEFAULT_DUE_DATE,
    isUrgent: false,
};
exports.DEFAULT_ABSOLUTE_REMINDER = {
    id: exports.DEFAULT_REMINDER_ID,
    itemId: exports.DEFAULT_TASK_ID,
    notifyUid: DEFAULT_USER_ID,
    isDeleted: false,
    type: 'absolute',
    due: exports.DEFAULT_DUE_DATE,
    isUrgent: true,
};
exports.DEFAULT_LOCATION_REMINDER = {
    id: exports.DEFAULT_REMINDER_ID,
    itemId: exports.DEFAULT_TASK_ID,
    notifyUid: DEFAULT_USER_ID,
    isDeleted: false,
    type: 'location',
    name: 'Aliados',
    locLat: '41.148581',
    locLong: '-8.610945000000015',
    locTrigger: 'on_enter',
    radius: 100,
};
exports.INVALID_COMMENT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { isDeleted: 'true' });
exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { fileAttachment: null, uidsToNotify: null, reactions: null });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_TASK = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK), { taskId: exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_TASK.itemId, itemId: undefined });
exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { fileAttachment: {
        resourceType: 'file',
        fileName: null,
        fileSize: null,
        fileType: null,
        fileDuration: null,
        uploadState: null,
        image: null,
        imageWidth: null,
        imageHeight: null,
        url: null,
        title: null,
    } });
exports.COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL), { taskId: exports.RAW_COMMENT_WITH_ATTACHMENT_WITH_OPTIONALS_AS_NULL.itemId, itemId: undefined });
exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = Object.assign(Object.assign({}, exports.DEFAULT_RAW_COMMENT), { itemId: undefined, projectId: exports.DEFAULT_PROJECT_ID });
exports.COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT = Object.assign(Object.assign({}, exports.RAW_COMMENT_WITH_OPTIONALS_AS_NULL_PROJECT), { taskId: undefined });
exports.DEFAULT_FOLDER = {
    id: '789',
    name: 'This is a folder',
    workspaceId: '100',
    defaultOrder: 3,
    childOrder: 3,
    isDeleted: false,
};
exports.DEFAULT_COLLABORATOR = {
    id: DEFAULT_USER_ID,
    email: DEFAULT_USER_EMAIL,
    fullName: DEFAULT_USER_NAME,
    timezone: 'UTC',
    imageId: null,
};
exports.DEFAULT_COLLABORATOR_STATE = {
    userId: DEFAULT_USER_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    state: 'active',
    isDeleted: false,
};
exports.DEFAULT_GOAL_ID = '550e8400-e29b-41d4-a716-446655440000';
exports.DEFAULT_GOAL = {
    id: exports.DEFAULT_GOAL_ID,
    ownerType: 'USER',
    ownerId: DEFAULT_USER_ID,
    name: 'Ship v2',
    description: 'Launch the new version',
    deadline: '2026-04-03',
    parentGoalId: null,
    childOrder: 0,
    isCompleted: false,
    completedAt: null,
    responsibleUid: null,
    isDeleted: false,
    progress: {
        totalTaskCount: 5,
        completedTaskCount: 2,
        percentage: 40,
    },
    creatorUid: DEFAULT_USER_ID,
    createdAt: defaultDate(),
    updatedAt: defaultDate(),
};
exports.DEFAULT_GOAL_NO_PROGRESS = Object.assign(Object.assign({}, exports.DEFAULT_GOAL), { id: 'e5f6g7h8-e29b-41d4-a716-446655440001', progress: undefined });
exports.DEFAULT_APP_ID = '42';
exports.DEFAULT_INSTALLATION_ID = '1001';
exports.DEFAULT_UI_EXTENSION_ID = '2002';
exports.DEFAULT_DISTRIBUTION_TOKEN = 'dist-token-abc';
exports.DEFAULT_ACCESS_TOKEN_ID = '3003';
exports.DEFAULT_APP = {
    id: exports.DEFAULT_APP_ID,
    clientId: 'client-id-xyz',
    status: 'public',
    displayName: 'Sample App',
    userId: DEFAULT_USER_ID,
    createdAt: defaultDate(),
    serviceUrl: 'https://example.com',
    oauthRedirectUri: 'https://example.com/callback',
    description: 'A sample app',
    iconSm: 'https://example.com/icon-sm.png',
    iconMd: 'https://example.com/icon-md.png',
    iconLg: 'https://example.com/icon-lg.png',
    appTokenScopes: ['data:read', 'task:add'],
};
exports.DEFAULT_APP_WITH_USER_COUNT = Object.assign(Object.assign({}, exports.DEFAULT_APP), { userCount: 12 });
exports.DEFAULT_APP_SECRETS = {
    clientId: 'client-id-xyz',
    clientSecret: 'client-secret-xyz',
};
exports.DEFAULT_APP_VERIFICATION_TOKEN = {
    verificationToken: 'verif-token-abc',
};
exports.DEFAULT_APP_TEST_TOKEN = {
    accessToken: 'test-token-xyz',
};
exports.DEFAULT_APP_DISTRIBUTION_TOKEN = {
    distributionToken: exports.DEFAULT_DISTRIBUTION_TOKEN,
};
exports.DEFAULT_APP_WEBHOOK = {
    status: 'active',
    callbackUrl: 'https://example.com/webhook',
    version: '1',
    events: ['item:added', 'item:completed'],
};
exports.DEFAULT_APP_INSTALLATION = {
    id: exports.DEFAULT_INSTALLATION_ID,
    creator: DEFAULT_USER_ID,
    createdTs: 1700000000,
    installationType: 'user',
    appTokenScopesValidated: true,
    integration: exports.DEFAULT_APP,
};
exports.DEFAULT_USER_AUTHORIZATION = {
    accessTokenId: exports.DEFAULT_ACCESS_TOKEN_ID,
    scope: ['data:read', 'task:add'],
    scopeDescriptions: ['Read data', 'Add tasks'],
    createdAt: new Date('2025-01-01T00:00:00Z'),
    app: {
        id: exports.DEFAULT_APP_ID,
        displayName: 'Sample App',
        description: 'A sample app',
        serviceUrl: 'https://example.com',
        iconMd: 'https://example.com/icon-md.png',
    },
};
exports.DEFAULT_UI_EXTENSION = {
    id: exports.DEFAULT_UI_EXTENSION_ID,
    integrationId: exports.DEFAULT_APP_ID,
    extensionId: 'ext-guid-1',
    extensionType: 'context-menu',
    contextType: 'task',
    url: 'https://example.com/extension',
    icon: 'https://example.com/icon.png',
    name: 'My Extension',
    description: 'Does things in the task context menu',
    width: 400,
    height: 600,
    defVersion: 2,
    minimumCardistVersion: '0.60',
};
exports.DEFAULT_NOTE = {
    id: DEFAULT_COMMENT_ID,
    projectId: exports.DEFAULT_PROJECT_ID,
    content: DEFAULT_COMMENT_CONTENT,
    postedAt: defaultDate(),
    fileAttachment: exports.DEFAULT_ATTACHMENT,
    postedUid: DEFAULT_USER_ID,
    uidsToNotify: null,
    reactions: DEFAULT_COMMENT_REACTIONS,
    isDeleted: false,
};
