// @doist/todoist-sdk/dist/cjs/types/billing/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PricingResponseSchema = exports.PricingVersionSchema = exports.PricingTermsSchema = exports.PricesResponseSchema = exports.ProPlanDetailsSchema = exports.PriceListingSchema = exports.FormattedPriceSchema = exports.SubscriptionInfoSchema = exports.BillingPlanPriceSchema = exports.TAX_BEHAVIORS = exports.BILLING_CYCLES = exports.BILLING_ACTIVATION_METHODS = exports.BILLING_PLANS = exports.BILLING_SUBSCRIPTION_STATUSES = void 0;
const zod_1 = require("zod");
const types_1 = require("../workspaces/types");
/** Available subscription statuses. */
exports.BILLING_SUBSCRIPTION_STATUSES = [
    'none',
    'autorenew',
    'canceled',
    'expired',
    'trial',
    'trial_canceled',
];
/** Available plan tiers. */
exports.BILLING_PLANS = ['free', 'pro', 'pro_gifted', 'pro_legacy'];
/** Ways an active subscription can be purchased. */
exports.BILLING_ACTIVATION_METHODS = [
    'stripe',
    'appstore',
    'playstore',
    'teams',
    'manual',
];
/** Available billing cycles. */
exports.BILLING_CYCLES = ['monthly', 'yearly'];
/** Available tax behaviors for a price. */
exports.TAX_BEHAVIORS = ['exclusive', 'inclusive', 'unspecified'];
exports.BillingPlanPriceSchema = zod_1.z.object({
    amount: zod_1.z.string(),
    rawAmount: zod_1.z.number(),
    currency: zod_1.z.string(),
    billingCycle: zod_1.z.enum(exports.BILLING_CYCLES).nullable(),
    taxBehavior: zod_1.z.enum(exports.TAX_BEHAVIORS),
});
exports.SubscriptionInfoSchema = zod_1.z.object({
    status: zod_1.z.enum(exports.BILLING_SUBSCRIPTION_STATUSES),
    plan: zod_1.z.enum(exports.BILLING_PLANS),
    expirationDate: zod_1.z.string().nullable(),
    activationMethod: zod_1.z.enum(exports.BILLING_ACTIVATION_METHODS),
    planPrice: exports.BillingPlanPriceSchema.nullable(),
    billingPortalUrl: zod_1.z.string().nullable(),
    billingPortalSwitchToAnnualUrl: zod_1.z.string().nullable(),
    hasBillingPortal: zod_1.z.boolean(),
    hasBillingPortalSwitchToAnnual: zod_1.z.boolean(),
    invoiceCreditBalance: zod_1.z.record(zod_1.z.string(), zod_1.z.number().int()).nullable(),
    hasSwitchLegacyToCurrent: zod_1.z.boolean(),
});
exports.FormattedPriceSchema = zod_1.z.object({
    currency: zod_1.z.string(),
    unitAmount: zod_1.z.number().int(),
    taxBehavior: zod_1.z.enum(exports.TAX_BEHAVIORS),
});
exports.PriceListingSchema = zod_1.z.object({
    billingCycle: zod_1.z.enum(exports.BILLING_CYCLES),
    prices: zod_1.z.array(exports.FormattedPriceSchema),
});
exports.ProPlanDetailsSchema = zod_1.z.object({
    currentPlanStatus: zod_1.z.enum(types_1.WORKSPACE_PLAN_STATUSES),
    downgradeAt: zod_1.z.string().nullable(),
    priceList: zod_1.z.array(exports.PriceListingSchema),
});
exports.PricesResponseSchema = zod_1.z.object({
    pro: zod_1.z.array(exports.PriceListingSchema),
    teams: zod_1.z.array(exports.PriceListingSchema),
});
/**
 * Monthly and yearly amounts for a single currency. Each amount is a number in
 * the currency's smallest unit, or a localized formatted string when the
 * `pricing` endpoint is called with `formatted: true`.
 */
exports.PricingTermsSchema = zod_1.z.object({
    monthly: zod_1.z.union([zod_1.z.number(), zod_1.z.string()]),
    yearly: zod_1.z.union([zod_1.z.number(), zod_1.z.string()]),
});
/**
 * A pricing version entry: plan name (`pro` / `biz`) → currency code →
 * {@link PricingTerms}.
 */
exports.PricingVersionSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.record(zod_1.z.string(), exports.PricingTermsSchema));
/**
 * Response of the `pricing` endpoint: fixed version-pointer string fields plus
 * dynamic version keys (`v1`, `v2`, …), each a {@link PricingVersion}.
 */
exports.PricingResponseSchema = zod_1.z
    .object({
    latestPro: zod_1.z.string(),
    latestBiz: zod_1.z.string(),
    sessionPro: zod_1.z.string(),
    sessionBiz: zod_1.z.string(),
})
    .catchall(exports.PricingVersionSchema);
