// @doist/todoist-mcp/dist/index.js
import { f as s, s as a, l as e, m as t, a as o, u as d, r as i, d as r, b as n, c as l, e as c, h as p, i as m, j as f, k, n as T, o as u, p as g, q as j, t as v, v as P, w as h, x as S, y as b, z as A, A as y, B as C, C as G, D as F, E as x, F as E, G as L, H as O, I as w, J as H, K as I, L as M, M as q, N as z, O as B, P as D, Q as N, R, S as W } from "./mcp-server-DUs6oq-a.js";
import { T as V, g as _ } from "./mcp-server-DUs6oq-a.js";
import { r as Y, v as Z } from "./require-valid-todoist-token-BeOx8xgI.js";
const K = {
  // Task management tools
  addTasks: W,
  completeTasks: R,
  uncompleteTasks: N,
  updateTasks: D,
  findTasks: B,
  findTasksByDate: z,
  findCompletedTasks: q,
  rescheduleTasks: M,
  // Project management tools
  addProjects: I,
  updateProjects: H,
  findProjects: w,
  // Section management tools
  addSections: O,
  updateSections: L,
  findSections: E,
  // Goal management tools
  findGoals: x,
  addGoals: F,
  updateGoals: G,
  completeGoals: C,
  linkGoalTasks: y,
  // Comment management tools
  addComments: A,
  updateComments: b,
  findComments: S,
  // Attachment tools
  viewAttachment: h,
  // Label management tools
  addLabels: P,
  updateLabels: v,
  findLabels: j,
  // Filter management tools
  findFilters: g,
  addFilters: u,
  updateFilters: T,
  // Activity and audit tools
  findActivity: k,
  getProductivityStats: f,
  // Health and insights tools
  getProjectHealth: m,
  getProjectActivityStats: p,
  analyzeProjectHealth: c,
  getWorkspaceInsights: l,
  // General tools
  getOverview: n,
  deleteObject: r,
  reorderObjects: i,
  userInfo: d,
  // Assignment and collaboration tools
  findProjectCollaborators: o,
  manageAssignments: t,
  // Workspace tools
  listWorkspaces: e,
  // OpenAI MCP tools
  search: a,
  fetch: s
};
export {
  V as FEATURE_NAMES,
  A as addComments,
  u as addFilters,
  F as addGoals,
  P as addLabels,
  I as addProjects,
  O as addSections,
  W as addTasks,
  c as analyzeProjectHealth,
  C as completeGoals,
  R as completeTasks,
  r as deleteObject,
  s as fetch,
  k as findActivity,
  S as findComments,
  q as findCompletedTasks,
  g as findFilters,
  x as findGoals,
  j as findLabels,
  o as findProjectCollaborators,
  w as findProjects,
  E as findSections,
  B as findTasks,
  z as findTasksByDate,
  _ as getMcpServer,
  n as getOverview,
  f as getProductivityStats,
  p as getProjectActivityStats,
  m as getProjectHealth,
  l as getWorkspaceInsights,
  y as linkGoalTasks,
  e as listWorkspaces,
  t as manageAssignments,
  i as reorderObjects,
  Y as requireValidTodoistToken,
  M as rescheduleTasks,
  a as search,
  K as tools,
  N as uncompleteTasks,
  b as updateComments,
  T as updateFilters,
  G as updateGoals,
  v as updateLabels,
  H as updateProjects,
  L as updateSections,
  D as updateTasks,
  d as userInfo,
  Z as validateTodoistToken,
  h as viewAttachment
};
