// @doist/todoist-sdk/dist/esm/authentication.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { v4 as uuid } from 'uuid';
import { getAuthBaseUri, getSyncBaseUri, ENDPOINT_AUTHORIZATION, ENDPOINT_GET_TOKEN, ENDPOINT_REVOKE, ENDPOINT_REGISTER, ENDPOINT_REST_ACCESS_TOKENS_MIGRATE, } from './consts/endpoints.js';
import { request, isSuccess } from './transport/http-client.js';
import { TodoistRequestError } from './types/index.js';
/** Available OAuth2 permission scopes. */
export const PERMISSIONS = [
    'task:add',
    'data:read',
    'data:read_write',
    'data:delete',
    'project:delete',
    'backups:read',
    'billing:read',
    'billing:read_write',
    'dev:app_console',
];
/** Supported token endpoint authentication methods for dynamic client registration. */
export const TOKEN_ENDPOINT_AUTH_METHODS = [
    'client_secret_post',
    'client_secret_basic',
    'none',
];
/**
 * Creates a Basic Authentication header value from client credentials.
 * @param clientId - The OAuth client ID
 * @param clientSecret - The OAuth client secret
 * @returns The Basic Auth header value (without the 'Basic ' prefix)
 */
function createBasicAuthHeader(clientId, clientSecret) {
    const credentials = `${clientId}:${clientSecret}`;
    return Buffer.from(credentials).toString('base64');
}
/**
 * Generates a random state parameter for OAuth2 authorization.
 * The state parameter helps prevent CSRF attacks.
 *
 * @example
 * ```typescript
 * const state = getAuthStateParameter()
 * // Store state in session
 * const authUrl = getAuthorizationUrl(clientId, ['data:read'], state)
 * ```
 *
 * @returns A random UUID v4 string
 */
export function getAuthStateParameter() {
    return uuid();
}
/**
 * Generates the authorization URL for the OAuth2 flow.
 *
 * The `clientId` can be either a traditional client ID string (e.g. from
 * {@link registerClient}) or an HTTPS URL pointing to a client metadata document,
 * as defined in {@link https://drafts.ietf.org/doc/draft-ietf-oauth-client-id-metadata-document/ RFC draft-ietf-oauth-client-id-metadata-document}.
 *
 * @example
 * ```typescript
 * const url = getAuthorizationUrl({
 *   clientId: 'your-client-id',
 *   permissions: ['data:read', 'task:add'],
 *   state,
 * })
 * // Redirect user to url
 * ```
 *
 * @returns The full authorization URL to redirect users to
 * @see https://developer.todoist.com/api/v1/#tag/Authorization/OAuth
 */
export function getAuthorizationUrl({ clientId, permissions, state, baseUrl, }) {
    if (!(permissions === null || permissions === void 0 ? void 0 : permissions.length)) {
        throw new Error('At least one scope value should be passed for permissions.');
    }
    const scope = permissions.join(',');
    return `${getAuthBaseUri(baseUrl)}${ENDPOINT_AUTHORIZATION}?client_id=${encodeURIComponent(clientId)}&scope=${scope}&state=${state}`;
}
/**
 * Exchanges an authorization code for an access token.
 *
 * @example
 * ```typescript
 * const { accessToken } = await getAuthToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   code: authCode
 * })
 * ```
 *
 * @returns The access token response
 * @throws {@link TodoistRequestError} If the token exchange fails
 */
export async function getAuthToken(args, options) {
    var _a;
    if (typeof options === 'string') {
        throw new TypeError('Passing baseUrl as a string is no longer supported. Use an options object instead: getAuthToken(args, { baseUrl })');
    }
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    try {
        const response = await request({
            httpMethod: 'POST',
            baseUri: getAuthBaseUri(baseUrl),
            relativePath: ENDPOINT_GET_TOKEN,
            apiToken: undefined,
            payload: args,
            customFetch,
        });
        if (response.status !== 200 || !((_a = response.data) === null || _a === void 0 ? void 0 : _a.accessToken)) {
            throw new TodoistRequestError('Authentication token exchange failed.', response.status, response.data);
        }
        return response.data;
    }
    catch (error) {
        // Re-throw with custom message for authentication failures
        const err = error;
        throw new TodoistRequestError('Authentication token exchange failed.', err.httpStatusCode, err.responseData);
    }
}
/**
 * Revokes a token using the RFC 7009 OAuth 2.0 Token Revocation standard.
 *
 * This function uses HTTP Basic Authentication with client credentials and follows
 * the RFC 7009 specification for token revocation.
 *
 * @example
 * ```typescript
 * await revokeToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   token: 'access-token-to-revoke'
 * })
 * ```
 *
 * @returns True if revocation was successful
 * @see https://datatracker.ietf.org/doc/html/rfc7009
 * @see https://developer.todoist.com/api/v1/#tag/Authorization
 */
export async function revokeToken(args, options) {
    if (typeof options === 'string') {
        throw new TypeError('Passing baseUrl as a string is no longer supported. Use an options object instead: revokeToken(args, { baseUrl })');
    }
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    const { clientId, clientSecret, token } = args;
    // Create Basic Auth header as per RFC 7009
    const basicAuth = createBasicAuthHeader(clientId, clientSecret);
    const customHeaders = {
        Authorization: `Basic ${basicAuth}`,
    };
    // Request body only contains the token and optional token_type_hint
    const requestBody = {
        token,
        token_type_hint: 'access_token',
    };
    const response = await request({
        httpMethod: 'POST',
        baseUri: getSyncBaseUri(baseUrl),
        relativePath: ENDPOINT_REVOKE,
        apiToken: undefined,
        payload: requestBody,
        requestId: undefined,
        hasSyncCommands: false,
        customHeaders: customHeaders,
        customFetch,
    });
    return isSuccess(response);
}
/**
 * Migrates a personal API token to an OAuth access token.
 *
 * This allows applications to transition users from personal API tokens
 * to proper OAuth tokens without requiring the user to go through the
 * full OAuth authorization flow.
 *
 * @example
 * ```typescript
 * const { accessToken } = await migratePersonalToken({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   personalToken: 'user-personal-token',
 *   scope: 'data:read_write,data:delete'
 * })
 * ```
 *
 * @returns The new OAuth token response
 * @throws {@link TodoistRequestError} If the migration fails
 */
export async function migratePersonalToken(args, options) {
    var _a;
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    try {
        const response = await request({
            httpMethod: 'POST',
            baseUri: getSyncBaseUri(baseUrl),
            relativePath: ENDPOINT_REST_ACCESS_TOKENS_MIGRATE,
            apiToken: undefined,
            payload: Object.assign(Object.assign({}, args), { scope: args.scope.join(',') }),
            customFetch,
        });
        if (response.status !== 200 || !((_a = response.data) === null || _a === void 0 ? void 0 : _a.accessToken)) {
            throw new TodoistRequestError('Personal token migration failed.', response.status, response.data);
        }
        return response.data;
    }
    catch (error) {
        const err = error;
        throw new TodoistRequestError('Personal token migration failed.', err.httpStatusCode, err.responseData);
    }
}
/**
 * Registers a new OAuth client via Dynamic Client Registration (RFC 7591).
 *
 * @example
 * ```typescript
 * const client = await registerClient({
 *   redirectUris: ['https://example.com/callback'],
 *   clientName: 'My App',
 *   scope: ['data:read_write', 'task:add'],
 * })
 * // Use client.clientId and client.clientSecret for OAuth flows
 * ```
 *
 * @returns The registered client details
 * @throws {@link TodoistRequestError} If the registration fails
 * @see {@link https://datatracker.ietf.org/doc/html/rfc7591 RFC 7591}
 */
export async function registerClient(args, options) {
    var _a, _b;
    const baseUrl = options === null || options === void 0 ? void 0 : options.baseUrl;
    const customFetch = options === null || options === void 0 ? void 0 : options.customFetch;
    try {
        const response = await request({
            httpMethod: 'POST',
            baseUri: getAuthBaseUri(baseUrl),
            relativePath: ENDPOINT_REGISTER,
            apiToken: undefined,
            payload: Object.assign(Object.assign({}, args), { scope: (_a = args.scope) === null || _a === void 0 ? void 0 : _a.join(' ') }),
            customFetch,
        });
        if (!isSuccess(response) || !((_b = response.data) === null || _b === void 0 ? void 0 : _b.clientId)) {
            throw new TodoistRequestError('Dynamic client registration failed.', response.status, response.data);
        }
        const _c = response.data, { clientIdIssuedAt, clientSecretExpiresAt, scope } = _c, rest = __rest(_c, ["clientIdIssuedAt", "clientSecretExpiresAt", "scope"]);
        return Object.assign(Object.assign({}, rest), { scope: scope ? scope.split(' ') : undefined, clientIdIssuedAt: clientIdIssuedAt !== undefined ? new Date(clientIdIssuedAt * 1000) : undefined, clientSecretExpiresAt: clientSecretExpiresAt === undefined
                ? undefined
                : clientSecretExpiresAt === 0
                    ? null
                    : new Date(clientSecretExpiresAt * 1000) });
    }
    catch (error) {
        const err = error;
        throw new TodoistRequestError('Dynamic client registration failed.', err.httpStatusCode, err.responseData);
    }
}
