// @doist/todoist-sdk/dist/cjs/types/sections/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SectionSchema = exports.SectionBaseSchema = void 0;
const zod_1 = require("zod");
const url_helpers_1 = require("../../utils/url-helpers");
/**
 * Raw section fields (the persisted shape). Exposed so schemas that extend
 * or override a field — notably the webhook variant — can reuse the base
 * without duplicating every field.
 */
exports.SectionBaseSchema = zod_1.z.object({
    id: zod_1.z.string(),
    userId: zod_1.z.string(),
    projectId: zod_1.z.string(),
    addedAt: zod_1.z.coerce.date(),
    updatedAt: zod_1.z.coerce.date(),
    archivedAt: zod_1.z.coerce.date().nullable(),
    name: zod_1.z.string(),
    description: zod_1.z.string().nullable(),
    sectionOrder: zod_1.z.number().int(),
    isArchived: zod_1.z.boolean(),
    isDeleted: zod_1.z.boolean(),
    isCollapsed: zod_1.z.boolean(),
});
exports.SectionSchema = exports.SectionBaseSchema.transform((data) => (Object.assign(Object.assign({}, data), { url: (0, url_helpers_1.getSectionUrl)(data.id, data.name) })));
