// @doist/todoist-sdk/dist/cjs/types/webhooks/sections.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SectionUnarchivedPayloadSchema = exports.SectionArchivedPayloadSchema = exports.SectionDeletedPayloadSchema = exports.SectionUpdatedPayloadSchema = exports.SectionAddedPayloadSchema = exports.WebhookSectionSchema = exports.SECTION_WEBHOOK_EVENTS = exports.SectionWebhookEventEnum = void 0;
const zod_1 = require("zod");
const url_helpers_1 = require("../../utils/url-helpers");
const types_1 = require("../sections/types");
const envelope_1 = require("./envelope");
/**
 * Enum-like object of section (`section:*`) webhook event names. Single
 * source of truth for the discriminator values — `SECTION_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
exports.SectionWebhookEventEnum = {
    Added: 'section:added',
    Updated: 'section:updated',
    Deleted: 'section:deleted',
    Archived: 'section:archived',
    Unarchived: 'section:unarchived',
};
/** Webhook events that carry a section as their `eventData`. */
exports.SECTION_WEBHOOK_EVENTS = [
    exports.SectionWebhookEventEnum.Added,
    exports.SectionWebhookEventEnum.Updated,
    exports.SectionWebhookEventEnum.Deleted,
    exports.SectionWebhookEventEnum.Archived,
    exports.SectionWebhookEventEnum.Unarchived,
];
/**
 * A section as delivered in a webhook payload.
 *
 * Reuses {@link SectionBaseSchema} and overrides `updatedAt` to tolerate
 * payloads that deliver it as `null` or omit the key entirely — tombstones
 * and never-updated sections come through without an `updated_at` value.
 * The result is normalised to `Date | null` so consumers don't have to
 * handle three shapes. `description` is similarly tolerated (older webhook
 * payloads may omit it) and normalised to `string | null`. Surfaces the
 * computed `url` so the exposed shape matches {@link Section}.
 */
exports.WebhookSectionSchema = types_1.SectionBaseSchema.extend({
    updatedAt: zod_1.z.coerce
        .date()
        .nullish()
        .transform((value) => value !== null && value !== void 0 ? value : null),
    description: zod_1.z
        .string()
        .nullish()
        .transform((value) => value !== null && value !== void 0 ? value : null),
}).transform((data) => (Object.assign(Object.assign({}, data), { url: (0, url_helpers_1.getSectionUrl)(data.id, data.name) })));
exports.SectionAddedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.SectionWebhookEventEnum.Added), eventData: exports.WebhookSectionSchema }));
exports.SectionUpdatedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.SectionWebhookEventEnum.Updated), eventData: exports.WebhookSectionSchema }));
exports.SectionDeletedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.SectionWebhookEventEnum.Deleted), eventData: exports.WebhookSectionSchema }));
exports.SectionArchivedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.SectionWebhookEventEnum.Archived), eventData: exports.WebhookSectionSchema }));
exports.SectionUnarchivedPayloadSchema = zod_1.z.object(Object.assign(Object.assign({}, envelope_1.BaseWebhookEnvelopeShape), { eventName: zod_1.z.literal(exports.SectionWebhookEventEnum.Unarchived), eventData: exports.WebhookSectionSchema }));
