// @doist/todoist-sdk/dist/esm/types/billing/types.js
import { z } from 'zod';
import { WORKSPACE_PLAN_STATUSES } from '../workspaces/types.js';
/** Available subscription statuses. */
export const BILLING_SUBSCRIPTION_STATUSES = [
    'none',
    'autorenew',
    'canceled',
    'expired',
    'trial',
    'trial_canceled',
];
/** Available plan tiers. */
export const BILLING_PLANS = ['free', 'pro', 'pro_gifted', 'pro_legacy'];
/** Ways an active subscription can be purchased. */
export const BILLING_ACTIVATION_METHODS = [
    'stripe',
    'appstore',
    'playstore',
    'teams',
    'manual',
];
/** Available billing cycles. */
export const BILLING_CYCLES = ['monthly', 'yearly'];
/** Available tax behaviors for a price. */
export const TAX_BEHAVIORS = ['exclusive', 'inclusive', 'unspecified'];
export const BillingPlanPriceSchema = z.object({
    amount: z.string(),
    rawAmount: z.number(),
    currency: z.string(),
    billingCycle: z.enum(BILLING_CYCLES).nullable(),
    taxBehavior: z.enum(TAX_BEHAVIORS),
});
export const SubscriptionInfoSchema = z.object({
    status: z.enum(BILLING_SUBSCRIPTION_STATUSES),
    plan: z.enum(BILLING_PLANS),
    expirationDate: z.string().nullable(),
    activationMethod: z.enum(BILLING_ACTIVATION_METHODS),
    planPrice: BillingPlanPriceSchema.nullable(),
    billingPortalUrl: z.string().nullable(),
    billingPortalSwitchToAnnualUrl: z.string().nullable(),
    hasBillingPortal: z.boolean(),
    hasBillingPortalSwitchToAnnual: z.boolean(),
    invoiceCreditBalance: z.record(z.string(), z.number().int()).nullable(),
    hasSwitchLegacyToCurrent: z.boolean(),
});
export const FormattedPriceSchema = z.object({
    currency: z.string(),
    unitAmount: z.number().int(),
    taxBehavior: z.enum(TAX_BEHAVIORS),
});
export const PriceListingSchema = z.object({
    billingCycle: z.enum(BILLING_CYCLES),
    prices: z.array(FormattedPriceSchema),
});
export const ProPlanDetailsSchema = z.object({
    currentPlanStatus: z.enum(WORKSPACE_PLAN_STATUSES),
    downgradeAt: z.string().nullable(),
    priceList: z.array(PriceListingSchema),
});
export const PricesResponseSchema = z.object({
    pro: z.array(PriceListingSchema),
    teams: z.array(PriceListingSchema),
});
/**
 * Monthly and yearly amounts for a single currency. Each amount is a number in
 * the currency's smallest unit, or a localized formatted string when the
 * `pricing` endpoint is called with `formatted: true`.
 */
export const PricingTermsSchema = z.object({
    monthly: z.union([z.number(), z.string()]),
    yearly: z.union([z.number(), z.string()]),
});
/**
 * A pricing version entry: plan name (`pro` / `biz`) → currency code →
 * {@link PricingTerms}.
 */
export const PricingVersionSchema = z.record(z.string(), z.record(z.string(), PricingTermsSchema));
/**
 * Response of the `pricing` endpoint: fixed version-pointer string fields plus
 * dynamic version keys (`v1`, `v2`, …), each a {@link PricingVersion}.
 */
export const PricingResponseSchema = z
    .object({
    latestPro: z.string(),
    latestBiz: z.string(),
    sessionPro: z.string(),
    sessionBiz: z.string(),
})
    .catchall(PricingVersionSchema);
