// @doist/todoist-sdk/dist/esm/types/webhooks/sections.js
import { z } from 'zod';
import { getSectionUrl } from '../../utils/url-helpers.js';
import { SectionBaseSchema } from '../sections/types.js';
import { BaseWebhookEnvelopeShape } from './envelope.js';
/**
 * Enum-like object of section (`section:*`) webhook event names. Single
 * source of truth for the discriminator values — `SECTION_WEBHOOK_EVENTS`
 * and every `z.literal()` below derive from it.
 */
export const SectionWebhookEventEnum = {
    Added: 'section:added',
    Updated: 'section:updated',
    Deleted: 'section:deleted',
    Archived: 'section:archived',
    Unarchived: 'section:unarchived',
};
/** Webhook events that carry a section as their `eventData`. */
export const SECTION_WEBHOOK_EVENTS = [
    SectionWebhookEventEnum.Added,
    SectionWebhookEventEnum.Updated,
    SectionWebhookEventEnum.Deleted,
    SectionWebhookEventEnum.Archived,
    SectionWebhookEventEnum.Unarchived,
];
/**
 * A section as delivered in a webhook payload.
 *
 * Reuses {@link SectionBaseSchema} and overrides `updatedAt` to tolerate
 * payloads that deliver it as `null` or omit the key entirely — tombstones
 * and never-updated sections come through without an `updated_at` value.
 * The result is normalised to `Date | null` so consumers don't have to
 * handle three shapes. `description` is similarly tolerated (older webhook
 * payloads may omit it) and normalised to `string | null`. Surfaces the
 * computed `url` so the exposed shape matches {@link Section}.
 */
export const WebhookSectionSchema = SectionBaseSchema.extend({
    updatedAt: z.coerce
        .date()
        .nullish()
        .transform((value) => value !== null && value !== void 0 ? value : null),
    description: z
        .string()
        .nullish()
        .transform((value) => value !== null && value !== void 0 ? value : null),
}).transform((data) => (Object.assign(Object.assign({}, data), { url: getSectionUrl(data.id, data.name) })));
export const SectionAddedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(SectionWebhookEventEnum.Added), eventData: WebhookSectionSchema }));
export const SectionUpdatedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(SectionWebhookEventEnum.Updated), eventData: WebhookSectionSchema }));
export const SectionDeletedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(SectionWebhookEventEnum.Deleted), eventData: WebhookSectionSchema }));
export const SectionArchivedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(SectionWebhookEventEnum.Archived), eventData: WebhookSectionSchema }));
export const SectionUnarchivedPayloadSchema = z.object(Object.assign(Object.assign({}, BaseWebhookEnvelopeShape), { eventName: z.literal(SectionWebhookEventEnum.Unarchived), eventData: WebhookSectionSchema }));
