// @doist/todoist-sdk/dist/cjs/clients/app-client.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppClient = void 0;
const zod_1 = require("zod");
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const apps_1 = require("../types/apps");
const multipart_upload_1 = require("../utils/multipart-upload");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all developer-app management endpoints
 * (apps, secrets, tokens, webhooks, installations, user authorizations).
 * All endpoints require the `dev:app_console` OAuth scope.
 *
 * Instantiated by `TodoistApi`; every public app-management method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing JSDoc.
 */
class AppClient extends base_client_1.BaseClient {
    async getApps(requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppArray)(response.data);
    }
    async getApp(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppWithUserCount)(response.data);
    }
    async addApp(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateApp)(response.data);
    }
    async updateApp(appId, args, requestId) {
        validators_1.IdSchema.parse(appId);
        const { appTokenScopes } = args, rest = __rest(args, ["appTokenScopes"]);
        const payload = Object.assign({}, rest);
        if (appTokenScopes !== undefined) {
            payload.appTokenScopes = appTokenScopes === null ? null : appTokenScopes.join(',');
        }
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: payload,
            requestId: requestId,
        });
        return (0, validators_1.validateAppWithUserCount)(response.data);
    }
    async deleteApp(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async getAppSecrets(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppSecretsEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppSecrets)(response.data);
    }
    async resetAppClientSecret(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppClientSecretEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppWithUserCount)(response.data);
    }
    async revokeAppTokens(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppTokensEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppWithUserCount)(response.data);
    }
    async uploadAppIcon({ appId, file, fileName = '', size = 'medium' }, requestId) {
        validators_1.IdSchema.parse(appId);
        if (Buffer.isBuffer(file) && file.length === 0) {
            throw new Error('Cannot upload empty image file');
        }
        const data = await (0, multipart_upload_1.uploadMultipartFile)({
            baseUrl: this.apiRootBase,
            authToken: this.authToken,
            endpoint: (0, endpoints_1.getAppIconEndpoint)(appId, size),
            file,
            fileName,
            additionalFields: { file_name: fileName },
            requestId: requestId,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateAppWithUserCount)(data);
    }
    async getAppTestToken(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppTestTokenEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppTestToken)(response.data);
    }
    async createAppTestToken(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'PUT',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppTestTokenEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppTestToken)(response.data);
    }
    async getAppDistributionToken(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppDistributionTokenEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppDistributionToken)(response.data);
    }
    async getAppVerificationToken(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppVerificationTokenEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppVerificationToken)(response.data);
    }
    async resetAppVerificationToken(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppVerificationTokenEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppVerificationToken)(response.data);
    }
    async getAppByDistributionToken(distributionToken, requestId) {
        zod_1.z.string().min(1).parse(distributionToken);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppByDistributionTokenEndpoint)(distributionToken),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppByDistributionToken)(response.data);
    }
    // --- App webhooks ---
    async getAppWebhook(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppWebhookEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return response.data ? (0, validators_1.validateAppWebhook)(response.data) : null;
    }
    async updateAppWebhook(args, requestId) {
        const { appId, events } = args, rest = __rest(args, ["appId", "events"]);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppWebhookEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, rest), { events: events.map(apps_1.webhookEventToWireName) }),
            requestId: requestId,
        });
        return (0, validators_1.validateAppWebhook)(response.data);
    }
    async deleteAppWebhook(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppWebhookEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    // --- App installations ---
    async getAppInstallations(requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS_INSTALLATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppInstallationArray)(response.data);
    }
    async installApp(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS_INSTALLATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateAppInstallation)(response.data);
    }
    async getAppInstallation(installationId, requestId) {
        validators_1.IdSchema.parse(installationId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppInstallationEndpoint)(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateAppInstallation)(response.data);
    }
    async updateAppInstallation(installationId, args, requestId) {
        validators_1.IdSchema.parse(installationId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppInstallationEndpoint)(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateAppInstallation)(response.data);
    }
    async uninstallApp(installationId, requestId) {
        validators_1.IdSchema.parse(installationId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getAppInstallationEndpoint)(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    // --- User authorizations ---
    async getUserAuthorizations(requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_USER_AUTHORIZATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateUserAuthorizationArray)(response.data);
    }
    async revokeUserAuthorization(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_USER_AUTHORIZATIONS_DELETE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.AppClient = AppClient;
