// @doist/todoist-sdk/dist/cjs/clients/folder-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FolderClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all folder-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public folder method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class FolderClient extends base_client_1.BaseClient {
    async getFolders(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_FOLDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateFolderArray)(results),
            nextCursor,
        };
    }
    async getFolder(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateFolder)(response.data);
    }
    async addFolder(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_FOLDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateFolder)(response.data);
    }
    async updateFolder(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateFolder)(response.data);
    }
    async deleteFolder(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.FolderClient = FolderClient;
