// @doist/todoist-sdk/dist/cjs/clients/insights-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InsightsClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all insights-domain endpoints
 * (project activity stats, health, progress, workspace insights).
 *
 * Instantiated by `TodoistApi`; every public insights method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class InsightsClient extends base_client_1.BaseClient {
    async getProjectActivityStats(projectId, args = {}) {
        validators_1.IdSchema.parse(projectId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getProjectInsightsActivityStatsEndpoint)(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign({ objectType: 'ITEM', eventType: 'COMPLETED' }, args),
        });
        return (0, validators_1.validateProjectActivityStats)(response.data);
    }
    async getProjectHealth(projectId) {
        validators_1.IdSchema.parse(projectId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getProjectInsightsHealthEndpoint)(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProjectHealth)(response.data);
    }
    async getProjectHealthContext(projectId) {
        validators_1.IdSchema.parse(projectId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getProjectInsightsHealthContextEndpoint)(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProjectHealthContext)(response.data);
    }
    async getProjectProgress(projectId) {
        validators_1.IdSchema.parse(projectId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getProjectInsightsProgressEndpoint)(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProjectProgress)(response.data);
    }
    async getWorkspaceInsights(workspaceId, args = {}) {
        validators_1.IdSchema.parse(workspaceId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getWorkspaceInsightsEndpoint)(workspaceId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, args), (args.projectIds ? { projectIds: args.projectIds.join(',') } : {})),
        });
        return (0, validators_1.validateWorkspaceInsights)(response.data);
    }
    async analyzeProjectHealth(projectId, requestId) {
        validators_1.IdSchema.parse(projectId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, endpoints_1.getProjectInsightsHealthAnalyzeEndpoint)(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateProjectHealth)(response.data);
    }
}
exports.InsightsClient = InsightsClient;
