// @doist/todoist-sdk/dist/cjs/clients/productivity-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductivityClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling the productivity-stats endpoint.
 *
 * Instantiated by `TodoistApi`; the public `getProductivityStats` method
 * on `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class ProductivityClient extends base_client_1.BaseClient {
    async getProductivityStats() {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PRODUCTIVITY,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProductivityStats)(response.data);
    }
}
exports.ProductivityClient = ProductivityClient;
