// @doist/todoist-sdk/dist/cjs/clients/project-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all project-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public project method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class ProjectClient extends base_client_1.BaseClient {
    async getProject(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProject)(response.data);
    }
    async getProjects(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateProjectArray)(results),
            nextCursor,
        };
    }
    async searchProjects(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateProjectArray)(results),
            nextCursor,
        };
    }
    async getArchivedProjects(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_ARCHIVED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateProjectArray)(results),
            nextCursor,
        };
    }
    async addProject(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data);
    }
    async updateProject(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data);
    }
    async deleteProject(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async archiveProject(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id, endpoints_1.PROJECT_ARCHIVE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data);
    }
    async unarchiveProject(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id, endpoints_1.PROJECT_UNARCHIVE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data);
    }
    async moveProjectToWorkspace(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_MOVE_TO_WORKSPACE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data.project);
    }
    async moveProjectToPersonal(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_MOVE_TO_PERSONAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateProject)(response.data.project);
    }
    async getArchivedProjectsCount(args = {}) {
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_ARCHIVED_COUNT,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return data;
    }
    async getProjectPermissions() {
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_PROJECTS_PERMISSIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return data;
    }
    async getFullProject(id, args = {}) {
        validators_1.IdSchema.parse(id);
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id, endpoints_1.ENDPOINT_REST_PROJECT_FULL),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            project: data.project ? (0, validators_1.validateProject)(data.project) : null,
            commentsCount: data.commentsCount,
            tasks: (0, validators_1.validateTaskArray)(data.tasks),
            sections: (0, validators_1.validateSectionArray)(data.sections),
            collaborators: (0, validators_1.validateUserArray)(data.collaborators),
            notes: (0, validators_1.validateCommentArray)(data.notes),
        };
    }
    async joinProject(id, requestId) {
        validators_1.IdSchema.parse(id);
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, id, endpoints_1.ENDPOINT_REST_PROJECT_JOIN),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return {
            project: (0, validators_1.validateWorkspaceProject)(data.project),
            tasks: (0, validators_1.validateTaskArray)(data.tasks),
            sections: (0, validators_1.validateSectionArray)(data.sections),
            comments: (0, validators_1.validateNoteArray)(data.comments),
            collaborators: (0, validators_1.validateCollaboratorArray)(data.collaborators),
            collaboratorStates: (0, validators_1.validateCollaboratorStateArray)(data.collaboratorStates),
            folder: data.folder ? (0, validators_1.validateFolder)(data.folder) : null,
            subprojects: (0, validators_1.validateWorkspaceProjectArray)(data.subprojects),
        };
    }
    async getProjectCollaborators(projectId, args = {}) {
        validators_1.IdSchema.parse(projectId);
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_PROJECTS, projectId, endpoints_1.ENDPOINT_REST_PROJECT_COLLABORATORS),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateUserArray)(results),
            nextCursor,
        };
    }
}
exports.ProjectClient = ProjectClient;
