// @doist/todoist-sdk/dist/cjs/clients/task-client.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskClient = void 0;
const uuid_1 = require("uuid");
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const sync_request_1 = require("../transport/sync-request");
const types_1 = require("../types");
const request_helpers_1 = require("../utils/request-helpers");
const uncompletable_helpers_1 = require("../utils/uncompletable-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
const MAX_COMMAND_COUNT = 100;
/**
 * Internal sub-client handling all task-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public task method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class TaskClient extends base_client_1.BaseClient {
    async getTask(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateTask)(response.data);
    }
    async getTasks(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateTaskArray)(results),
            nextCursor,
        };
    }
    async getTasksByFilter(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS_FILTER,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateTaskArray)(results),
            nextCursor,
        };
    }
    async getCompletedTasksByCompletionDate(args) {
        const { data: { items, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: (0, validators_1.validateTaskArray)(items),
            nextCursor,
        };
    }
    async getCompletedTasksByDueDate(args) {
        const { data: { items, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: (0, validators_1.validateTaskArray)(items),
            nextCursor,
        };
    }
    async searchCompletedTasks(args) {
        const { data: { items, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS_COMPLETED_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: (0, validators_1.validateTaskArray)(items),
            nextCursor,
        };
    }
    async getAllCompletedTasks(args = {}) {
        const { since, until } = args, rest = __rest(args, ["since", "until"]);
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS_COMPLETED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign(Object.assign({}, rest), (since ? { since: since.toISOString() } : {})), (until ? { until: until.toISOString() } : {})),
        });
        return {
            projects: data.projects,
            sections: data.sections,
            items: (0, validators_1.validateTaskArray)(data.items),
        };
    }
    async addTask(args, requestId) {
        // Process content based on isUncompletable flag
        const processedArgs = Object.assign(Object.assign({}, args), { content: (0, uncompletable_helpers_1.processTaskContent)(args.content, args.isUncompletable) });
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TASKS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
            requestId: requestId,
        });
        return (0, validators_1.validateTask)(response.data);
    }
    async quickAddTask(args) {
        // Process text based on isUncompletable flag
        const processedArgs = Object.assign(Object.assign({}, args), { text: (0, uncompletable_helpers_1.processTaskContent)(args.text, args.isUncompletable) });
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_SYNC_QUICK_ADD,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
        });
        return (0, validators_1.validateTask)(response.data);
    }
    async updateTask(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        // Translate SDK alias for due-date clearing to Todoist's accepted payload value.
        const normalizedArgs = args.dueString === null ? Object.assign(Object.assign({}, args), { dueString: 'no date' }) : args;
        // Process content if both content and isUncompletable are provided
        const processedArgs = normalizedArgs.content && normalizedArgs.isUncompletable !== undefined
            ? Object.assign(Object.assign({}, normalizedArgs), { content: (0, uncompletable_helpers_1.processTaskContent)(normalizedArgs.content, normalizedArgs.isUncompletable) }) : normalizedArgs;
        // Remap `order` → `childOrder` so snakeCaseKeys() produces `child_order`
        const { order } = processedArgs, argsWithoutOrder = __rest(processedArgs, ["order"]);
        const remappedArgs = order !== undefined ? Object.assign(Object.assign({}, argsWithoutOrder), { childOrder: order }) : argsWithoutOrder;
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: remappedArgs,
            requestId: requestId,
        });
        return (0, validators_1.validateTask)(response.data);
    }
    async moveTasks(ids, args, requestId) {
        var _a;
        if (ids.length > MAX_COMMAND_COUNT) {
            throw new types_1.TodoistRequestError(`Maximum number of items is ${MAX_COMMAND_COUNT}`, 400);
        }
        const commands = ids.map((id) => ({
            type: 'item_move',
            uuid: (0, uuid_1.v4)(),
            args: Object.assign(Object.assign(Object.assign({ id }, (0, request_helpers_1.spreadIfDefined)(args.projectId, (v) => ({ projectId: v }))), (0, request_helpers_1.spreadIfDefined)(args.sectionId, (v) => ({ sectionId: v }))), (0, request_helpers_1.spreadIfDefined)(args.parentId, (v) => ({ parentId: v }))),
        }));
        const syncRequest = {
            commands,
            resourceTypes: ['items'],
        };
        const syncResponse = await (0, sync_request_1.performSyncRequest)(this.syncContext, syncRequest, {
            requestId,
            hasSyncCommands: true,
        });
        if (!((_a = syncResponse.items) === null || _a === void 0 ? void 0 : _a.length)) {
            throw new types_1.TodoistRequestError('Tasks not found', 404);
        }
        const syncTasks = syncResponse.items.filter((task) => ids.includes(task.id));
        if (!syncTasks.length) {
            throw new types_1.TodoistRequestError('Tasks not found', 404);
        }
        return (0, validators_1.validateTaskArray)(syncTasks);
    }
    async moveTask(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id, endpoints_1.ENDPOINT_REST_TASK_MOVE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign(Object.assign({}, (0, request_helpers_1.spreadIfDefined)(args.projectId, (v) => ({ project_id: v }))), (0, request_helpers_1.spreadIfDefined)(args.sectionId, (v) => ({ section_id: v }))), (0, request_helpers_1.spreadIfDefined)(args.parentId, (v) => ({ parent_id: v }))),
            requestId: requestId,
        });
        return (0, validators_1.validateTask)(response.data);
    }
    async closeTask(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id, endpoints_1.ENDPOINT_REST_TASK_CLOSE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async reopenTask(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id, endpoints_1.ENDPOINT_REST_TASK_REOPEN),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async deleteTask(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.TaskClient = TaskClient;
