// @doist/todoist-sdk/dist/cjs/types/sync/resources/reminders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReminderSchema = exports.RelativeReminderSchema = exports.AbsoluteReminderSchema = exports.LocationReminderSchema = exports.LOCATION_TRIGGERS = exports.REMINDER_TYPES = exports.ReminderTypeEnum = exports.ReminderBaseSchema = void 0;
const zod_1 = require("zod");
const types_1 = require("../../tasks/types");
exports.ReminderBaseSchema = zod_1.z.looseObject({
    id: zod_1.z.string(),
    notifyUid: zod_1.z.string(),
    itemId: zod_1.z.string(),
    projectId: zod_1.z.string().optional(),
    isDeleted: zod_1.z.boolean(),
});
/** Enum-like object of reminder type values. */
exports.ReminderTypeEnum = {
    Relative: 'relative',
    Absolute: 'absolute',
    Location: 'location',
};
/** Available reminder types. */
exports.REMINDER_TYPES = [
    exports.ReminderTypeEnum.Relative,
    exports.ReminderTypeEnum.Absolute,
    exports.ReminderTypeEnum.Location,
];
/** Available location reminder triggers. */
exports.LOCATION_TRIGGERS = ['on_enter', 'on_leave'];
exports.LocationReminderSchema = exports.ReminderBaseSchema.extend({
    type: zod_1.z.literal(exports.ReminderTypeEnum.Location),
    name: zod_1.z.string(),
    locLat: zod_1.z.string(),
    locLong: zod_1.z.string(),
    locTrigger: zod_1.z.enum(exports.LOCATION_TRIGGERS),
    radius: zod_1.z.number().int(),
});
exports.AbsoluteReminderSchema = exports.ReminderBaseSchema.extend({
    type: zod_1.z.literal(exports.ReminderTypeEnum.Absolute),
    due: types_1.DueDateSchema,
    isUrgent: zod_1.z.boolean().optional(),
});
exports.RelativeReminderSchema = exports.ReminderBaseSchema.extend({
    type: zod_1.z.literal(exports.ReminderTypeEnum.Relative),
    minuteOffset: zod_1.z.number().int(),
    due: types_1.DueDateSchema.optional(),
    isUrgent: zod_1.z.boolean().optional(),
});
exports.ReminderSchema = zod_1.z.discriminatedUnion('type', [
    exports.LocationReminderSchema,
    exports.AbsoluteReminderSchema,
    exports.RelativeReminderSchema,
]);
