// @doist/todoist-sdk/dist/cjs/types/sync/resources/suggestions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuggestionSchema = exports.SyncWorkspaceTemplateSuggestionsSchema = exports.SyncTemplateSuggestionsSchema = exports.SUGGESTION_SECTION_TYPES = exports.WorkspaceTemplateSuggestionSchema = exports.TemplateSuggestionSchema = exports.TEMPLATE_TYPES = void 0;
const zod_1 = require("zod");
/** Available template types. */
exports.TEMPLATE_TYPES = ['project', 'setup'];
exports.TemplateSuggestionSchema = zod_1.z.looseObject({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    templateType: zod_1.z.enum(exports.TEMPLATE_TYPES),
});
exports.WorkspaceTemplateSuggestionSchema = exports.TemplateSuggestionSchema.extend({
    workspaceId: zod_1.z.string().nullable(),
});
/** Available suggestion section types. */
exports.SUGGESTION_SECTION_TYPES = ['templates', 'most_used_user_templates'];
exports.SyncTemplateSuggestionsSchema = zod_1.z.looseObject({
    type: zod_1.z.enum(exports.SUGGESTION_SECTION_TYPES),
    content: zod_1.z.object({
        templates: zod_1.z.array(exports.TemplateSuggestionSchema),
        locale: zod_1.z.string(),
    }),
    isDeleted: zod_1.z.boolean(),
});
exports.SyncWorkspaceTemplateSuggestionsSchema = zod_1.z.looseObject({
    type: zod_1.z.literal('most_used_workspace_templates'),
    content: zod_1.z.object({
        templates: zod_1.z.array(exports.WorkspaceTemplateSuggestionSchema),
        locale: zod_1.z.string(),
    }),
    isDeleted: zod_1.z.boolean(),
});
exports.SuggestionSchema = zod_1.z.union([
    exports.SyncWorkspaceTemplateSuggestionsSchema,
    exports.SyncTemplateSuggestionsSchema,
]);
