// @doist/todoist-sdk/dist/cjs/types/goals/types.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoalSchema = exports.GoalProgressSchema = exports.GOAL_OWNER_TYPES = void 0;
const zod_1 = require("zod");
/** Available goal owner types. */
exports.GOAL_OWNER_TYPES = ['USER', 'WORKSPACE'];
/**
 * Progress of a {@link Goal}.
 *
 * The REST API returns these counts under `total_item_count` / `completed_item_count`,
 * reflecting the backend's generic "item" terminology. The SDK normalizes them to
 * `totalTaskCount` / `completedTaskCount` so the public surface consistently speaks
 * in "tasks" (matching the rest of the SDK). The `preprocess` step below performs
 * that rename after the transport layer's snake_case → camelCase conversion — do not
 * remove it without also changing the downstream field names.
 */
exports.GoalProgressSchema = zod_1.z.preprocess((raw) => {
    if (raw === null || raw === undefined || typeof raw !== 'object' || Array.isArray(raw)) {
        return raw;
    }
    const _a = raw, { totalItemCount, completedItemCount, totalTaskCount, completedTaskCount } = _a, rest = __rest(_a, ["totalItemCount", "completedItemCount", "totalTaskCount", "completedTaskCount"]);
    return Object.assign(Object.assign({}, rest), { totalTaskCount: totalTaskCount !== null && totalTaskCount !== void 0 ? totalTaskCount : totalItemCount, completedTaskCount: completedTaskCount !== null && completedTaskCount !== void 0 ? completedTaskCount : completedItemCount });
}, zod_1.z.looseObject({
    totalTaskCount: zod_1.z.number().int(),
    completedTaskCount: zod_1.z.number().int(),
    percentage: zod_1.z.number(),
}));
exports.GoalSchema = zod_1.z.looseObject({
    id: zod_1.z.string(),
    ownerType: zod_1.z.enum(exports.GOAL_OWNER_TYPES),
    ownerId: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string().nullable(),
    deadline: zod_1.z.string().nullable(),
    parentGoalId: zod_1.z.string().nullable(),
    childOrder: zod_1.z.number().int(),
    isCompleted: zod_1.z.boolean(),
    completedAt: zod_1.z.coerce.date().nullable(),
    responsibleUid: zod_1.z.string().nullable(),
    isDeleted: zod_1.z.boolean(),
    progress: exports.GoalProgressSchema.optional(),
    creatorUid: zod_1.z.string(),
    createdAt: zod_1.z.coerce.date(),
    updatedAt: zod_1.z.coerce.date(),
});
