// @doist/todoist-sdk/dist/cjs/types/productivity/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorSchema = exports.ProductivityStatsSchema = exports.KarmaUpdateSchema = exports.ItemsWithDateSchema = exports.CompletedItemSchema = exports.StreakSchema = void 0;
const zod_1 = require("zod");
const user_preferences_1 = require("../sync/user-preferences");
exports.StreakSchema = zod_1.z.object({
    count: zod_1.z.number(),
    start: zod_1.z.string(),
    end: zod_1.z.string(),
});
exports.CompletedItemSchema = zod_1.z.object({
    id: zod_1.z.string(),
    completed: zod_1.z.number(),
});
exports.ItemsWithDateSchema = zod_1.z.object({
    items: zod_1.z.array(exports.CompletedItemSchema),
    totalCompleted: zod_1.z.number(),
});
exports.KarmaUpdateSchema = zod_1.z.object({
    time: zod_1.z.string(),
    newKarma: zod_1.z.number(),
    positiveKarma: zod_1.z.number(),
    negativeKarma: zod_1.z.number(),
    positiveKarmaReasons: zod_1.z.array(zod_1.z.any()),
    negativeKarmaReasons: zod_1.z.array(zod_1.z.any()),
});
exports.ProductivityStatsSchema = zod_1.z.object({
    completedCount: zod_1.z.number(),
    daysItems: zod_1.z.array(exports.ItemsWithDateSchema.extend({
        date: zod_1.z.string(),
    })),
    goals: zod_1.z.object({
        currentDailyStreak: exports.StreakSchema,
        currentWeeklyStreak: exports.StreakSchema,
        dailyGoal: zod_1.z.number(),
        ignoreDays: zod_1.z.array(user_preferences_1.DayOfWeekSchema),
        karmaDisabled: user_preferences_1.BooleanFromZeroOneSchema,
        lastDailyStreak: exports.StreakSchema,
        lastWeeklyStreak: exports.StreakSchema,
        maxDailyStreak: exports.StreakSchema,
        maxWeeklyStreak: exports.StreakSchema,
        user: zod_1.z.string(),
        userId: zod_1.z.string(),
        vacationMode: user_preferences_1.BooleanFromZeroOneSchema,
        weeklyGoal: zod_1.z.number(),
    }),
    karma: zod_1.z.number(),
    karmaGraphData: zod_1.z.array(zod_1.z.object({
        date: zod_1.z.string(),
        karmaAvg: zod_1.z.number(),
    })),
    karmaLastUpdate: zod_1.z.number(),
    karmaTrend: zod_1.z.string(),
    karmaUpdateReasons: zod_1.z.array(exports.KarmaUpdateSchema),
    projectColors: zod_1.z.record(zod_1.z.string(), zod_1.z.string()),
    weekItems: zod_1.z.array(exports.ItemsWithDateSchema.extend({
        from: zod_1.z.string(),
        to: zod_1.z.string(),
    })),
});
exports.ColorSchema = zod_1.z.object({
    /** The key of the color (i.e. 'berry_red') */
    key: zod_1.z.string(),
    /** The display name of the color (i.e. 'Berry Red') */
    displayName: zod_1.z.string(),
    /** The hex value of the color (i.e. '#b8255f') */
    hexValue: zod_1.z.string(),
});
