// @doist/todoist-sdk/dist/cjs/types/workspaces/types.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceUserTaskSchema = exports.MemberActivityInfoSchema = exports.WorkspaceSchema = exports.WorkspacePropertiesSchema = exports.WorkspaceLimitsSchema = exports.WorkspacePlanSchema = exports.WORKSPACE_PLANS = exports.JoinWorkspaceResultSchema = exports.WorkspacePlanDetailsSchema = exports.WORKSPACE_PLAN_STATUSES = exports.WORKSPACE_CURRENT_PLANS = exports.FormattedPriceListingSchema = exports.PlanPriceSchema = exports.WorkspaceInvitationSchema = exports.WorkspaceUserSchema = exports.WorkspaceRoleSchema = exports.WORKSPACE_ROLES = exports.COLLABORATOR_ROLES = void 0;
const zod_1 = require("zod");
const common_1 = require("../common");
const types_1 = require("../tasks/types");
/** Available project collaborator roles. */
exports.COLLABORATOR_ROLES = [
    'CREATOR',
    'ADMIN',
    'READ_WRITE',
    'EDIT_ONLY',
    'COMPLETE_ONLY',
];
/**
 * Available workspace roles.
 */
exports.WORKSPACE_ROLES = ['ADMIN', 'MEMBER', 'GUEST'];
exports.WorkspaceRoleSchema = zod_1.z.enum(exports.WORKSPACE_ROLES);
exports.WorkspaceUserSchema = zod_1.z.object({
    userId: zod_1.z.string(),
    workspaceId: zod_1.z.string(),
    userEmail: zod_1.z.string(),
    fullName: zod_1.z.string(),
    timezone: zod_1.z.string(),
    role: exports.WorkspaceRoleSchema,
    imageId: zod_1.z.string().nullable(),
    isDeleted: zod_1.z.boolean().default(false),
});
exports.WorkspaceInvitationSchema = zod_1.z.object({
    id: zod_1.z.string().default('0'),
    inviterId: zod_1.z.string(),
    userEmail: zod_1.z.string(),
    workspaceId: zod_1.z.string(),
    role: exports.WorkspaceRoleSchema,
    isExistingUser: zod_1.z.boolean(),
});
exports.PlanPriceSchema = zod_1.z.object({
    currency: zod_1.z.string(),
    amount: zod_1.z.union([zod_1.z.number(), zod_1.z.string()]),
    interval: zod_1.z.string().optional(),
});
exports.FormattedPriceListingSchema = zod_1.z.object({
    currency: zod_1.z.string().optional(),
    amount: zod_1.z.number().optional(),
    interval: zod_1.z.string().optional(),
    formatted: zod_1.z.string().optional(),
});
/** Available workspace plan names. */
exports.WORKSPACE_CURRENT_PLANS = ['Business', 'Starter'];
/** Available workspace plan statuses. */
exports.WORKSPACE_PLAN_STATUSES = [
    'Active',
    'Downgraded',
    'Cancelled',
    'NeverSubscribed',
];
exports.WorkspacePlanDetailsSchema = zod_1.z.object({
    currentMemberCount: zod_1.z.number(),
    currentPlan: zod_1.z.enum(exports.WORKSPACE_CURRENT_PLANS),
    currentPlanStatus: zod_1.z.enum(exports.WORKSPACE_PLAN_STATUSES),
    downgradeAt: zod_1.z.string().nullable(),
    currentActiveProjects: zod_1.z.number(),
    maximumActiveProjects: zod_1.z.number(),
    priceList: zod_1.z.array(exports.FormattedPriceListingSchema),
    workspaceId: common_1.StringOrNumberSchema,
    isTrialing: zod_1.z.boolean(),
    trialEndsAt: zod_1.z.string().nullable(),
    cancelAtPeriodEnd: zod_1.z.boolean(),
    hasTrialed: zod_1.z.boolean(),
    planPrice: exports.PlanPriceSchema.nullable(),
    hasBillingPortal: zod_1.z.boolean(),
    hasBillingPortalSwitchToAnnual: zod_1.z.boolean(),
});
exports.JoinWorkspaceResultSchema = zod_1.z.object({
    customSortingApplied: zod_1.z.boolean(),
    projectSortPreference: zod_1.z.string(),
    role: exports.WorkspaceRoleSchema,
    userId: zod_1.z.string(),
    workspaceId: zod_1.z.string(),
});
/**
 * Available workspace plans.
 */
exports.WORKSPACE_PLANS = ['STARTER', 'BUSINESS'];
exports.WorkspacePlanSchema = zod_1.z.enum(exports.WORKSPACE_PLANS);
/**
 * Workspace resource limits.
 */
exports.WorkspaceLimitsSchema = zod_1.z
    .object({
    current: zod_1.z.record(zod_1.z.string(), zod_1.z.any()).nullable(),
    next: zod_1.z.record(zod_1.z.string(), zod_1.z.any()).nullable(),
})
    .catchall(zod_1.z.any());
/**
 * Workspace properties (flexible object for unknown fields).
 */
exports.WorkspacePropertiesSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.unknown());
/**
 * Represents a workspace in Todoist.
 */
exports.WorkspaceSchema = zod_1.z
    .object({
    id: common_1.StringOrNumberSchema,
    name: zod_1.z.string(),
    plan: exports.WorkspacePlanSchema,
    role: exports.WorkspaceRoleSchema.optional(),
    inviteCode: zod_1.z.string(),
    isLinkSharingEnabled: zod_1.z.boolean(),
    isGuestAllowed: zod_1.z.boolean(),
    limits: exports.WorkspaceLimitsSchema.optional(),
    logoBig: zod_1.z.string().nullish(),
    logoMedium: zod_1.z.string().nullish(),
    logoSmall: zod_1.z.string().nullish(),
    logoS640: zod_1.z.string().nullish(),
    createdAt: zod_1.z.coerce.date().optional(),
    dateCreated: zod_1.z.coerce.date().optional(),
    creatorId: common_1.StringOrNumberSchema,
    properties: exports.WorkspacePropertiesSchema,
})
    .transform((_a) => {
    var { dateCreated, createdAt } = _a, rest = __rest(_a, ["dateCreated", "createdAt"]);
    return (Object.assign(Object.assign({}, rest), { createdAt: createdAt !== null && createdAt !== void 0 ? createdAt : dateCreated }));
});
exports.MemberActivityInfoSchema = zod_1.z.object({
    userId: zod_1.z.string(),
    tasksAssigned: zod_1.z.number().int(),
    tasksOverdue: zod_1.z.number().int(),
});
exports.WorkspaceUserTaskSchema = zod_1.z.object({
    id: zod_1.z.string(),
    content: zod_1.z.string(),
    responsibleUid: zod_1.z.string().nullable(),
    due: types_1.DueDateSchema.nullable(),
    deadline: types_1.DeadlineSchema.nullable(),
    labels: zod_1.z.array(zod_1.z.string()),
    notesCount: zod_1.z.number().int(),
    projectId: zod_1.z.string(),
    projectName: zod_1.z.string(),
    priority: zod_1.z.number().int(),
    description: zod_1.z.string(),
    isOverdue: zod_1.z.boolean(),
});
