// @doist/todoist-sdk/dist/esm/clients/activity-client.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { ENDPOINT_REST_ACTIVITIES } from '../consts/endpoints.js';
import { request } from '../transport/http-client.js';
import { denormalizeObjectTypeFromApi, normalizeObjectEventTypeForApi, } from '../utils/activity-helpers.js';
import { formatDateToYYYYMMDD } from '../utils/url-helpers.js';
import { validateActivityEventArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling the activity-log endpoint.
 *
 * Instantiated by `TodoistApi`; the public `getActivityLogs` method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
export class ActivityClient extends BaseClient {
    async getActivityLogs(args = {}) {
        // Convert Date objects to YYYY-MM-DD strings
        const dateFrom = args.dateFrom instanceof Date ? formatDateToYYYYMMDD(args.dateFrom) : args.dateFrom;
        const dateTo = args.dateTo instanceof Date ? formatDateToYYYYMMDD(args.dateTo) : args.dateTo;
        // Destructure out raw date, filter-type, and removed legacy fields so they don't leak into payload
        const _a = args, { dateFrom: _dateFrom, dateTo: _dateTo, objectEventTypes, objectType: _objectType, eventType: _eventType, since: _since, until: _until } = _a, rest = __rest(_a, ["dateFrom", "dateTo", "objectEventTypes", "objectType", "eventType", "since", "until"]);
        // Build normalized objectEventTypes for the API
        let normalizedObjectEventTypes;
        if (objectEventTypes !== undefined) {
            const arr = Array.isArray(objectEventTypes) ? objectEventTypes : [objectEventTypes];
            normalizedObjectEventTypes = arr.map(normalizeObjectEventTypeForApi);
        }
        const processedArgs = Object.assign(Object.assign(Object.assign(Object.assign({}, rest), (dateFrom !== undefined ? { dateFrom } : {})), (dateTo !== undefined ? { dateTo } : {})), (normalizedObjectEventTypes !== undefined
            ? { objectEventTypes: normalizedObjectEventTypes }
            : {}));
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_ACTIVITIES,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
        });
        // Convert legacy API object types back to modern SDK types
        const normalizedResults = results.map((event) => {
            const normalizedType = denormalizeObjectTypeFromApi(event.objectType);
            return Object.assign(Object.assign({}, event), { objectType: normalizedType || event.objectType });
        });
        return {
            results: validateActivityEventArray(normalizedResults),
            nextCursor,
        };
    }
}
