// @doist/todoist-sdk/dist/esm/clients/comment-client.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { ENDPOINT_REST_COMMENTS } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { generatePath } from '../utils/request-helpers.js';
import { IdSchema, validateComment, validateCommentArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all comment-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public comment method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class CommentClient extends BaseClient {
    async getComments(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_COMMENTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateCommentArray(results),
            nextCursor,
        };
    }
    async getComment(id) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateComment(response.data);
    }
    async addComment(args, requestId) {
        const { uidsToNotify } = args, rest = __rest(args, ["uidsToNotify"]);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_COMMENTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, rest), (uidsToNotify ? { uidsToNotify: uidsToNotify.join(',') } : {})),
            requestId: requestId,
        });
        return validateComment(response.data);
    }
    async updateComment(id, args, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateComment(response.data);
    }
    async deleteComment(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
}
