// @doist/todoist-sdk/dist/esm/clients/folder-client.js
import { ENDPOINT_REST_FOLDERS } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { generatePath } from '../utils/request-helpers.js';
import { IdSchema, validateFolder, validateFolderArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all folder-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public folder method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class FolderClient extends BaseClient {
    async getFolders(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_FOLDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateFolderArray(results),
            nextCursor,
        };
    }
    async getFolder(id) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateFolder(response.data);
    }
    async addFolder(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_FOLDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateFolder(response.data);
    }
    async updateFolder(id, args, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateFolder(response.data);
    }
    async deleteFolder(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_FOLDERS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
}
