// @doist/todoist-sdk/dist/esm/clients/insights-client.js
import { getProjectInsightsActivityStatsEndpoint, getProjectInsightsHealthAnalyzeEndpoint, getProjectInsightsHealthContextEndpoint, getProjectInsightsHealthEndpoint, getProjectInsightsProgressEndpoint, getWorkspaceInsightsEndpoint, } from '../consts/endpoints.js';
import { request } from '../transport/http-client.js';
import { IdSchema, validateProjectActivityStats, validateProjectHealth, validateProjectHealthContext, validateProjectProgress, validateWorkspaceInsights, } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all insights-domain endpoints
 * (project activity stats, health, progress, workspace insights).
 *
 * Instantiated by `TodoistApi`; every public insights method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
export class InsightsClient extends BaseClient {
    async getProjectActivityStats(projectId, args = {}) {
        IdSchema.parse(projectId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: getProjectInsightsActivityStatsEndpoint(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign({ objectType: 'ITEM', eventType: 'COMPLETED' }, args),
        });
        return validateProjectActivityStats(response.data);
    }
    async getProjectHealth(projectId) {
        IdSchema.parse(projectId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: getProjectInsightsHealthEndpoint(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateProjectHealth(response.data);
    }
    async getProjectHealthContext(projectId) {
        IdSchema.parse(projectId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: getProjectInsightsHealthContextEndpoint(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateProjectHealthContext(response.data);
    }
    async getProjectProgress(projectId) {
        IdSchema.parse(projectId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: getProjectInsightsProgressEndpoint(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateProjectProgress(response.data);
    }
    async getWorkspaceInsights(workspaceId, args = {}) {
        IdSchema.parse(workspaceId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: getWorkspaceInsightsEndpoint(workspaceId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, args), (args.projectIds ? { projectIds: args.projectIds.join(',') } : {})),
        });
        return validateWorkspaceInsights(response.data);
    }
    async analyzeProjectHealth(projectId, requestId) {
        IdSchema.parse(projectId);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: getProjectInsightsHealthAnalyzeEndpoint(projectId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateProjectHealth(response.data);
    }
}
