// @doist/todoist-sdk/dist/esm/types/sync/resources/reminders.js
import { z } from 'zod';
import { DueDateSchema } from '../../tasks/types.js';
export const ReminderBaseSchema = z.looseObject({
    id: z.string(),
    notifyUid: z.string(),
    itemId: z.string(),
    projectId: z.string().optional(),
    isDeleted: z.boolean(),
});
/** Enum-like object of reminder type values. */
export const ReminderTypeEnum = {
    Relative: 'relative',
    Absolute: 'absolute',
    Location: 'location',
};
/** Available reminder types. */
export const REMINDER_TYPES = [
    ReminderTypeEnum.Relative,
    ReminderTypeEnum.Absolute,
    ReminderTypeEnum.Location,
];
/** Available location reminder triggers. */
export const LOCATION_TRIGGERS = ['on_enter', 'on_leave'];
export const LocationReminderSchema = ReminderBaseSchema.extend({
    type: z.literal(ReminderTypeEnum.Location),
    name: z.string(),
    locLat: z.string(),
    locLong: z.string(),
    locTrigger: z.enum(LOCATION_TRIGGERS),
    radius: z.number().int(),
});
export const AbsoluteReminderSchema = ReminderBaseSchema.extend({
    type: z.literal(ReminderTypeEnum.Absolute),
    due: DueDateSchema,
    isUrgent: z.boolean().optional(),
});
export const RelativeReminderSchema = ReminderBaseSchema.extend({
    type: z.literal(ReminderTypeEnum.Relative),
    minuteOffset: z.number().int(),
    due: DueDateSchema.optional(),
    isUrgent: z.boolean().optional(),
});
export const ReminderSchema = z.discriminatedUnion('type', [
    LocationReminderSchema,
    AbsoluteReminderSchema,
    RelativeReminderSchema,
]);
