// @doist/todoist-sdk/dist/esm/types/sync/resources/workspaces.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
import { StringOrNumberSchema } from '../../common.js';
import { WorkspaceRoleSchema, WorkspacePlanSchema, WorkspaceLimitsSchema, WorkspacePropertiesSchema, } from '../../workspaces/types.js';
/**
 * Sync API workspace resource.
 *
 * This is a superset of the REST `WorkspaceSchema` — the Sync API returns
 * additional fields like description, isDeleted, isCollapsed, member counts,
 * domain settings, and sorting preferences.
 */
export const SyncWorkspaceSchema = z
    .looseObject({
    id: StringOrNumberSchema,
    name: z.string(),
    description: z.string(),
    logoBig: z.string().nullish(),
    logoMedium: z.string().nullish(),
    logoSmall: z.string().nullish(),
    logoS640: z.string().nullish(),
    creatorId: StringOrNumberSchema,
    createdAt: z.coerce.date().optional(),
    dateCreated: z.coerce.date().optional(),
    isDeleted: z.boolean(),
    isCollapsed: z.boolean(),
    role: WorkspaceRoleSchema.optional(),
    plan: WorkspacePlanSchema,
    limits: WorkspaceLimitsSchema.optional(),
    inviteCode: z.string().nullable().optional(),
    isLinkSharingEnabled: z.boolean().nullable().optional(),
    isGuestAllowed: z.boolean().nullable().optional(),
    currentActiveProjects: z.number().nullable(),
    currentMemberCount: z.number().nullable(),
    memberCountByType: z
        .object({
        adminCount: z.number().int(),
        guestCount: z.number().int(),
        memberCount: z.number().int(),
    })
        .optional(),
    currentTemplateCount: z.number().nullable(),
    pendingInvitations: z.array(z.string()).nullable().optional(),
    domainName: z.string().nullish(),
    domainDiscovery: z.boolean().nullish(),
    restrictEmailDomains: z.boolean().nullish(),
    adminSortingApplied: z.boolean(),
    projectSortPreference: z.string().nullish(),
    defaultCollaborators: z
        .object({
        predefinedGroupIds: z.array(z.string()),
        userIds: z.array(StringOrNumberSchema),
    })
        .optional(),
    properties: WorkspacePropertiesSchema.optional(),
})
    .transform((_a) => {
    var { dateCreated, createdAt } = _a, rest = __rest(_a, ["dateCreated", "createdAt"]);
    return (Object.assign(Object.assign({}, rest), { createdAt: createdAt !== null && createdAt !== void 0 ? createdAt : dateCreated }));
});
