// @doist/todoist-sdk/dist/esm/types/goals/types.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
/** Available goal owner types. */
export const GOAL_OWNER_TYPES = ['USER', 'WORKSPACE'];
/**
 * Progress of a {@link Goal}.
 *
 * The REST API returns these counts under `total_item_count` / `completed_item_count`,
 * reflecting the backend's generic "item" terminology. The SDK normalizes them to
 * `totalTaskCount` / `completedTaskCount` so the public surface consistently speaks
 * in "tasks" (matching the rest of the SDK). The `preprocess` step below performs
 * that rename after the transport layer's snake_case → camelCase conversion — do not
 * remove it without also changing the downstream field names.
 */
export const GoalProgressSchema = z.preprocess((raw) => {
    if (raw === null || raw === undefined || typeof raw !== 'object' || Array.isArray(raw)) {
        return raw;
    }
    const _a = raw, { totalItemCount, completedItemCount, totalTaskCount, completedTaskCount } = _a, rest = __rest(_a, ["totalItemCount", "completedItemCount", "totalTaskCount", "completedTaskCount"]);
    return Object.assign(Object.assign({}, rest), { totalTaskCount: totalTaskCount !== null && totalTaskCount !== void 0 ? totalTaskCount : totalItemCount, completedTaskCount: completedTaskCount !== null && completedTaskCount !== void 0 ? completedTaskCount : completedItemCount });
}, z.looseObject({
    totalTaskCount: z.number().int(),
    completedTaskCount: z.number().int(),
    percentage: z.number(),
}));
export const GoalSchema = z.looseObject({
    id: z.string(),
    ownerType: z.enum(GOAL_OWNER_TYPES),
    ownerId: z.string(),
    name: z.string(),
    description: z.string().nullable(),
    deadline: z.string().nullable(),
    parentGoalId: z.string().nullable(),
    childOrder: z.number().int(),
    isCompleted: z.boolean(),
    completedAt: z.coerce.date().nullable(),
    responsibleUid: z.string().nullable(),
    isDeleted: z.boolean(),
    progress: GoalProgressSchema.optional(),
    creatorUid: z.string(),
    createdAt: z.coerce.date(),
    updatedAt: z.coerce.date(),
});
