// @doist/todoist-sdk/dist/esm/clients/app-client.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { z } from 'zod';
import { ENDPOINT_REST_APPS, ENDPOINT_REST_APPS_INSTALLATIONS, ENDPOINT_REST_USER_AUTHORIZATIONS, ENDPOINT_REST_USER_AUTHORIZATIONS_DELETE, getAppByDistributionTokenEndpoint, getAppClientSecretEndpoint, getAppDistributionTokenEndpoint, getAppEndpoint, getAppIconEndpoint, getAppInstallationEndpoint, getAppSecretsEndpoint, getAppTestTokenEndpoint, getAppTokensEndpoint, getAppVerificationTokenEndpoint, getAppWebhookEndpoint, } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { webhookEventToWireName } from '../types/apps/index.js';
import { uploadMultipartFile } from '../utils/multipart-upload.js';
import { IdSchema, validateApp, validateAppArray, validateAppByDistributionToken, validateAppDistributionToken, validateAppInstallation, validateAppInstallationArray, validateAppSecrets, validateAppTestToken, validateAppVerificationToken, validateAppWebhook, validateAppWithUserCount, validateUserAuthorizationArray, } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all developer-app management endpoints
 * (apps, secrets, tokens, webhooks, installations, user authorizations).
 * All endpoints require the `dev:app_console` OAuth scope.
 *
 * Instantiated by `TodoistApi`; every public app-management method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing JSDoc.
 */
export class AppClient extends BaseClient {
    async getApps(requestId) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppArray(response.data);
    }
    async getApp(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppWithUserCount(response.data);
    }
    async addApp(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateApp(response.data);
    }
    async updateApp(appId, args, requestId) {
        IdSchema.parse(appId);
        const { appTokenScopes } = args, rest = __rest(args, ["appTokenScopes"]);
        const payload = Object.assign({}, rest);
        if (appTokenScopes !== undefined) {
            payload.appTokenScopes = appTokenScopes === null ? null : appTokenScopes.join(',');
        }
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: getAppEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: payload,
            requestId: requestId,
        });
        return validateAppWithUserCount(response.data);
    }
    async deleteApp(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async getAppSecrets(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppSecretsEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppSecrets(response.data);
    }
    async resetAppClientSecret(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppClientSecretEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppWithUserCount(response.data);
    }
    async revokeAppTokens(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppTokensEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppWithUserCount(response.data);
    }
    async uploadAppIcon({ appId, file, fileName = '', size = 'medium' }, requestId) {
        IdSchema.parse(appId);
        if (Buffer.isBuffer(file) && file.length === 0) {
            throw new Error('Cannot upload empty image file');
        }
        const data = await uploadMultipartFile({
            baseUrl: this.apiRootBase,
            authToken: this.authToken,
            endpoint: getAppIconEndpoint(appId, size),
            file,
            fileName,
            additionalFields: { file_name: fileName },
            requestId: requestId,
            customFetch: this.customFetch,
        });
        return validateAppWithUserCount(data);
    }
    async getAppTestToken(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppTestTokenEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppTestToken(response.data);
    }
    async createAppTestToken(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'PUT',
            baseUri: this.apiRootBase,
            relativePath: getAppTestTokenEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppTestToken(response.data);
    }
    async getAppDistributionToken(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppDistributionTokenEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppDistributionToken(response.data);
    }
    async getAppVerificationToken(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppVerificationTokenEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppVerificationToken(response.data);
    }
    async resetAppVerificationToken(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppVerificationTokenEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppVerificationToken(response.data);
    }
    async getAppByDistributionToken(distributionToken, requestId) {
        z.string().min(1).parse(distributionToken);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppByDistributionTokenEndpoint(distributionToken),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppByDistributionToken(response.data);
    }
    // --- App webhooks ---
    async getAppWebhook(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppWebhookEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return response.data ? validateAppWebhook(response.data) : null;
    }
    async updateAppWebhook(args, requestId) {
        const { appId, events } = args, rest = __rest(args, ["appId", "events"]);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: getAppWebhookEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, rest), { events: events.map(webhookEventToWireName) }),
            requestId: requestId,
        });
        return validateAppWebhook(response.data);
    }
    async deleteAppWebhook(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppWebhookEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    // --- App installations ---
    async getAppInstallations(requestId) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS_INSTALLATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppInstallationArray(response.data);
    }
    async installApp(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS_INSTALLATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateAppInstallation(response.data);
    }
    async getAppInstallation(installationId, requestId) {
        IdSchema.parse(installationId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getAppInstallationEndpoint(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateAppInstallation(response.data);
    }
    async updateAppInstallation(installationId, args, requestId) {
        IdSchema.parse(installationId);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: getAppInstallationEndpoint(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateAppInstallation(response.data);
    }
    async uninstallApp(installationId, requestId) {
        IdSchema.parse(installationId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getAppInstallationEndpoint(installationId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    // --- User authorizations ---
    async getUserAuthorizations(requestId) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_USER_AUTHORIZATIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateUserAuthorizationArray(response.data);
    }
    async revokeUserAuthorization(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_USER_AUTHORIZATIONS_DELETE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return isSuccess(response);
    }
}
