// @doist/todoist-sdk/dist/esm/clients/billing-client.js
import { ENDPOINT_PAYMENTS_CANCEL_PLAN, ENDPOINT_PAYMENTS_REACTIVATE_PLAN, ENDPOINT_PAYMENTS_SUBSCRIPTION_INFO, ENDPOINT_PRICES, ENDPOINT_PRICING, ENDPOINT_PRO_BILLING_PORTAL, ENDPOINT_PRO_PLAN_DETAILS, ENDPOINT_PRO_TRIAL, ENDPOINT_PRO_UPGRADE, ENDPOINT_WORKSPACE_BILLING_PORTAL, ENDPOINT_WORKSPACE_TRIAL, ENDPOINT_WORKSPACE_UPGRADE, } from '../consts/endpoints.js';
import { request } from '../transport/http-client.js';
import { validatePricing, validatePrices, validateProPlanDetails, validateSubscriptionInfo, } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all billing-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public billing method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class BillingClient extends BaseClient {
    async getSubscriptionInfo() {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PAYMENTS_SUBSCRIPTION_INFO,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateSubscriptionInfo(response.data);
    }
    async cancelPlan(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PAYMENTS_CANCEL_PLAN,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async reactivatePlan(requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PAYMENTS_REACTIVATE_PLAN,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateSubscriptionInfo(response.data);
    }
    async upgradeToPro(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRO_UPGRADE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async startProTrial(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRO_TRIAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async createProBillingPortalSession(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRO_BILLING_PORTAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async getProPlanDetails() {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRO_PLAN_DETAILS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateProPlanDetails(response.data);
    }
    async upgradeWorkspace(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_WORKSPACE_UPGRADE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async startWorkspaceTrial(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_WORKSPACE_TRIAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async createWorkspaceBillingPortalSession(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_WORKSPACE_BILLING_PORTAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async getPrices() {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRICES,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validatePrices(response.data);
    }
    async getPricing(args) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_PRICING,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return validatePricing(response.data);
    }
}
