// @doist/todoist-sdk/dist/esm/clients/goals-client.js
import { ENDPOINT_REST_GOALS, ENDPOINT_REST_GOALS_SEARCH, GOAL_COMPLETE, GOAL_TASKS, GOAL_UNCOMPLETE, } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { generatePath } from '../utils/request-helpers.js';
import { IdSchema, validateGoal, validateGoalArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all goal-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public goal method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class GoalsClient extends BaseClient {
    async getGoals(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_GOALS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateGoalArray(results),
            nextCursor,
        };
    }
    async searchGoals(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_GOALS_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateGoalArray(results),
            nextCursor,
        };
    }
    async getGoal(id) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateGoal(response.data);
    }
    async addGoal(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_GOALS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateGoal(response.data);
    }
    async updateGoal(id, args, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateGoal(response.data);
    }
    async deleteGoal(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async completeGoal(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, id, GOAL_COMPLETE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateGoal(response.data);
    }
    async uncompleteGoal(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, id, GOAL_UNCOMPLETE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateGoal(response.data);
    }
    async linkTaskToGoal({ goalId, taskId }, requestId) {
        IdSchema.parse(goalId);
        IdSchema.parse(taskId);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, goalId, GOAL_TASKS),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: { itemId: taskId },
            requestId: requestId,
        });
        return validateGoal(response.data);
    }
    async unlinkTaskFromGoal({ goalId, taskId }, requestId) {
        IdSchema.parse(goalId);
        IdSchema.parse(taskId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_GOALS, goalId, GOAL_TASKS, taskId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
}
