// @doist/todoist-sdk/dist/esm/clients/id-mapping-client.js
import { ENDPOINT_REST_ID_MAPPINGS, ENDPOINT_REST_MOVED_IDS } from '../consts/endpoints.js';
import { request } from '../transport/http-client.js';
import { generatePath } from '../utils/request-helpers.js';
import { validateIdMappingArray, validateMovedIdArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling ID-mapping endpoints.
 *
 * Instantiated by `TodoistApi`; every public id-mapping method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
export class IdMappingClient extends BaseClient {
    async getIdMappings(args) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_ID_MAPPINGS, args.objName, args.objIds.join(',')),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateIdMappingArray(response.data);
    }
    async getMovedIds(args) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_MOVED_IDS, args.objName),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args.oldIds ? { oldIds: args.oldIds.join(',') } : undefined,
        });
        return validateMovedIdArray(response.data);
    }
}
