// @doist/todoist-sdk/dist/esm/clients/label-client.js
import { ENDPOINT_REST_LABELS, ENDPOINT_REST_LABELS_SEARCH, ENDPOINT_REST_LABELS_SHARED, ENDPOINT_REST_LABELS_SHARED_REMOVE, ENDPOINT_REST_LABELS_SHARED_RENAME, } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { generatePath } from '../utils/request-helpers.js';
import { IdSchema, validateLabel, validateLabelArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all label-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public label method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class LabelClient extends BaseClient {
    async getLabel(id) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateLabel(response.data);
    }
    async getLabels(args = {}) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateLabelArray(results),
            nextCursor,
        };
    }
    async searchLabels(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateLabelArray(results),
            nextCursor,
        };
    }
    async addLabel(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateLabel(response.data);
    }
    async updateLabel(id, args, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateLabel(response.data);
    }
    async deleteLabel(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async getSharedLabels(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS_SHARED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return { results, nextCursor };
    }
    async renameSharedLabel(args) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS_SHARED_RENAME,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return isSuccess(response);
    }
    async removeSharedLabel(args) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_LABELS_SHARED_REMOVE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return isSuccess(response);
    }
}
