// @doist/todoist-sdk/dist/esm/clients/task-client.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { v4 as uuidv4 } from 'uuid';
import { ENDPOINT_REST_TASKS, ENDPOINT_REST_TASKS_COMPLETED, ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE, ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE, ENDPOINT_REST_TASKS_COMPLETED_SEARCH, ENDPOINT_REST_TASKS_FILTER, ENDPOINT_REST_TASK_CLOSE, ENDPOINT_REST_TASK_MOVE, ENDPOINT_REST_TASK_REOPEN, ENDPOINT_SYNC_QUICK_ADD, } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { performSyncRequest } from '../transport/sync-request.js';
import { TodoistRequestError } from '../types/index.js';
import { generatePath, spreadIfDefined } from '../utils/request-helpers.js';
import { processTaskContent } from '../utils/uncompletable-helpers.js';
import { IdSchema, validateTask, validateTaskArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
const MAX_COMMAND_COUNT = 100;
/**
 * Internal sub-client handling all task-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public task method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
export class TaskClient extends BaseClient {
    async getTask(id) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return validateTask(response.data);
    }
    async getTasks(args = {}) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateTaskArray(results),
            nextCursor,
        };
    }
    async getTasksByFilter(args) {
        const { data: { results, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS_FILTER,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: validateTaskArray(results),
            nextCursor,
        };
    }
    async getCompletedTasksByCompletionDate(args) {
        const { data: { items, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS_COMPLETED_BY_COMPLETION_DATE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: validateTaskArray(items),
            nextCursor,
        };
    }
    async getCompletedTasksByDueDate(args) {
        const { data: { items, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS_COMPLETED_BY_DUE_DATE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: validateTaskArray(items),
            nextCursor,
        };
    }
    async searchCompletedTasks(args) {
        const { data: { items, nextCursor }, } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS_COMPLETED_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            items: validateTaskArray(items),
            nextCursor,
        };
    }
    async getAllCompletedTasks(args = {}) {
        const { since, until } = args, rest = __rest(args, ["since", "until"]);
        const { data } = await request({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS_COMPLETED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign(Object.assign({}, rest), (since ? { since: since.toISOString() } : {})), (until ? { until: until.toISOString() } : {})),
        });
        return {
            projects: data.projects,
            sections: data.sections,
            items: validateTaskArray(data.items),
        };
    }
    async addTask(args, requestId) {
        // Process content based on isUncompletable flag
        const processedArgs = Object.assign(Object.assign({}, args), { content: processTaskContent(args.content, args.isUncompletable) });
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_REST_TASKS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
            requestId: requestId,
        });
        return validateTask(response.data);
    }
    async quickAddTask(args) {
        // Process text based on isUncompletable flag
        const processedArgs = Object.assign(Object.assign({}, args), { text: processTaskContent(args.text, args.isUncompletable) });
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: ENDPOINT_SYNC_QUICK_ADD,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
        });
        return validateTask(response.data);
    }
    async updateTask(id, args, requestId) {
        IdSchema.parse(id);
        // Translate SDK alias for due-date clearing to Todoist's accepted payload value.
        const normalizedArgs = args.dueString === null ? Object.assign(Object.assign({}, args), { dueString: 'no date' }) : args;
        // Process content if both content and isUncompletable are provided
        const processedArgs = normalizedArgs.content && normalizedArgs.isUncompletable !== undefined
            ? Object.assign(Object.assign({}, normalizedArgs), { content: processTaskContent(normalizedArgs.content, normalizedArgs.isUncompletable) }) : normalizedArgs;
        // Remap `order` → `childOrder` so snakeCaseKeys() produces `child_order`
        const { order } = processedArgs, argsWithoutOrder = __rest(processedArgs, ["order"]);
        const remappedArgs = order !== undefined ? Object.assign(Object.assign({}, argsWithoutOrder), { childOrder: order }) : argsWithoutOrder;
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: remappedArgs,
            requestId: requestId,
        });
        return validateTask(response.data);
    }
    async moveTasks(ids, args, requestId) {
        var _a;
        if (ids.length > MAX_COMMAND_COUNT) {
            throw new TodoistRequestError(`Maximum number of items is ${MAX_COMMAND_COUNT}`, 400);
        }
        const commands = ids.map((id) => ({
            type: 'item_move',
            uuid: uuidv4(),
            args: Object.assign(Object.assign(Object.assign({ id }, spreadIfDefined(args.projectId, (v) => ({ projectId: v }))), spreadIfDefined(args.sectionId, (v) => ({ sectionId: v }))), spreadIfDefined(args.parentId, (v) => ({ parentId: v }))),
        }));
        const syncRequest = {
            commands,
            resourceTypes: ['items'],
        };
        const syncResponse = await performSyncRequest(this.syncContext, syncRequest, {
            requestId,
            hasSyncCommands: true,
        });
        if (!((_a = syncResponse.items) === null || _a === void 0 ? void 0 : _a.length)) {
            throw new TodoistRequestError('Tasks not found', 404);
        }
        const syncTasks = syncResponse.items.filter((task) => ids.includes(task.id));
        if (!syncTasks.length) {
            throw new TodoistRequestError('Tasks not found', 404);
        }
        return validateTaskArray(syncTasks);
    }
    async moveTask(id, args, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id, ENDPOINT_REST_TASK_MOVE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign(Object.assign({}, spreadIfDefined(args.projectId, (v) => ({ project_id: v }))), spreadIfDefined(args.sectionId, (v) => ({ section_id: v }))), spreadIfDefined(args.parentId, (v) => ({ parent_id: v }))),
            requestId: requestId,
        });
        return validateTask(response.data);
    }
    async closeTask(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id, ENDPOINT_REST_TASK_CLOSE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async reopenTask(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id, ENDPOINT_REST_TASK_REOPEN),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async deleteTask(id, requestId) {
        IdSchema.parse(id);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: generatePath(ENDPOINT_REST_TASKS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
}
