// @doist/todoist-sdk/dist/esm/clients/ui-extension-client.js
import { ENDPOINT_REST_APPS_UI_EXTENSIONS, ENDPOINT_REST_APPS_UI_EXTENSIONS_INSTALLED, getUiExtensionEndpoint, getUiExtensionIconEndpoint, getUiExtensionsByIntegrationEndpoint, } from '../consts/endpoints.js';
import { isSuccess, request } from '../transport/http-client.js';
import { uploadMultipartFile } from '../utils/multipart-upload.js';
import { IdSchema, validateUiExtension, validateUiExtensionArray } from '../utils/validators.js';
import { BaseClient } from './base-client.js';
/**
 * Internal sub-client handling all UI-extension endpoints for developer
 * applications. All endpoints require the `dev:app_console` OAuth scope.
 *
 * Instantiated by `TodoistApi`; every public UI-extension method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing JSDoc.
 */
export class UiExtensionClient extends BaseClient {
    async getUiExtension(uiExtensionId, requestId) {
        IdSchema.parse(uiExtensionId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getUiExtensionEndpoint(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateUiExtension(response.data);
    }
    async getInstalledUiExtensions(requestId) {
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS_UI_EXTENSIONS_INSTALLED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateUiExtensionArray(response.data);
    }
    async getUiExtensionsForApp(appId, requestId) {
        IdSchema.parse(appId);
        const response = await request({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: getUiExtensionsByIntegrationEndpoint(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return validateUiExtensionArray(response.data);
    }
    async addUiExtension(args, requestId) {
        const response = await request({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: ENDPOINT_REST_APPS_UI_EXTENSIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateUiExtension(response.data);
    }
    async updateUiExtension(uiExtensionId, args, requestId) {
        IdSchema.parse(uiExtensionId);
        const response = await request({
            httpMethod: 'PATCH',
            baseUri: this.apiRootBase,
            relativePath: getUiExtensionEndpoint(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return validateUiExtension(response.data);
    }
    async deleteUiExtension(uiExtensionId, requestId) {
        IdSchema.parse(uiExtensionId);
        const response = await request({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: getUiExtensionEndpoint(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return isSuccess(response);
    }
    async uploadUiExtensionIcon(args, requestId) {
        var _a;
        if (Buffer.isBuffer(args.file) && args.file.length === 0) {
            throw new Error('Cannot upload empty image file');
        }
        const data = await uploadMultipartFile({
            baseUrl: this.apiRootBase,
            authToken: this.authToken,
            endpoint: getUiExtensionIconEndpoint(args.uiExtensionId),
            file: args.file,
            fileName: args.fileName,
            additionalFields: { file_name: (_a = args.fileName) !== null && _a !== void 0 ? _a : '' },
            requestId: requestId,
            customFetch: this.customFetch,
        });
        return validateUiExtension(data);
    }
}
