// @doist/todoist-sdk/dist/cjs/clients/activity-client.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const activity_helpers_1 = require("../utils/activity-helpers");
const url_helpers_1 = require("../utils/url-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling the activity-log endpoint.
 *
 * Instantiated by `TodoistApi`; the public `getActivityLogs` method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class ActivityClient extends base_client_1.BaseClient {
    async getActivityLogs(args = {}) {
        // Convert Date objects to YYYY-MM-DD strings
        const dateFrom = args.dateFrom instanceof Date ? (0, url_helpers_1.formatDateToYYYYMMDD)(args.dateFrom) : args.dateFrom;
        const dateTo = args.dateTo instanceof Date ? (0, url_helpers_1.formatDateToYYYYMMDD)(args.dateTo) : args.dateTo;
        // Destructure out raw date, filter-type, and removed legacy fields so they don't leak into payload
        const _a = args, { dateFrom: _dateFrom, dateTo: _dateTo, objectEventTypes, objectType: _objectType, eventType: _eventType, since: _since, until: _until } = _a, rest = __rest(_a, ["dateFrom", "dateTo", "objectEventTypes", "objectType", "eventType", "since", "until"]);
        // Build normalized objectEventTypes for the API
        let normalizedObjectEventTypes;
        if (objectEventTypes !== undefined) {
            const arr = Array.isArray(objectEventTypes) ? objectEventTypes : [objectEventTypes];
            normalizedObjectEventTypes = arr.map(activity_helpers_1.normalizeObjectEventTypeForApi);
        }
        const processedArgs = Object.assign(Object.assign(Object.assign(Object.assign({}, rest), (dateFrom !== undefined ? { dateFrom } : {})), (dateTo !== undefined ? { dateTo } : {})), (normalizedObjectEventTypes !== undefined
            ? { objectEventTypes: normalizedObjectEventTypes }
            : {}));
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_ACTIVITIES,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: processedArgs,
        });
        // Convert legacy API object types back to modern SDK types
        const normalizedResults = results.map((event) => {
            const normalizedType = (0, activity_helpers_1.denormalizeObjectTypeFromApi)(event.objectType);
            return Object.assign(Object.assign({}, event), { objectType: normalizedType || event.objectType });
        });
        return {
            results: (0, validators_1.validateActivityEventArray)(normalizedResults),
            nextCursor,
        };
    }
}
exports.ActivityClient = ActivityClient;
