// @doist/todoist-sdk/dist/cjs/clients/backup-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const file_response_1 = require("../utils/file-response");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling backup-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public backup method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class BackupClient extends base_client_1.BaseClient {
    async getBackups(args = {}) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_BACKUPS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return (0, validators_1.validateBackupArray)(response.data);
    }
    async downloadBackup(args) {
        const url = `${this.syncApiBase}${endpoints_1.ENDPOINT_REST_BACKUPS_DOWNLOAD}?file=${encodeURIComponent(args.file)}`;
        const fetchOptions = {
            headers: { Authorization: `Bearer ${this.authToken}` },
        };
        if (this.customFetch) {
            const response = await this.customFetch(url, fetchOptions);
            if (!response.ok) {
                throw new Error(`Failed to download backup: ${response.status} ${response.statusText}`);
            }
            return (0, file_response_1.wrapAsFileResponse)(response, 'downloadBackup');
        }
        const response = await fetch(url, fetchOptions);
        if (!response.ok) {
            throw new Error(`Failed to download backup: ${response.status} ${response.statusText}`);
        }
        return (0, file_response_1.wrapAsFileResponse)(response, 'downloadBackup');
    }
}
exports.BackupClient = BackupClient;
