// @doist/todoist-sdk/dist/cjs/clients/base-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseClient = void 0;
/**
 * Internal base class for every domain sub-client (tasks, projects, …).
 *
 * Not exported from the package — sub-clients are an implementation detail
 * of `TodoistApi`. Holds the three pieces of HTTP identity every sub-client
 * needs and exposes a typed accessor to feed the Sync API helper.
 *
 * The constructor takes a {@link SyncRequestContext} — the same record
 * shape used by `performSyncRequest` — so there is a single type describing
 * "the HTTP identity an SDK component needs", not two types that must stay
 * in sync.
 */
class BaseClient {
    constructor(context) {
        this.authToken = context.authToken;
        this.syncApiBase = context.syncApiBase;
        this.apiRootBase = new URL('/', context.syncApiBase).toString();
        this.customFetch = context.customFetch;
    }
    get syncContext() {
        return {
            authToken: this.authToken,
            syncApiBase: this.syncApiBase,
            customFetch: this.customFetch,
        };
    }
}
exports.BaseClient = BaseClient;
