// @doist/todoist-sdk/dist/cjs/clients/billing-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BillingClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all billing-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public billing method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class BillingClient extends base_client_1.BaseClient {
    async getSubscriptionInfo() {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PAYMENTS_SUBSCRIPTION_INFO,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateSubscriptionInfo)(response.data);
    }
    async cancelPlan(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PAYMENTS_CANCEL_PLAN,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async reactivatePlan(requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PAYMENTS_REACTIVATE_PLAN,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateSubscriptionInfo)(response.data);
    }
    async upgradeToPro(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRO_UPGRADE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async startProTrial(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRO_TRIAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async createProBillingPortalSession(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRO_BILLING_PORTAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async getProPlanDetails() {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRO_PLAN_DETAILS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateProPlanDetails)(response.data);
    }
    async upgradeWorkspace(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_WORKSPACE_UPGRADE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async startWorkspaceTrial(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_WORKSPACE_TRIAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async createWorkspaceBillingPortalSession(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_WORKSPACE_BILLING_PORTAL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return response.data;
    }
    async getPrices() {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRICES,
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validatePrices)(response.data);
    }
    async getPricing(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_PRICING,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return (0, validators_1.validatePricing)(response.data);
    }
}
exports.BillingClient = BillingClient;
