// @doist/todoist-sdk/dist/cjs/clients/comment-client.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommentClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all comment-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public comment method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class CommentClient extends base_client_1.BaseClient {
    async getComments(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_COMMENTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateCommentArray)(results),
            nextCursor,
        };
    }
    async getComment(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateComment)(response.data);
    }
    async addComment(args, requestId) {
        const { uidsToNotify } = args, rest = __rest(args, ["uidsToNotify"]);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_COMMENTS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, rest), (uidsToNotify ? { uidsToNotify: uidsToNotify.join(',') } : {})),
            requestId: requestId,
        });
        return (0, validators_1.validateComment)(response.data);
    }
    async updateComment(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateComment)(response.data);
    }
    async deleteComment(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_COMMENTS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.CommentClient = CommentClient;
