// @doist/todoist-sdk/dist/cjs/clients/email-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling email-forwarding endpoints.
 *
 * Instantiated by `TodoistApi`; every public email-forwarding method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class EmailClient extends base_client_1.BaseClient {
    async getOrCreateEmailForwarding(args, requestId) {
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'PUT',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_EMAILS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return data;
    }
    async disableEmailForwarding(args, requestId) {
        const queryParams = new URLSearchParams({
            obj_type: args.objType,
            obj_id: args.objId,
        });
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: `${endpoints_1.ENDPOINT_REST_EMAILS}?${queryParams.toString()}`,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.EmailClient = EmailClient;
