// @doist/todoist-sdk/dist/cjs/clients/goals-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoalsClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all goal-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public goal method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class GoalsClient extends base_client_1.BaseClient {
    async getGoals(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_GOALS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateGoalArray)(results),
            nextCursor,
        };
    }
    async searchGoals(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_GOALS_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateGoalArray)(results),
            nextCursor,
        };
    }
    async getGoal(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async addGoal(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_GOALS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async updateGoal(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async deleteGoal(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async completeGoal(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, id, endpoints_1.GOAL_COMPLETE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async uncompleteGoal(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, id, endpoints_1.GOAL_UNCOMPLETE),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async linkTaskToGoal({ goalId, taskId }, requestId) {
        validators_1.IdSchema.parse(goalId);
        validators_1.IdSchema.parse(taskId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, goalId, endpoints_1.GOAL_TASKS),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: { itemId: taskId },
            requestId: requestId,
        });
        return (0, validators_1.validateGoal)(response.data);
    }
    async unlinkTaskFromGoal({ goalId, taskId }, requestId) {
        validators_1.IdSchema.parse(goalId);
        validators_1.IdSchema.parse(taskId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_GOALS, goalId, endpoints_1.GOAL_TASKS, taskId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.GoalsClient = GoalsClient;
