// @doist/todoist-sdk/dist/cjs/clients/id-mapping-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdMappingClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling ID-mapping endpoints.
 *
 * Instantiated by `TodoistApi`; every public id-mapping method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class IdMappingClient extends base_client_1.BaseClient {
    async getIdMappings(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_ID_MAPPINGS, args.objName, args.objIds.join(',')),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateIdMappingArray)(response.data);
    }
    async getMovedIds(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_MOVED_IDS, args.objName),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args.oldIds ? { oldIds: args.oldIds.join(',') } : undefined,
        });
        return (0, validators_1.validateMovedIdArray)(response.data);
    }
}
exports.IdMappingClient = IdMappingClient;
