// @doist/todoist-sdk/dist/cjs/clients/label-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabelClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all label-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public label method on `TodoistApi`
 * delegates here. See `todoist-api.ts` for the user-facing JSDoc.
 */
class LabelClient extends base_client_1.BaseClient {
    async getLabel(id) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateLabel)(response.data);
    }
    async getLabels(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateLabelArray)(results),
            nextCursor,
        };
    }
    async searchLabels(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS_SEARCH,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateLabelArray)(results),
            nextCursor,
        };
    }
    async addLabel(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateLabel)(response.data);
    }
    async updateLabel(id, args, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateLabel)(response.data);
    }
    async deleteLabel(id, requestId) {
        validators_1.IdSchema.parse(id);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.syncApiBase,
            relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LABELS, id),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async getSharedLabels(args) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS_SHARED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return { results, nextCursor };
    }
    async renameSharedLabel(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS_SHARED_RENAME,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async removeSharedLabel(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LABELS_SHARED_REMOVE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return (0, http_client_1.isSuccess)(response);
    }
}
exports.LabelClient = LabelClient;
