// @doist/todoist-sdk/dist/cjs/clients/reminder-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReminderClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const types_1 = require("../types");
const reminders_1 = require("../types/reminders");
const reminders_2 = require("../types/sync/resources/reminders");
const request_helpers_1 = require("../utils/request-helpers");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all reminder-domain endpoints
 * (both time-based and location-based reminders).
 *
 * Instantiated by `TodoistApi`; every public reminder method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc. The time-vs-location 404 translation is preserved here so a
 * caller who used the wrong endpoint gets a TodoistArgumentError
 * pointing at the correct method, rather than a bare 404.
 */
class ReminderClient extends base_client_1.BaseClient {
    async getReminders(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_REMINDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateReminderArray)(results),
            nextCursor,
        };
    }
    async getLocationReminders(args = {}) {
        const { data: { results, nextCursor }, } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LOCATION_REMINDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return {
            results: (0, validators_1.validateLocationReminderArray)(results),
            nextCursor,
        };
    }
    async getReminder(id) {
        reminders_1.ReminderIdSchema.parse(id);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'GET',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
            });
            return (0, validators_1.validateReminder)(response.data);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Reminder ${id} was not found on the time-based reminder endpoint. If this is a location reminder, use getLocationReminder instead.`);
        }
    }
    async getLocationReminder(id) {
        reminders_1.ReminderIdSchema.parse(id);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'GET',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LOCATION_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
            });
            return (0, validators_1.validateReminder)(response.data);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Location reminder ${id} was not found on the location reminder endpoint. If this is a time-based reminder, use getReminder instead.`);
        }
    }
    async addReminder(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_REMINDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateReminder)(response.data);
    }
    async addLocationReminder(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_LOCATION_REMINDERS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: Object.assign(Object.assign({}, args), { reminderType: reminders_2.ReminderTypeEnum.Location }),
            requestId: requestId,
        });
        return (0, validators_1.validateReminder)(response.data);
    }
    async updateReminder(id, args, requestId) {
        reminders_1.ReminderIdSchema.parse(id);
        const payload = reminders_1.UpdateReminderArgsSchema.parse(args);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'POST',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
                payload,
                requestId: requestId,
            });
            return (0, validators_1.validateReminder)(response.data);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Reminder ${id} was not found on the time-based reminder endpoint. If this is a location reminder, use updateLocationReminder instead.`);
        }
    }
    async updateLocationReminder(id, args, requestId) {
        reminders_1.ReminderIdSchema.parse(id);
        const payload = reminders_1.UpdateLocationReminderArgsSchema.parse(args);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'POST',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LOCATION_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
                payload,
                requestId: requestId,
            });
            return (0, validators_1.validateReminder)(response.data);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Location reminder ${id} was not found on the location reminder endpoint. If this is a time-based reminder, use updateReminder instead.`);
        }
    }
    async deleteReminder(id, requestId) {
        reminders_1.ReminderIdSchema.parse(id);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'DELETE',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
                requestId: requestId,
            });
            return (0, http_client_1.isSuccess)(response);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Reminder ${id} was not found on the time-based reminder endpoint. If this is a location reminder, use deleteLocationReminder instead.`);
        }
    }
    async deleteLocationReminder(id, requestId) {
        reminders_1.ReminderIdSchema.parse(id);
        try {
            const response = await (0, http_client_1.request)({
                httpMethod: 'DELETE',
                baseUri: this.syncApiBase,
                relativePath: (0, request_helpers_1.generatePath)(endpoints_1.ENDPOINT_REST_LOCATION_REMINDERS, id),
                apiToken: this.authToken,
                customFetch: this.customFetch,
                requestId: requestId,
            });
            return (0, http_client_1.isSuccess)(response);
        }
        catch (error) {
            if (!(error instanceof types_1.TodoistRequestError) || error.httpStatusCode !== 404) {
                throw error;
            }
            throw new types_1.TodoistArgumentError(`Location reminder ${id} was not found on the location reminder endpoint. If this is a time-based reminder, use deleteReminder instead.`);
        }
    }
}
exports.ReminderClient = ReminderClient;
