// @doist/todoist-sdk/dist/cjs/clients/template-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const multipart_upload_1 = require("../utils/multipart-upload");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all template-domain endpoints.
 *
 * Instantiated by `TodoistApi`; every public template method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing
 * JSDoc.
 */
class TemplateClient extends base_client_1.BaseClient {
    async exportTemplateAsFile(args) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TEMPLATES_FILE,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return response.data;
    }
    async exportTemplateAsUrl(args) {
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TEMPLATES_URL,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
        });
        return data;
    }
    async createProjectFromTemplate(args, requestId) {
        const { file, fileName, name, workspaceId } = args;
        const additionalFields = { name };
        if (workspaceId !== undefined && workspaceId !== null) {
            additionalFields.workspace_id = workspaceId;
        }
        const data = await (0, multipart_upload_1.uploadMultipartFile)({
            baseUrl: this.syncApiBase,
            authToken: this.authToken,
            endpoint: endpoints_1.ENDPOINT_REST_TEMPLATES_CREATE_FROM_FILE,
            file,
            fileName,
            additionalFields,
            customFetch: this.customFetch,
            requestId,
        });
        return this.validateTemplateResponse(data);
    }
    async importTemplateIntoProject(args, requestId) {
        const { file, fileName, projectId } = args;
        const data = await (0, multipart_upload_1.uploadMultipartFile)({
            baseUrl: this.syncApiBase,
            authToken: this.authToken,
            endpoint: endpoints_1.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_FILE,
            file,
            fileName,
            additionalFields: { project_id: projectId },
            customFetch: this.customFetch,
            requestId,
        });
        return this.validateTemplateResponse(data);
    }
    async importTemplateFromId(args, requestId) {
        const { data } = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.syncApiBase,
            relativePath: endpoints_1.ENDPOINT_REST_TEMPLATES_IMPORT_FROM_ID,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return this.validateTemplateResponse(data);
    }
    validateTemplateResponse(data) {
        var _a, _b, _c, _d;
        return Object.assign(Object.assign({}, data), { projects: (0, validators_1.validateProjectArray)((_a = data.projects) !== null && _a !== void 0 ? _a : []), sections: (0, validators_1.validateSectionArray)((_b = data.sections) !== null && _b !== void 0 ? _b : []), tasks: (0, validators_1.validateTaskArray)((_c = data.tasks) !== null && _c !== void 0 ? _c : []), comments: (0, validators_1.validateCommentArray)((_d = data.comments) !== null && _d !== void 0 ? _d : []) });
    }
}
exports.TemplateClient = TemplateClient;
