// @doist/todoist-sdk/dist/cjs/clients/ui-extension-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UiExtensionClient = void 0;
const endpoints_1 = require("../consts/endpoints");
const http_client_1 = require("../transport/http-client");
const multipart_upload_1 = require("../utils/multipart-upload");
const validators_1 = require("../utils/validators");
const base_client_1 = require("./base-client");
/**
 * Internal sub-client handling all UI-extension endpoints for developer
 * applications. All endpoints require the `dev:app_console` OAuth scope.
 *
 * Instantiated by `TodoistApi`; every public UI-extension method on
 * `TodoistApi` delegates here. See `todoist-api.ts` for user-facing JSDoc.
 */
class UiExtensionClient extends base_client_1.BaseClient {
    async getUiExtension(uiExtensionId, requestId) {
        validators_1.IdSchema.parse(uiExtensionId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getUiExtensionEndpoint)(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateUiExtension)(response.data);
    }
    async getInstalledUiExtensions(requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS_UI_EXTENSIONS_INSTALLED,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateUiExtensionArray)(response.data);
    }
    async getUiExtensionsForApp(appId, requestId) {
        validators_1.IdSchema.parse(appId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'GET',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getUiExtensionsByIntegrationEndpoint)(appId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, validators_1.validateUiExtensionArray)(response.data);
    }
    async addUiExtension(args, requestId) {
        const response = await (0, http_client_1.request)({
            httpMethod: 'POST',
            baseUri: this.apiRootBase,
            relativePath: endpoints_1.ENDPOINT_REST_APPS_UI_EXTENSIONS,
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateUiExtension)(response.data);
    }
    async updateUiExtension(uiExtensionId, args, requestId) {
        validators_1.IdSchema.parse(uiExtensionId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'PATCH',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getUiExtensionEndpoint)(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            payload: args,
            requestId: requestId,
        });
        return (0, validators_1.validateUiExtension)(response.data);
    }
    async deleteUiExtension(uiExtensionId, requestId) {
        validators_1.IdSchema.parse(uiExtensionId);
        const response = await (0, http_client_1.request)({
            httpMethod: 'DELETE',
            baseUri: this.apiRootBase,
            relativePath: (0, endpoints_1.getUiExtensionEndpoint)(uiExtensionId),
            apiToken: this.authToken,
            customFetch: this.customFetch,
            requestId: requestId,
        });
        return (0, http_client_1.isSuccess)(response);
    }
    async uploadUiExtensionIcon(args, requestId) {
        var _a;
        if (Buffer.isBuffer(args.file) && args.file.length === 0) {
            throw new Error('Cannot upload empty image file');
        }
        const data = await (0, multipart_upload_1.uploadMultipartFile)({
            baseUrl: this.apiRootBase,
            authToken: this.authToken,
            endpoint: (0, endpoints_1.getUiExtensionIconEndpoint)(args.uiExtensionId),
            file: args.file,
            fileName: args.fileName,
            additionalFields: { file_name: (_a = args.fileName) !== null && _a !== void 0 ? _a : '' },
            requestId: requestId,
            customFetch: this.customFetch,
        });
        return (0, validators_1.validateUiExtension)(data);
    }
}
exports.UiExtensionClient = UiExtensionClient;
