// @doist/todoist-sdk/dist/cjs/types/reminders/requests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateLocationReminderArgsSchema = exports.UpdateReminderArgsSchema = exports.UpdateAbsoluteReminderArgsSchema = exports.UpdateRelativeReminderArgsSchema = exports.AddLocationReminderArgsSchema = exports.AddReminderArgsSchema = exports.AddAbsoluteReminderArgsSchema = exports.AddRelativeReminderArgsSchema = exports.ReminderDueDateSchema = exports.ReminderIdSchema = exports.ReminderDeliveryServiceSchema = exports.REMINDER_DELIVERY_SERVICES = void 0;
const zod_1 = require("zod");
const reminders_1 = require("../sync/resources/reminders");
const types_1 = require("../tasks/types");
/** Available reminder delivery services. */
exports.REMINDER_DELIVERY_SERVICES = ['email', 'push'];
exports.ReminderDeliveryServiceSchema = zod_1.z.enum(exports.REMINDER_DELIVERY_SERVICES);
exports.ReminderIdSchema = zod_1.z.string();
exports.ReminderDueDateSchema = types_1.DueDateSchema.pick({
    date: true,
    string: true,
    timezone: true,
    lang: true,
    isRecurring: true,
})
    .partial()
    .strict();
// ── Add schemas ──────────────────────────────────────────────────────
exports.AddRelativeReminderArgsSchema = zod_1.z
    .object({
    taskId: zod_1.z.string(),
    reminderType: zod_1.z.literal(reminders_1.ReminderTypeEnum.Relative).optional(),
    minuteOffset: zod_1.z.number().int(),
    service: exports.ReminderDeliveryServiceSchema.optional(),
    notifyUid: zod_1.z.string().optional(),
    isUrgent: zod_1.z.boolean().optional(),
})
    .strict();
exports.AddAbsoluteReminderArgsSchema = zod_1.z
    .object({
    taskId: zod_1.z.string(),
    reminderType: zod_1.z.literal(reminders_1.ReminderTypeEnum.Absolute),
    due: exports.ReminderDueDateSchema,
    service: exports.ReminderDeliveryServiceSchema.optional(),
    notifyUid: zod_1.z.string().optional(),
    isUrgent: zod_1.z.boolean().optional(),
})
    .strict();
exports.AddReminderArgsSchema = zod_1.z.union([
    exports.AddRelativeReminderArgsSchema,
    exports.AddAbsoluteReminderArgsSchema,
]);
exports.AddLocationReminderArgsSchema = zod_1.z
    .object({
    taskId: zod_1.z.string(),
    notifyUid: zod_1.z.string().optional(),
    name: zod_1.z.string(),
    locLat: zod_1.z.string(),
    locLong: zod_1.z.string(),
    locTrigger: zod_1.z.enum(reminders_1.LOCATION_TRIGGERS),
    radius: zod_1.z.number().int().optional(),
})
    .strict();
// ── Update schemas ───────────────────────────────────────────────────
exports.UpdateRelativeReminderArgsSchema = zod_1.z
    .object({
    reminderType: zod_1.z.literal(reminders_1.ReminderTypeEnum.Relative),
    minuteOffset: zod_1.z.number().int().optional(),
    notifyUid: zod_1.z.string().optional(),
    service: exports.ReminderDeliveryServiceSchema.optional(),
    isUrgent: zod_1.z.boolean().optional(),
})
    .strict();
exports.UpdateAbsoluteReminderArgsSchema = zod_1.z
    .object({
    reminderType: zod_1.z.literal(reminders_1.ReminderTypeEnum.Absolute),
    due: exports.ReminderDueDateSchema.optional(),
    notifyUid: zod_1.z.string().optional(),
    service: exports.ReminderDeliveryServiceSchema.optional(),
    isUrgent: zod_1.z.boolean().optional(),
})
    .strict();
exports.UpdateReminderArgsSchema = zod_1.z
    .discriminatedUnion('reminderType', [
    exports.UpdateRelativeReminderArgsSchema,
    exports.UpdateAbsoluteReminderArgsSchema,
])
    .refine((args) => Object.entries(args).some(([key, value]) => key !== 'reminderType' && value !== undefined), {
    message: 'At least one reminder field must be provided to updateReminder',
});
exports.UpdateLocationReminderArgsSchema = zod_1.z
    .object({
    notifyUid: zod_1.z.string().optional(),
    name: zod_1.z.string().optional(),
    locLat: zod_1.z.string().optional(),
    locLong: zod_1.z.string().optional(),
    locTrigger: zod_1.z.enum(reminders_1.LOCATION_TRIGGERS).optional(),
    radius: zod_1.z.number().int().optional(),
})
    .strict()
    .refine((args) => Object.values(args).some((value) => value !== undefined), {
    message: 'At least one reminder field must be provided to updateLocationReminder',
});
