// @doist/todoist-sdk/dist/cjs/types/users/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrentUserSchema = exports.PREMIUM_STATUSES = exports.TimezoneInfoSchema = exports.UserSchema = void 0;
const zod_1 = require("zod");
exports.UserSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    email: zod_1.z.string(),
});
exports.TimezoneInfoSchema = zod_1.z.object({
    gmtString: zod_1.z.string(),
    hours: zod_1.z.number().int(),
    isDst: zod_1.z.number().int(),
    minutes: zod_1.z.number().int(),
    timezone: zod_1.z.string(),
});
/** Available user premium statuses. */
exports.PREMIUM_STATUSES = [
    'not_premium',
    'current_personal_plan',
    'legacy_personal_plan',
    'teams_business_member',
];
exports.CurrentUserSchema = zod_1.z.object({
    id: zod_1.z.string(),
    email: zod_1.z.string(),
    fullName: zod_1.z.string(),
    avatarBig: zod_1.z.string().nullish(),
    avatarMedium: zod_1.z.string().nullish(),
    avatarS640: zod_1.z.string().nullish(),
    avatarSmall: zod_1.z.string().nullish(),
    businessAccountId: zod_1.z.string().nullable(),
    isPremium: zod_1.z.boolean(),
    premiumStatus: zod_1.z.enum(exports.PREMIUM_STATUSES),
    dateFormat: zod_1.z.number().int(),
    timeFormat: zod_1.z.number().int(),
    weeklyGoal: zod_1.z.number().int(),
    dailyGoal: zod_1.z.number().int(),
    completedCount: zod_1.z.number().int(),
    completedToday: zod_1.z.number().int(),
    karma: zod_1.z.number(),
    karmaTrend: zod_1.z.string(),
    lang: zod_1.z.string(),
    nextWeek: zod_1.z.number().int(),
    startDay: zod_1.z.number().int(),
    startPage: zod_1.z.string(),
    tzInfo: exports.TimezoneInfoSchema,
    inboxProjectId: zod_1.z.string(),
    daysOff: zod_1.z.array(zod_1.z.number().int()),
    weekendStartDay: zod_1.z.number().int(),
});
