// @doist/todoist-sdk/dist/cjs/types/sync/resources/calendars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarAccountSchema = exports.CALENDAR_SYNC_STATES = exports.CALENDAR_ACCOUNT_TYPES = exports.CalendarSchema = void 0;
const zod_1 = require("zod");
exports.CalendarSchema = zod_1.z.looseObject({
    id: zod_1.z.string(),
    summary: zod_1.z.string(),
    color: zod_1.z.string().nullable(),
    accountId: zod_1.z.string(),
    isVisible: zod_1.z.boolean(),
    isTaskCalendar: zod_1.z.boolean().optional(),
});
/** Available calendar account provider types. */
exports.CALENDAR_ACCOUNT_TYPES = ['google', 'microsoft', 'apple'];
/** Available calendar sync states. */
exports.CALENDAR_SYNC_STATES = ['synced', 'syncing', 'error'];
exports.CalendarAccountSchema = zod_1.z.looseObject({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    type: zod_1.z.enum(exports.CALENDAR_ACCOUNT_TYPES),
    isDeleted: zod_1.z.boolean().optional(),
    isEventsEnabled: zod_1.z.boolean().optional(),
    isTasksEnabled: zod_1.z.boolean().optional(),
    isAllDayTasksEnabled: zod_1.z.boolean().optional(),
    pendingOperationUntil: zod_1.z.string().nullable().optional(),
    calendarsSyncState: zod_1.z.enum(exports.CALENDAR_SYNC_STATES).optional(),
});
