// @doist/todoist-sdk/dist/cjs/types/sync/resources/user-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSettingsSchema = exports.QuickAddFeatureSchema = exports.NavigationFeatureSchema = void 0;
const zod_1 = require("zod");
exports.NavigationFeatureSchema = zod_1.z.object({
    name: zod_1.z.string(),
    shown: zod_1.z.boolean(),
});
exports.QuickAddFeatureSchema = zod_1.z.object({
    name: zod_1.z.string(),
    shown: zod_1.z.boolean(),
});
/**
 * Sync API user settings resource.
 */
exports.UserSettingsSchema = zod_1.z.looseObject({
    completedSoundDesktop: zod_1.z.boolean(),
    completedSoundMobile: zod_1.z.boolean(),
    debugLogSendingEnabledUntil: zod_1.z.string().nullish(),
    legacyPricing: zod_1.z.union([zod_1.z.boolean(), zod_1.z.literal(0), zod_1.z.literal(1)]).transform(Boolean),
    navigation: zod_1.z
        .object({
        countsShown: zod_1.z.boolean(),
        features: zod_1.z.array(exports.NavigationFeatureSchema),
    })
        .loose(),
    reminderDesktop: zod_1.z.boolean(),
    reminderEmail: zod_1.z.boolean(),
    reminderPush: zod_1.z.boolean(),
    resetRecurringSubtasks: zod_1.z.boolean(),
    aiEmailAssist: zod_1.z.boolean(),
    theme: zod_1.z.string().nullable().optional(),
    syncTheme: zod_1.z.boolean().optional(),
    quickAdd: zod_1.z
        .object({
        labelsShown: zod_1.z.boolean(),
        features: zod_1.z.array(exports.QuickAddFeatureSchema),
    })
        .loose(),
});
