// @doist/todoist-sdk/dist/cjs/types/sync/resources/view-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectViewOptionsDefaultsSchema = exports.ViewOptionsSchema = exports.CalendarSettingsSchema = exports.CALENDAR_LAYOUTS = exports.SortOrderSchema = exports.SORT_ORDERS = exports.SortedBySchema = exports.SORTED_BY_OPTIONS = exports.GroupedBySchema = exports.GROUPED_BY_OPTIONS = exports.ViewModeSchema = exports.VIEW_MODES = exports.ViewTypeSchema = exports.VIEW_TYPES = void 0;
const zod_1 = require("zod");
/** Available view types. */
exports.VIEW_TYPES = [
    'TODAY',
    'UPCOMING',
    'PROJECT',
    'LABEL',
    'FILTER',
    'WORKSPACE_FILTER',
    'SEARCH',
    'TEMPLATE_PREVIEW',
    'TASK_DETAIL',
    'AUTOMATION',
    'ASSIGNED',
    'OVERDUE',
    'WORKSPACE_OVERVIEW',
];
exports.ViewTypeSchema = zod_1.z.enum(exports.VIEW_TYPES);
/** Available view modes. */
exports.VIEW_MODES = ['LIST', 'BOARD', 'CALENDAR'];
exports.ViewModeSchema = zod_1.z.enum(exports.VIEW_MODES);
/** Available grouping options. */
exports.GROUPED_BY_OPTIONS = [
    'ASSIGNEE',
    'ADDED_DATE',
    'DUE_DATE',
    'DEADLINE',
    'LABEL',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
];
exports.GroupedBySchema = zod_1.z.enum(exports.GROUPED_BY_OPTIONS).nullable();
/** Available sorting options. */
exports.SORTED_BY_OPTIONS = [
    'MANUAL',
    'ALPHABETICALLY',
    'ASSIGNEE',
    'DUE_DATE',
    'DEADLINE',
    'ADDED_DATE',
    'PRIORITY',
    'PROJECT',
    'WORKSPACE',
];
exports.SortedBySchema = zod_1.z.enum(exports.SORTED_BY_OPTIONS).nullable();
/** Available sort directions. */
exports.SORT_ORDERS = ['ASC', 'DESC'];
exports.SortOrderSchema = zod_1.z.enum(exports.SORT_ORDERS).nullable();
/** Available calendar layout modes. */
exports.CALENDAR_LAYOUTS = ['WEEK', 'MONTH'];
exports.CalendarSettingsSchema = zod_1.z.looseObject({
    layout: zod_1.z.enum(exports.CALENDAR_LAYOUTS).optional(),
});
exports.ViewOptionsSchema = zod_1.z.looseObject({
    viewType: exports.ViewTypeSchema,
    objectId: zod_1.z.string().optional(),
    groupedBy: exports.GroupedBySchema.optional(),
    filteredBy: zod_1.z.string().nullable().optional(),
    viewMode: exports.ViewModeSchema.optional(),
    showCompletedTasks: zod_1.z.boolean().optional(),
    sortedBy: exports.SortedBySchema.optional(),
    sortOrder: exports.SortOrderSchema.optional(),
});
exports.ProjectViewOptionsDefaultsSchema = zod_1.z.looseObject({
    projectId: zod_1.z.string(),
    viewMode: exports.ViewModeSchema.nullable().optional(),
    groupedBy: exports.GroupedBySchema.optional(),
    sortedBy: exports.SortedBySchema.optional(),
    sortOrder: exports.SortOrderSchema.optional(),
    showCompletedTasks: zod_1.z.boolean().optional(),
    filteredBy: zod_1.z.string().nullable().optional(),
    calendarSettings: exports.CalendarSettingsSchema.nullable().optional(),
});
