// @doist/todoist-sdk/dist/cjs/types/sync/resources/workspaces.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncWorkspaceSchema = void 0;
const zod_1 = require("zod");
const common_1 = require("../../common");
const types_1 = require("../../workspaces/types");
/**
 * Sync API workspace resource.
 *
 * This is a superset of the REST `WorkspaceSchema` — the Sync API returns
 * additional fields like description, isDeleted, isCollapsed, member counts,
 * domain settings, and sorting preferences.
 */
exports.SyncWorkspaceSchema = zod_1.z
    .looseObject({
    id: common_1.StringOrNumberSchema,
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    logoBig: zod_1.z.string().nullish(),
    logoMedium: zod_1.z.string().nullish(),
    logoSmall: zod_1.z.string().nullish(),
    logoS640: zod_1.z.string().nullish(),
    creatorId: common_1.StringOrNumberSchema,
    createdAt: zod_1.z.coerce.date().optional(),
    dateCreated: zod_1.z.coerce.date().optional(),
    isDeleted: zod_1.z.boolean(),
    isCollapsed: zod_1.z.boolean(),
    role: types_1.WorkspaceRoleSchema.optional(),
    plan: types_1.WorkspacePlanSchema,
    limits: types_1.WorkspaceLimitsSchema.optional(),
    inviteCode: zod_1.z.string().nullable().optional(),
    isLinkSharingEnabled: zod_1.z.boolean().nullable().optional(),
    isGuestAllowed: zod_1.z.boolean().nullable().optional(),
    currentActiveProjects: zod_1.z.number().nullable(),
    currentMemberCount: zod_1.z.number().nullable(),
    memberCountByType: zod_1.z
        .object({
        adminCount: zod_1.z.number().int(),
        guestCount: zod_1.z.number().int(),
        memberCount: zod_1.z.number().int(),
    })
        .optional(),
    currentTemplateCount: zod_1.z.number().nullable(),
    pendingInvitations: zod_1.z.array(zod_1.z.string()).nullable().optional(),
    domainName: zod_1.z.string().nullish(),
    domainDiscovery: zod_1.z.boolean().nullish(),
    restrictEmailDomains: zod_1.z.boolean().nullish(),
    adminSortingApplied: zod_1.z.boolean(),
    projectSortPreference: zod_1.z.string().nullish(),
    defaultCollaborators: zod_1.z
        .object({
        predefinedGroupIds: zod_1.z.array(zod_1.z.string()),
        userIds: zod_1.z.array(common_1.StringOrNumberSchema),
    })
        .optional(),
    properties: types_1.WorkspacePropertiesSchema.optional(),
})
    .transform((_a) => {
    var { dateCreated, createdAt } = _a, rest = __rest(_a, ["dateCreated", "createdAt"]);
    return (Object.assign(Object.assign({}, rest), { createdAt: createdAt !== null && createdAt !== void 0 ? createdAt : dateCreated }));
});
