// @doist/todoist-sdk/dist/cjs/types/tasks/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskSchema = exports.DeadlineSchema = exports.DurationSchema = exports.DURATION_UNITS = exports.DueDateSchema = void 0;
const zod_1 = require("zod");
const uncompletable_helpers_1 = require("../../utils/uncompletable-helpers");
const url_helpers_1 = require("../../utils/url-helpers");
exports.DueDateSchema = zod_1.z
    .object({
    isRecurring: zod_1.z.boolean(),
    string: zod_1.z.string(),
    date: zod_1.z.string(),
})
    .extend({
    datetime: zod_1.z.string().nullable().optional(),
    timezone: zod_1.z.string().nullable().optional(),
    lang: zod_1.z.string().nullable().optional(),
});
/** Available duration units for task deadlines. */
exports.DURATION_UNITS = ['minute', 'day'];
exports.DurationSchema = zod_1.z.object({
    amount: zod_1.z.number().positive('Value should be greater than zero'),
    unit: zod_1.z.enum(exports.DURATION_UNITS),
});
exports.DeadlineSchema = zod_1.z.object({
    date: zod_1.z.string(),
    lang: zod_1.z.string(),
});
exports.TaskSchema = zod_1.z
    .object({
    id: zod_1.z.string(),
    userId: zod_1.z.string(),
    projectId: zod_1.z.string(),
    sectionId: zod_1.z.string().nullable(),
    parentId: zod_1.z.string().nullable(),
    addedByUid: zod_1.z.string().nullable(),
    assignedByUid: zod_1.z.string().nullable(),
    responsibleUid: zod_1.z.string().nullable(),
    labels: zod_1.z.array(zod_1.z.string()),
    deadline: exports.DeadlineSchema.nullable(),
    duration: exports.DurationSchema.nullable(),
    checked: zod_1.z.boolean(),
    isDeleted: zod_1.z.boolean(),
    addedAt: zod_1.z.coerce.date().nullable(),
    completedAt: zod_1.z.coerce.date().nullable(),
    updatedAt: zod_1.z.coerce.date().nullable(),
    due: exports.DueDateSchema.nullable(),
    priority: zod_1.z.number().int(),
    childOrder: zod_1.z.number().int(),
    content: zod_1.z.string(),
    description: zod_1.z.string(),
    dayOrder: zod_1.z.number().int(),
    isCollapsed: zod_1.z.boolean(),
    isUncompletable: zod_1.z.boolean().default(false),
})
    .transform((data) => {
    // Auto-detect uncompletable status from content prefix
    const isUncompletable = (0, uncompletable_helpers_1.hasUncompletablePrefix)(data.content);
    return Object.assign(Object.assign({}, data), { isUncompletable, url: (0, url_helpers_1.getTaskUrl)(data.id, data.content) });
});
