// @doist/todoist-sdk/dist/esm/types/reminders/requests.js
import { z } from 'zod';
import { LOCATION_TRIGGERS, ReminderTypeEnum } from '../sync/resources/reminders.js';
import { DueDateSchema } from '../tasks/types.js';
/** Available reminder delivery services. */
export const REMINDER_DELIVERY_SERVICES = ['email', 'push'];
export const ReminderDeliveryServiceSchema = z.enum(REMINDER_DELIVERY_SERVICES);
export const ReminderIdSchema = z.string();
export const ReminderDueDateSchema = DueDateSchema.pick({
    date: true,
    string: true,
    timezone: true,
    lang: true,
    isRecurring: true,
})
    .partial()
    .strict();
// ── Add schemas ──────────────────────────────────────────────────────
export const AddRelativeReminderArgsSchema = z
    .object({
    taskId: z.string(),
    reminderType: z.literal(ReminderTypeEnum.Relative).optional(),
    minuteOffset: z.number().int(),
    service: ReminderDeliveryServiceSchema.optional(),
    notifyUid: z.string().optional(),
    isUrgent: z.boolean().optional(),
})
    .strict();
export const AddAbsoluteReminderArgsSchema = z
    .object({
    taskId: z.string(),
    reminderType: z.literal(ReminderTypeEnum.Absolute),
    due: ReminderDueDateSchema,
    service: ReminderDeliveryServiceSchema.optional(),
    notifyUid: z.string().optional(),
    isUrgent: z.boolean().optional(),
})
    .strict();
export const AddReminderArgsSchema = z.union([
    AddRelativeReminderArgsSchema,
    AddAbsoluteReminderArgsSchema,
]);
export const AddLocationReminderArgsSchema = z
    .object({
    taskId: z.string(),
    notifyUid: z.string().optional(),
    name: z.string(),
    locLat: z.string(),
    locLong: z.string(),
    locTrigger: z.enum(LOCATION_TRIGGERS),
    radius: z.number().int().optional(),
})
    .strict();
// ── Update schemas ───────────────────────────────────────────────────
export const UpdateRelativeReminderArgsSchema = z
    .object({
    reminderType: z.literal(ReminderTypeEnum.Relative),
    minuteOffset: z.number().int().optional(),
    notifyUid: z.string().optional(),
    service: ReminderDeliveryServiceSchema.optional(),
    isUrgent: z.boolean().optional(),
})
    .strict();
export const UpdateAbsoluteReminderArgsSchema = z
    .object({
    reminderType: z.literal(ReminderTypeEnum.Absolute),
    due: ReminderDueDateSchema.optional(),
    notifyUid: z.string().optional(),
    service: ReminderDeliveryServiceSchema.optional(),
    isUrgent: z.boolean().optional(),
})
    .strict();
export const UpdateReminderArgsSchema = z
    .discriminatedUnion('reminderType', [
    UpdateRelativeReminderArgsSchema,
    UpdateAbsoluteReminderArgsSchema,
])
    .refine((args) => Object.entries(args).some(([key, value]) => key !== 'reminderType' && value !== undefined), {
    message: 'At least one reminder field must be provided to updateReminder',
});
export const UpdateLocationReminderArgsSchema = z
    .object({
    notifyUid: z.string().optional(),
    name: z.string().optional(),
    locLat: z.string().optional(),
    locLong: z.string().optional(),
    locTrigger: z.enum(LOCATION_TRIGGERS).optional(),
    radius: z.number().int().optional(),
})
    .strict()
    .refine((args) => Object.values(args).some((value) => value !== undefined), {
    message: 'At least one reminder field must be provided to updateLocationReminder',
});
