// @doist/todoist-sdk/dist/cjs/types/apps/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserAuthorizationSchema = exports.UserAuthorizationAppSchema = exports.AppInstallationSchema = exports.InstalledIntegrationSchema = exports.AppWebhookSchema = exports.AppByDistributionTokenSchema = exports.AppByDistributionTokenUiExtensionSchema = exports.AppDistributionTokenSchema = exports.AppTestTokenSchema = exports.AppVerificationTokenSchema = exports.AppSecretsSchema = exports.AppWithUserCountSchema = exports.AppSchema = exports.AppTokenScopesSchema = exports.InstallationTypeSchema = exports.INSTALLATION_TYPES = exports.WebhookVersionSchema = exports.WEBHOOK_VERSIONS = exports.WebhookStatusSchema = exports.WEBHOOK_STATUSES = exports.WebhookEventWireNameSchema = exports.WEBHOOK_EVENT_WIRE_NAMES = exports.WebhookEventSchema = exports.WEBHOOK_EVENTS = exports.AppStatusSchema = exports.APP_STATUSES = void 0;
exports.webhookEventFromWireName = webhookEventFromWireName;
exports.webhookEventToWireName = webhookEventToWireName;
const zod_1 = require("zod");
const common_1 = require("../common");
const types_1 = require("../ui-extensions/types");
/**
 * Available statuses for a developer application.
 */
exports.APP_STATUSES = ['public'];
exports.AppStatusSchema = zod_1.z.enum(exports.APP_STATUSES);
/**
 * Webhook event types that can be subscribed to.
 *
 * These are the friendly public names the backend uses in webhook payloads.
 * The on-the-wire config endpoint uses a different (legacy) form (see
 * {@link WEBHOOK_EVENT_WIRE_NAMES}); the SDK converts between the two
 * so consumers only ever work with these pretty names.
 *
 * @see https://developer.todoist.com/sync/v1/#webhooks
 */
exports.WEBHOOK_EVENTS = [
    'item:added',
    'item:updated',
    'item:completed',
    'item:uncompleted',
    'item:deleted',
    'note:added',
    'note:updated',
    'note:deleted',
    'project:added',
    'project:updated',
    'project:deleted',
    'project:archived',
    'project:unarchived',
    'section:added',
    'section:updated',
    'section:deleted',
    'section:archived',
    'section:unarchived',
    'label:added',
    'label:updated',
    'label:deleted',
    'filter:added',
    'filter:updated',
    'filter:deleted',
    'reminder:fired',
];
exports.WebhookEventSchema = zod_1.z.enum(exports.WEBHOOK_EVENTS);
/**
 * Raw wire-format webhook event names used by the config endpoint
 * (`POST/GET /apps/<id>/webhook`). Exported for completeness — prefer
 * {@link WEBHOOK_EVENTS} when working with the SDK.
 */
exports.WEBHOOK_EVENT_WIRE_NAMES = [
    'e_item_added',
    'e_item_updated',
    'e_item_completed',
    'e_item_uncompleted',
    'e_item_deleted',
    'e_note_added',
    'e_note_updated',
    'e_note_deleted',
    'e_project_added',
    'e_project_updated',
    'e_project_deleted',
    'e_project_archived',
    'e_project_unarchived',
    'e_section_added',
    'e_section_updated',
    'e_section_deleted',
    'e_section_archived',
    'e_section_unarchived',
    'e_label_added',
    'e_label_updated',
    'e_label_deleted',
    'e_filter_added',
    'e_filter_updated',
    'e_filter_deleted',
    'e_reminder_fired',
];
exports.WebhookEventWireNameSchema = zod_1.z.enum(exports.WEBHOOK_EVENT_WIRE_NAMES);
/**
 * Translates a wire-format event name (`'e_item_added'`) to its pretty form (`'item:added'`).
 */
function webhookEventFromWireName(wireName) {
    // Strip leading `e_`, replace first `_` with `:`. The resulting string is
    // guaranteed to be a `WebhookEvent` by construction of the two const arrays.
    return wireName.slice(2).replace('_', ':');
}
/**
 * Translates a pretty event name (`'item:added'`) to its wire-format form (`'e_item_added'`).
 */
function webhookEventToWireName(event) {
    return ('e_' + event.replace(':', '_'));
}
/**
 * Available statuses for an app webhook.
 */
exports.WEBHOOK_STATUSES = ['active', 'invalid'];
exports.WebhookStatusSchema = zod_1.z.enum(exports.WEBHOOK_STATUSES);
/**
 * Supported webhook payload versions. v9 is deprecated and intentionally excluded.
 */
exports.WEBHOOK_VERSIONS = ['1'];
exports.WebhookVersionSchema = zod_1.z.enum(exports.WEBHOOK_VERSIONS);
/**
 * Available installation types for an app installation.
 */
exports.INSTALLATION_TYPES = ['user'];
exports.InstallationTypeSchema = zod_1.z.enum(exports.INSTALLATION_TYPES);
/**
 * Transforms a comma-separated scope string into `string[] | null`.
 *
 * The backend stores `app_token_scopes` as a single comma-separated string
 * (e.g. `"data:read,task:add"`). Reads are exposed as `string[]` (rather than
 * `Permission[]`) so that newly added backend scopes don't break parsing —
 * writes accept `Permission` for autocomplete while still permitting `string`
 * (via `UpdateAppArgs`), so a fetched `App.appTokenScopes` can be round-tripped
 * back into `updateApp()` without casts.
 *
 * Preserves `null` on read (backend returns `null` for apps that have had
 * their scopes cleared via `updateApp({ appTokenScopes: null })`) so the
 * read and write shapes line up and information is not lost.
 */
exports.AppTokenScopesSchema = zod_1.z
    .string()
    .nullable()
    .transform((value) => {
    if (value === null)
        return null;
    if (value === '')
        return [];
    return value.split(',').filter(Boolean);
});
/**
 * Schema for a developer application.
 * @see https://developer.todoist.com/api/v1/#tag/Apps
 */
exports.AppSchema = zod_1.z.object({
    id: common_1.StringOrNumberSchema,
    clientId: zod_1.z.string(),
    status: exports.AppStatusSchema,
    displayName: zod_1.z.string(),
    userId: common_1.StringOrNumberSchema,
    createdAt: zod_1.z.coerce.date(),
    serviceUrl: zod_1.z.string().nullable(),
    oauthRedirectUri: zod_1.z.string().nullable(),
    description: zod_1.z.string().nullable(),
    iconSm: zod_1.z.string().nullable(),
    iconMd: zod_1.z.string().nullable(),
    iconLg: zod_1.z.string().nullable(),
    appTokenScopes: exports.AppTokenScopesSchema,
});
/**
 * Schema for a developer application that includes the user count.
 * Returned by most endpoints that manage a single app (get, update, delete).
 */
exports.AppWithUserCountSchema = exports.AppSchema.extend({
    userCount: zod_1.z.number().int(),
});
/**
 * OAuth client credentials for a developer application.
 */
exports.AppSecretsSchema = zod_1.z.object({
    clientId: zod_1.z.string(),
    clientSecret: zod_1.z.string(),
});
/**
 * Webhook verification token for a developer application.
 */
exports.AppVerificationTokenSchema = zod_1.z.object({
    verificationToken: zod_1.z.string(),
});
/**
 * Developer test access token for a developer application.
 */
exports.AppTestTokenSchema = zod_1.z.object({
    accessToken: zod_1.z.string().nullish(),
});
/**
 * Distribution token for a developer application.
 */
exports.AppDistributionTokenSchema = zod_1.z.object({
    distributionToken: zod_1.z.string(),
});
/**
 * Subset of an app returned when resolving by distribution token — used by the installation flow.
 * Includes a lightweight list of the app's UI extensions (not full `UIExtension` records).
 */
exports.AppByDistributionTokenUiExtensionSchema = zod_1.z.object({
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    icon: zod_1.z.string().nullable(),
    extensionType: zod_1.z.enum(types_1.UI_EXTENSION_TYPES),
});
exports.AppByDistributionTokenSchema = zod_1.z.object({
    id: common_1.StringOrNumberSchema,
    displayName: zod_1.z.string(),
    description: zod_1.z.string().nullable(),
    iconMd: zod_1.z.string().nullable(),
    appTokenScopes: exports.AppTokenScopesSchema,
    uiExtensions: zod_1.z.array(exports.AppByDistributionTokenUiExtensionSchema),
});
/**
 * Schema for an app webhook.
 *
 * Events arrive from the wire in their `e_*` form and are transformed to
 * their pretty `item:added`-style names for SDK consumers.
 */
exports.AppWebhookSchema = zod_1.z.object({
    status: exports.WebhookStatusSchema,
    callbackUrl: zod_1.z.string(),
    version: exports.WebhookVersionSchema,
    events: zod_1.z.array(exports.WebhookEventWireNameSchema.transform(webhookEventFromWireName)),
});
/**
 * Lightweight view of an app embedded in an installation record.
 * Mirrors `AppSchema` shape — backend returns the same `ApplicationRestView`.
 */
exports.InstalledIntegrationSchema = exports.AppSchema;
/**
 * Schema for an installed application (integration installed to a user's account).
 */
exports.AppInstallationSchema = zod_1.z.object({
    id: common_1.StringOrNumberSchema,
    creator: common_1.StringOrNumberSchema,
    createdTs: zod_1.z.number().int(),
    installationType: exports.InstallationTypeSchema,
    appTokenScopesValidated: zod_1.z.boolean(),
    integration: exports.InstalledIntegrationSchema.nullable(),
});
/**
 * Lightweight app info included on a user authorization record.
 */
exports.UserAuthorizationAppSchema = zod_1.z.object({
    id: common_1.StringOrNumberSchema,
    displayName: zod_1.z.string(),
    description: zod_1.z.string().nullable(),
    serviceUrl: zod_1.z.string().nullable(),
    iconMd: zod_1.z.string().nullable(),
});
/**
 * Schema for a user authorization entry — a token the user has granted to a third-party app.
 */
exports.UserAuthorizationSchema = zod_1.z.object({
    accessTokenId: common_1.StringOrNumberSchema,
    scope: zod_1.z.array(zod_1.z.string()),
    scopeDescriptions: zod_1.z.array(zod_1.z.string()),
    createdAt: zod_1.z.coerce.date(),
    app: exports.UserAuthorizationAppSchema.nullable(),
});
